{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "---\n",
    "description: Functions related to generating spectra from FASTA files\n",
    "output-file: fasta.html\n",
    "title: FASTA\n",
    "\n",
    "---\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook contains all functions related to creating spectra from FASTA files. In brief, what we are doing is the following:\n",
    "\n",
    "1. Read a FASTA file and in silico digest the proteins to generate peptides\n",
    "2. For each peptide, calculate a theoretical spectrum and precursor mass\n",
    "3. Save spectra\n",
    "\n",
    "Currently, `numba` has only limited string support. A lot of the functions are therefore Python-native."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cleaving\n",
    "\n",
    "We use regular expressions to find potential cleavage sites for cleaving and write the wrapper [`cleave_sequence`](https://mannlabs.github.io/alphapept/fasta.html#cleave_sequence) to use it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L33){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### cleave_sequence\n",
       "\n",
       ">      cleave_sequence (sequence:str='', n_missed_cleavages:int=0,\n",
       ">                       protease:str='trypsin', pep_length_min:int=6,\n",
       ">                       pep_length_max:int=65, **kwargs)\n",
       "\n",
       "Cleave a sequence with a given protease. Filters to have a minimum and maximum length.\n",
       "Args:\n",
       "    sequence (str): the given (protein) sequence.\n",
       "    n_missed_cleavages (int): the number of max missed cleavages.\n",
       "    protease (str): the protease/enzyme name, the regular expression can be found in alphapept.constants.protease_dict.\n",
       "    pep_length_min (int): min peptide length.\n",
       "    pep_length_max (int): max peptide length.\n",
       "Returns:\n",
       "    list (of str): cleaved peptide sequences with missed cleavages."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L33){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### cleave_sequence\n",
       "\n",
       ">      cleave_sequence (sequence:str='', n_missed_cleavages:int=0,\n",
       ">                       protease:str='trypsin', pep_length_min:int=6,\n",
       ">                       pep_length_max:int=65, **kwargs)\n",
       "\n",
       "Cleave a sequence with a given protease. Filters to have a minimum and maximum length.\n",
       "Args:\n",
       "    sequence (str): the given (protein) sequence.\n",
       "    n_missed_cleavages (int): the number of max missed cleavages.\n",
       "    protease (str): the protease/enzyme name, the regular expression can be found in alphapept.constants.protease_dict.\n",
       "    pep_length_min (int): min peptide length.\n",
       "    pep_length_max (int): max peptide length.\n",
       "Returns:\n",
       "    list (of str): cleaved peptide sequences with missed cleavages."
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(cleave_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L17){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_missed_cleavages\n",
       "\n",
       ">      get_missed_cleavages (sequences:list, n_missed_cleavages:int)\n",
       "\n",
       "Combine cleaved sequences to get sequences with missed cleavages\n",
       "Args:\n",
       "    seqeuences (list of str): the list of cleaved sequences, no missed cleavages are there.\n",
       "    n_missed_cleavages (int): the number of miss cleavage sites.\n",
       "Returns:\n",
       "    list (of str): the sequences with missed cleavages."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L17){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_missed_cleavages\n",
       "\n",
       ">      get_missed_cleavages (sequences:list, n_missed_cleavages:int)\n",
       "\n",
       "Combine cleaved sequences to get sequences with missed cleavages\n",
       "Args:\n",
       "    seqeuences (list of str): the list of cleaved sequences, no missed cleavages are there.\n",
       "    n_missed_cleavages (int): the number of miss cleavage sites.\n",
       "Returns:\n",
       "    list (of str): the sequences with missed cleavages."
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(get_missed_cleavages)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ABCDEFGHIJK', 'LMNOPQR']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protease = \"trypsin\"\n",
    "n_missed_cleavages = 0\n",
    "pep_length_min, pep_length_max = 6, 65\n",
    "\n",
    "cleave_sequence('ABCDEFGHIJKLMNOPQRST', n_missed_cleavages, protease, pep_length_min, pep_length_max)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Counting missed and internal cleavages\n",
    "The following are helper functions to retrieve the number of missed cleavages and internal cleavage sites for each sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L92){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### count_internal_cleavages\n",
       "\n",
       ">      count_internal_cleavages (sequence:str='', protease:str='trypsin',\n",
       ">                                **kwargs)\n",
       "\n",
       "Counts the number of internal cleavage sites for a given sequence and protease\n",
       "Args:\n",
       "    sequence (str): the given (peptide) sequence.\n",
       "    protease (str): the protease/enzyme name, the regular expression can be found in alphapept.constants.protease_dict.\n",
       "Returns:\n",
       "    int (0 or 1): if the sequence is from internal cleavage."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L92){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### count_internal_cleavages\n",
       "\n",
       ">      count_internal_cleavages (sequence:str='', protease:str='trypsin',\n",
       ">                                **kwargs)\n",
       "\n",
       "Counts the number of internal cleavage sites for a given sequence and protease\n",
       "Args:\n",
       "    sequence (str): the given (peptide) sequence.\n",
       "    protease (str): the protease/enzyme name, the regular expression can be found in alphapept.constants.protease_dict.\n",
       "Returns:\n",
       "    int (0 or 1): if the sequence is from internal cleavage."
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(count_internal_cleavages)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L77){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### count_missed_cleavages\n",
       "\n",
       ">      count_missed_cleavages (sequence:str='', protease:str='trypsin',\n",
       ">                              **kwargs)\n",
       "\n",
       "Counts the number of missed cleavages for a given sequence and protease\n",
       "Args:\n",
       "    sequence (str): the given (peptide) sequence.\n",
       "    protease (str): the protease/enzyme name, the regular expression can be found in alphapept.constants.protease_dict.\n",
       "Returns:\n",
       "    int: the number of miss cleavages"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L77){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### count_missed_cleavages\n",
       "\n",
       ">      count_missed_cleavages (sequence:str='', protease:str='trypsin',\n",
       ">                              **kwargs)\n",
       "\n",
       "Counts the number of missed cleavages for a given sequence and protease\n",
       "Args:\n",
       "    sequence (str): the given (peptide) sequence.\n",
       "    protease (str): the protease/enzyme name, the regular expression can be found in alphapept.constants.protease_dict.\n",
       "Returns:\n",
       "    int: the number of miss cleavages"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(count_missed_cleavages)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "protease = \"trypsin\"\n",
    "print(count_missed_cleavages('ABCDEFGHIJKLMNOPQRST', protease))\n",
    "\n",
    "protease = \"trypsin\"\n",
    "print(count_internal_cleavages('ABCDEFGHIJKLMNOPQRST', protease))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parsing\n",
    "\n",
    "Peptides are composed out of amino acids that are written in capital letters - `PEPTIDE`. To distinguish modifications, they are written in lowercase such as `PEPTIoxDE` and can be of arbitrary length. For a modified amino acid (AA), the modification precedes the letter of the amino acid. Decoys are indicated with an underscore. Therefore, the [`parse`](https://mannlabs.github.io/alphapept/fasta.html#parse) function splits after `_`. When parsing, the peptide string is converted into a `numba`-compatible list, like so: `PEPoxTIDE` -> `[P, E, P, oxT, I, D, E]`. This allows that we can use the `mass_dict` from `alphapept.constants` to directly determine the masses for the corresponding amino acids."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L136){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### list_to_numba\n",
       "\n",
       ">      list_to_numba (a_list)\n",
       "\n",
       "Convert Python list to numba.typed.List\n",
       "Args: \n",
       "    a_list (list): Python list.\n",
       "Return:\n",
       "    List (numba.typed.List): Numba typed list."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L136){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### list_to_numba\n",
       "\n",
       ">      list_to_numba (a_list)\n",
       "\n",
       "Convert Python list to numba.typed.List\n",
       "Args: \n",
       "    a_list (list): Python list.\n",
       "Return:\n",
       "    List (numba.typed.List): Numba typed list."
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(list_to_numba)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L115){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### parse\n",
       "\n",
       ">      parse (peptide:str)\n",
       "\n",
       "Parser to parse peptide strings\n",
       "Args: \n",
       "    peptide (str): modified peptide sequence.\n",
       "Return:\n",
       "    List (numba.typed.List): a list of animo acids and modified amono acids"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L115){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### parse\n",
       "\n",
       ">      parse (peptide:str)\n",
       "\n",
       "Parser to parse peptide strings\n",
       "Args: \n",
       "    peptide (str): modified peptide sequence.\n",
       "Return:\n",
       "    List (numba.typed.List): a list of animo acids and modified amono acids"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(parse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[P, E, P, T, I, D, E, ...]\n",
      "[P, E, P, oxT, I, D, E, ...]\n"
     ]
    }
   ],
   "source": [
    "print(parse('PEPTIDE'))\n",
    "print(parse('PEPoxTIDE'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Decoy\n",
    "\n",
    "The decoy strategy employed is a pseudo-reversal of the peptide sequence, keeping only the terminal amino acid and reversing the rest. Additionally, we can call the functions [`swap_KR`](https://mannlabs.github.io/alphapept/fasta.html#swap_kr) and and [`swap_AL`](https://mannlabs.github.io/alphapept/fasta.html#swap_al) that will swap the respective AAs. The function [`swap_KR`](https://mannlabs.github.io/alphapept/fasta.html#swap_kr) will only swap terminal AAs. The swapping functions only work if the AA is not modified."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L240){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_decoy_tag\n",
       "\n",
       ">      add_decoy_tag (peptides)\n",
       "\n",
       "Adds a '_decoy' tag to a list of peptides"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L240){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_decoy_tag\n",
       "\n",
       ">      add_decoy_tag (peptides)\n",
       "\n",
       "Adds a '_decoy' tag to a list of peptides"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(add_decoy_tag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L225){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_decoys\n",
       "\n",
       ">      get_decoys (peptide_list, pseudo_reverse=False, AL_swap=False,\n",
       ">                  KR_swap=False, **kwargs)\n",
       "\n",
       "Wrapper to get decoys for lists of peptides\n",
       "Args:\n",
       "    peptide_list (list): the list of peptides to be reversed.\n",
       "    pseudo_reverse (bool): If True, reverse the peptide bug keep the C-terminal amino acid; otherwise reverse the whole peptide. (Default: False)\n",
       "    AL_swap (bool): replace A with L, and vice versa. (Default: False)\n",
       "    KR_swap (bool): replace K with R at the C-terminal, and vice versa. (Default: False)\n",
       "Returns:\n",
       "    list (of str): a list of decoy peptides"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L225){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_decoys\n",
       "\n",
       ">      get_decoys (peptide_list, pseudo_reverse=False, AL_swap=False,\n",
       ">                  KR_swap=False, **kwargs)\n",
       "\n",
       "Wrapper to get decoys for lists of peptides\n",
       "Args:\n",
       "    peptide_list (list): the list of peptides to be reversed.\n",
       "    pseudo_reverse (bool): If True, reverse the peptide bug keep the C-terminal amino acid; otherwise reverse the whole peptide. (Default: False)\n",
       "    AL_swap (bool): replace A with L, and vice versa. (Default: False)\n",
       "    KR_swap (bool): replace K with R at the C-terminal, and vice versa. (Default: False)\n",
       "Returns:\n",
       "    list (of str): a list of decoy peptides"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(get_decoys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L202){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### swap_AL\n",
       "\n",
       ">      swap_AL (peptide:str)\n",
       "\n",
       "Swaps a A with L. Note: Only if AA is not modified.\n",
       "Args:\n",
       "    peptide (str): peptide.\n",
       "\n",
       "Returns:\n",
       "    str: peptide with swapped ALs."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L202){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### swap_AL\n",
       "\n",
       ">      swap_AL (peptide:str)\n",
       "\n",
       "Swaps a A with L. Note: Only if AA is not modified.\n",
       "Args:\n",
       "    peptide (str): peptide.\n",
       "\n",
       "Returns:\n",
       "    str: peptide with swapped ALs."
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(swap_AL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L183){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### swap_KR\n",
       "\n",
       ">      swap_KR (peptide:str)\n",
       "\n",
       "Swaps a terminal K or R. Note: Only if AA is not modified.\n",
       "\n",
       "Args:\n",
       "    peptide (str): peptide.\n",
       "\n",
       "Returns:\n",
       "    str: peptide with swapped KRs."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L183){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### swap_KR\n",
       "\n",
       ">      swap_KR (peptide:str)\n",
       "\n",
       "Swaps a terminal K or R. Note: Only if AA is not modified.\n",
       "\n",
       "Args:\n",
       "    peptide (str): peptide.\n",
       "\n",
       "Returns:\n",
       "    str: peptide with swapped KRs."
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(swap_KR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L153){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_decoy_sequence\n",
       "\n",
       ">      get_decoy_sequence (peptide:str, pseudo_reverse:bool=False,\n",
       ">                          AL_swap:bool=False, KR_swap:bool=False)\n",
       "\n",
       "Reverses a sequence and adds the '_decoy' tag.\n",
       "Args:\n",
       "    peptide (str): modified peptide to be reversed.\n",
       "    pseudo_reverse (bool): If True, reverse the peptide bug keep the C-terminal amino acid; otherwise reverse the whole peptide. (Default: False)\n",
       "    AL_swap (bool): replace A with L, and vice versa. (Default: False)\n",
       "    KR_swap (bool): replace K with R at the C-terminal, and vice versa. (Default: False)\n",
       "Returns:\n",
       "    str: reversed peptide ending with the '_decoy' tag."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L153){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_decoy_sequence\n",
       "\n",
       ">      get_decoy_sequence (peptide:str, pseudo_reverse:bool=False,\n",
       ">                          AL_swap:bool=False, KR_swap:bool=False)\n",
       "\n",
       "Reverses a sequence and adds the '_decoy' tag.\n",
       "Args:\n",
       "    peptide (str): modified peptide to be reversed.\n",
       "    pseudo_reverse (bool): If True, reverse the peptide bug keep the C-terminal amino acid; otherwise reverse the whole peptide. (Default: False)\n",
       "    AL_swap (bool): replace A with L, and vice versa. (Default: False)\n",
       "    KR_swap (bool): replace K with R at the C-terminal, and vice versa. (Default: False)\n",
       "Returns:\n",
       "    str: reversed peptide ending with the '_decoy' tag."
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(get_decoy_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[K, K, K, L, A, K, K, K, ...]\n",
      "[A, A, A, K, R, A, A, A, ...]\n"
     ]
    }
   ],
   "source": [
    "print(swap_AL(parse('KKKALKKK')))\n",
    "print(swap_KR(parse('AAAKRAAA')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EDITPEP\n"
     ]
    }
   ],
   "source": [
    "print(get_decoy_sequence('PEPTIDE'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CBA', 'FED', 'IHG']\n"
     ]
    }
   ],
   "source": [
    "print(get_decoys(['ABC','DEF','GHI']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modifications\n",
    "\n",
    "To add modifications to the peptides, we distinguish fixed and variable modifications. Additionally, we make a distinction between whether the modification is only terminal or not. \n",
    "\n",
    "### Fixed Modifications\n",
    "Fixed modifications are implemented by passing a list with modified AAs that should be replaced. As a AA is only one letter, the remainder is the modification. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L247){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_fixed_mods\n",
       "\n",
       ">      add_fixed_mods (seqs:list, mods_fixed:list, **kwargs)\n",
       "\n",
       "Adds fixed modifications to sequences.\n",
       "Args:\n",
       "    seqs (list of str): sequences to add fixed modifications\n",
       "    mods_fixed (list of str): the string list of fixed modifications. Each modification string must be in lower case, except for that the last letter must be the modified amino acid (e.g. oxidation on M should be oxM).\n",
       "Returns:\n",
       "    list (of str): the list of the modified sequences. 'ABCDEF' with fixed mod 'cC' will be 'ABcCDEF'."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L247){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_fixed_mods\n",
       "\n",
       ">      add_fixed_mods (seqs:list, mods_fixed:list, **kwargs)\n",
       "\n",
       "Adds fixed modifications to sequences.\n",
       "Args:\n",
       "    seqs (list of str): sequences to add fixed modifications\n",
       "    mods_fixed (list of str): the string list of fixed modifications. Each modification string must be in lower case, except for that the last letter must be the modified amino acid (e.g. oxidation on M should be oxM).\n",
       "Returns:\n",
       "    list (of str): the list of the modified sequences. 'ABCDEF' with fixed mod 'cC' will be 'ABcCDEF'."
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(add_fixed_mods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AbBcCDEF']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mods_fixed = ['cC','bB']\n",
    "peptide_list = ['ABCDEF']\n",
    "\n",
    "add_fixed_mods(peptide_list, mods_fixed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['caAbBcCDEF']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mods_fixed = ['aA','cC','bB']\n",
    "peptide_list = ['cABCDEF']\n",
    "\n",
    "add_fixed_mods(peptide_list, mods_fixed)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Variable Modifications\n",
    "\n",
    "To employ variable modifications, we loop through each variable modification and each position of the peptide and add them to the peptide list. For each iteration in get_isoforms, one more variable modification will be added."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L285){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_isoforms\n",
       "\n",
       ">      get_isoforms (mods_variable_dict:dict, peptide:str, isoforms_max:int,\n",
       ">                    n_modifications_max:int=None)\n",
       "\n",
       "Function to generate modified forms (with variable modifications) for a given peptide - returns a list of modified forms.\n",
       "The original sequence is included in the list\n",
       "Args:\n",
       "    mods_variable_dict (dict): Dicitionary with modifications. The key is AA, and value is the modified form (e.g. oxM).\n",
       "    peptide (str): the peptide sequence to generate modified forms.\n",
       "    isoforms_max (int): max number of modified forms to generate per peptide.\n",
       "    n_modifications_max (int, optional): max number of variable modifications per peptide.\n",
       "Returns:\n",
       "    list (of str): the list of peptide forms for the given peptide"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L285){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_isoforms\n",
       "\n",
       ">      get_isoforms (mods_variable_dict:dict, peptide:str, isoforms_max:int,\n",
       ">                    n_modifications_max:int=None)\n",
       "\n",
       "Function to generate modified forms (with variable modifications) for a given peptide - returns a list of modified forms.\n",
       "The original sequence is included in the list\n",
       "Args:\n",
       "    mods_variable_dict (dict): Dicitionary with modifications. The key is AA, and value is the modified form (e.g. oxM).\n",
       "    peptide (str): the peptide sequence to generate modified forms.\n",
       "    isoforms_max (int): max number of modified forms to generate per peptide.\n",
       "    n_modifications_max (int, optional): max number of variable modifications per peptide.\n",
       "Returns:\n",
       "    list (of str): the list of peptide forms for the given peptide"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(get_isoforms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L264){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_variable_mod\n",
       "\n",
       ">      add_variable_mod (peps:list, mods_variable_dict:dict)\n",
       "\n",
       "Function to add variable modification to a list of peptides.\n",
       "Args:\n",
       "    peps (list): List of peptides.\n",
       "    mods_variable_dict (dict): Dicitionary with modifications. The key is AA, and value is the modified form (e.g. oxM). \n",
       "Returns:\n",
       "    list : the list of peptide forms for the given peptide."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L264){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_variable_mod\n",
       "\n",
       ">      add_variable_mod (peps:list, mods_variable_dict:dict)\n",
       "\n",
       "Function to add variable modification to a list of peptides.\n",
       "Args:\n",
       "    peps (list): List of peptides.\n",
       "    mods_variable_dict (dict): Dicitionary with modifications. The key is AA, and value is the modified form (e.g. oxM). \n",
       "Returns:\n",
       "    list : the list of peptide forms for the given peptide."
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(add_variable_mod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['PEPTIDE', 'pPEPTIDE', 'PEpPTIDE', 'pPEpPTIDE']\n",
      "['AMAMA', 'AoxMAMA', 'AMAoxMA', 'AoxMAoxMA']\n",
      "['AMAMA', 'AoxMAMA', 'AMAoxMA']\n"
     ]
    }
   ],
   "source": [
    "mods_variable_dict = {'S':'pS','P':'pP','M':'oxM'}\n",
    "isoforms_max = 1024\n",
    "\n",
    "print(get_isoforms(mods_variable_dict, 'PEPTIDE', isoforms_max))\n",
    "print(get_isoforms(mods_variable_dict, 'AMAMA', isoforms_max))\n",
    "print(get_isoforms(mods_variable_dict, 'AMAMA', isoforms_max, n_modifications_max=1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lastly, we define the wrapper [`add_variable_mods`](https://mannlabs.github.io/alphapept/fasta.html#add_variable_mods) so that the functions can be called for lists of peptides and a list of variable modifications."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L333){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_variable_mods\n",
       "\n",
       ">      add_variable_mods (peptide_list:list, mods_variable:list,\n",
       ">                         isoforms_max:int, n_modifications_max:int, **kwargs)\n",
       "\n",
       "Add variable modifications to the peptide list\n",
       "Args:\n",
       "    peptide_list (list of str): peptide list.\n",
       "    mods_variable (list of str): modification list.\n",
       "    isoforms_max (int): max number of modified forms per peptide sequence.\n",
       "    n_modifications_max (int): max number of variable modifications per peptide.\n",
       "Returns:\n",
       "    list (of str): list of modified sequences for the given peptide list."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L333){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_variable_mods\n",
       "\n",
       ">      add_variable_mods (peptide_list:list, mods_variable:list,\n",
       ">                         isoforms_max:int, n_modifications_max:int, **kwargs)\n",
       "\n",
       "Add variable modifications to the peptide list\n",
       "Args:\n",
       "    peptide_list (list of str): peptide list.\n",
       "    mods_variable (list of str): modification list.\n",
       "    isoforms_max (int): max number of modified forms per peptide sequence.\n",
       "    n_modifications_max (int): max number of variable modifications per peptide.\n",
       "Returns:\n",
       "    list (of str): list of modified sequences for the given peptide list."
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(add_variable_mods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AMA', 'AoxMA', 'AAC', 'AAamC']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peptide_list = ['AMA', 'AAC']\n",
    "mods_variable = ['oxM','amC']\n",
    "isoforms_max = 1024\n",
    "n_modifications_max = 10\n",
    "\n",
    "add_variable_mods(peptide_list, mods_variable, isoforms_max, n_modifications_max)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Terminal Modifications - Fixed\n",
    "\n",
    "To handle terminal modifications, we use the following convention:\n",
    "\n",
    "* `<` for the left side (N-terminal)\n",
    "* `>` for the right side (C-Terminal)\n",
    "\n",
    "Additionally, if we want to have a terminal modification on any AA we indicate this `^`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L388){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_fixed_mods_terminal\n",
       "\n",
       ">      add_fixed_mods_terminal (peptides:list, mods_fixed_terminal:list,\n",
       ">                               **kwargs)\n",
       "\n",
       "Wrapper to add fixed mods on sequences and lists of mods\n",
       "Args:\n",
       "    peptides (list of str): peptide list.\n",
       "    mods_fixed_terminal (list of str): list of fixed terminal mods.\n",
       "Raises:\n",
       "    \"Invalid fixed terminal modification {mod}\" exception for the given mod.\n",
       "Returns:\n",
       "    list (of str): list of peptides with modification added."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L388){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_fixed_mods_terminal\n",
       "\n",
       ">      add_fixed_mods_terminal (peptides:list, mods_fixed_terminal:list,\n",
       ">                               **kwargs)\n",
       "\n",
       "Wrapper to add fixed mods on sequences and lists of mods\n",
       "Args:\n",
       "    peptides (list of str): peptide list.\n",
       "    mods_fixed_terminal (list of str): list of fixed terminal mods.\n",
       "Raises:\n",
       "    \"Invalid fixed terminal modification {mod}\" exception for the given mod.\n",
       "Returns:\n",
       "    list (of str): list of peptides with modification added."
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(add_fixed_mods_terminal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L363){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_fixed_mod_terminal\n",
       "\n",
       ">      add_fixed_mod_terminal (peptides:list, mod:str)\n",
       "\n",
       "Adds fixed terminal modifications\n",
       "Args:\n",
       "    peptides (list of str): peptide list.\n",
       "    mod (str): n-term mod contains '<^' (e.g. a<^ for Acetyl@N-term); c-term mod contains '>^'.\n",
       "Raises:\n",
       "    \"Invalid fixed terminal modification 'mod name'\" for the given mod.\n",
       "Returns:\n",
       "    list (of str): list of peptides with modification added."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L363){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_fixed_mod_terminal\n",
       "\n",
       ">      add_fixed_mod_terminal (peptides:list, mod:str)\n",
       "\n",
       "Adds fixed terminal modifications\n",
       "Args:\n",
       "    peptides (list of str): peptide list.\n",
       "    mod (str): n-term mod contains '<^' (e.g. a<^ for Acetyl@N-term); c-term mod contains '>^'.\n",
       "Raises:\n",
       "    \"Invalid fixed terminal modification 'mod name'\" for the given mod.\n",
       "Returns:\n",
       "    list (of str): list of peptides with modification added."
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(add_fixed_mod_terminal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting with peptide ['AMAMA']\n",
      "Any n-term modified with x (x<^): ['xAMAMA']\n",
      "Any c-term modified with x (x>^): ['AMAMxA']\n",
      "Only A on n-term modified with x (x<A): ['xAMAMA']\n",
      "Only A on c-term modified with x (x<A): ['AMAMxA']\n"
     ]
    }
   ],
   "source": [
    "peptide = ['AMAMA']\n",
    "\n",
    "print(f'Starting with peptide {peptide}')\n",
    "print('Any n-term modified with x (x<^):', add_fixed_mods_terminal(peptide, ['x<^']))\n",
    "print('Any c-term modified with x (x>^):', add_fixed_mods_terminal(peptide, ['x>^']))\n",
    "print('Only A on n-term modified with x (x<A):', add_fixed_mods_terminal(peptide, ['x<A']))\n",
    "print('Only A on c-term modified with x (x<A):', add_fixed_mods_terminal(peptide, ['x>A']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Terminal Modifications - Variable\n",
    "\n",
    "Lastly, to handle terminal variable modifications, we use the function [`add_variable_mods_terminal`](https://mannlabs.github.io/alphapept/fasta.html#add_variable_mods_terminal). As the modification can only be at the terminal end, this function only adds a peptide where the terminal end is modified."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L440){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_unique_peptides\n",
       "\n",
       ">      get_unique_peptides (peptides:list)\n",
       "\n",
       "Function to return unique elements from list.\n",
       "Args:\n",
       "    peptides (list of str): peptide list.\n",
       "Returns:\n",
       "    list (of str): list of peptides (unique)."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L440){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_unique_peptides\n",
       "\n",
       ">      get_unique_peptides (peptides:list)\n",
       "\n",
       "Function to return unique elements from list.\n",
       "Args:\n",
       "    peptides (list of str): peptide list.\n",
       "Returns:\n",
       "    list (of str): list of peptides (unique)."
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(get_unique_peptides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L408){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_variable_mods_terminal\n",
       "\n",
       ">      add_variable_mods_terminal (peptides:list, mods_variable_terminal:list,\n",
       ">                                  **kwargs)\n",
       "\n",
       "Function to add variable terminal modifications.\n",
       "Args:\n",
       "    peptides (list of str): peptide list.\n",
       "    mods_variable_terminal (list of str): list of variable terminal mods.\n",
       "Returns:\n",
       "    list (of str): list of peptides with modification added."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L408){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_variable_mods_terminal\n",
       "\n",
       ">      add_variable_mods_terminal (peptides:list, mods_variable_terminal:list,\n",
       ">                                  **kwargs)\n",
       "\n",
       "Function to add variable terminal modifications.\n",
       "Args:\n",
       "    peptides (list of str): peptide list.\n",
       "    mods_variable_terminal (list of str): list of variable terminal mods.\n",
       "Returns:\n",
       "    list (of str): list of peptides with modification added."
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(add_variable_mods_terminal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['xAMAMA', 'AMAMA']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peptide_list = ['AMAMA']\n",
    "add_variable_mods_terminal(peptide_list, ['x<^'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generating Peptides\n",
    "\n",
    "Lastly, we put all the functions into a wrapper [`generate_peptides`](https://mannlabs.github.io/alphapept/fasta.html#generate_peptides). It will accept a peptide and a dictionary with settings so that we can get all modified peptides."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L506){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### check_peptide\n",
       "\n",
       ">      check_peptide (peptide:str, AAs:set)\n",
       "\n",
       "Check if the peptide contains non-AA letters.\n",
       "Args:\n",
       "    peptide (str): peptide sequence.\n",
       "    AAs (set): the set of legal amino acids. See alphapept.constants.AAs\n",
       "Returns:\n",
       "    bool: True if all letters in the peptide is the subset of AAs, otherwise False"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L506){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### check_peptide\n",
       "\n",
       ">      check_peptide (peptide:str, AAs:set)\n",
       "\n",
       "Check if the peptide contains non-AA letters.\n",
       "Args:\n",
       "    peptide (str): peptide sequence.\n",
       "    AAs (set): the set of legal amino acids. See alphapept.constants.AAs\n",
       "Returns:\n",
       "    bool: True if all letters in the peptide is the subset of AAs, otherwise False"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(check_peptide)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L451){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### generate_peptides\n",
       "\n",
       ">      generate_peptides (peptide:str, **kwargs)\n",
       "\n",
       "Wrapper to get modified peptides (fixed and variable mods) from a peptide.\n",
       "\n",
       "Args:\n",
       "    peptide (str): the given peptide sequence.\n",
       "Returns:\n",
       "    list (of str): all modified peptides.\n",
       "\n",
       "TODO:\n",
       "    There can be some edge-cases which are not defined yet.\n",
       "    Example:\n",
       "    Setting the same fixed modification - once for all peptides and once for only terminal for the protein.\n",
       "    The modification will then be applied twice."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L451){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### generate_peptides\n",
       "\n",
       ">      generate_peptides (peptide:str, **kwargs)\n",
       "\n",
       "Wrapper to get modified peptides (fixed and variable mods) from a peptide.\n",
       "\n",
       "Args:\n",
       "    peptide (str): the given peptide sequence.\n",
       "Returns:\n",
       "    list (of str): all modified peptides.\n",
       "\n",
       "TODO:\n",
       "    There can be some edge-cases which are not defined yet.\n",
       "    Example:\n",
       "    Setting the same fixed modification - once for all peptides and once for only terminal for the protein.\n",
       "    The modification will then be applied twice."
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(generate_peptides)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['aPEPTIDEM',\n",
       " 'aPEPTIDEoxM',\n",
       " 'MEDITPEaP_decoy',\n",
       " 'oxMEDITPEaP_decoy',\n",
       " 'PEPTIDEM',\n",
       " 'PEPTIDEoxM',\n",
       " 'MEDITPEP_decoy',\n",
       " 'oxMEDITPEP_decoy']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kwargs = {}\n",
    "\n",
    "kwargs[\"protease\"] = \"trypsin\"\n",
    "kwargs[\"n_missed_cleavages\"] = 2\n",
    "kwargs[\"pep_length_min\"] = 6\n",
    "kwargs[\"pep_length_max\"] = 27\n",
    "kwargs[\"mods_variable\"] = [\"oxM\"]\n",
    "kwargs[\"mods_variable_terminal\"] = []\n",
    "kwargs[\"mods_fixed\"] = [\"cC\"]\n",
    "kwargs[\"mods_fixed_terminal\"] = []\n",
    "kwargs[\"mods_fixed_terminal_prot\"] = []\n",
    "kwargs[\"mods_variable_terminal_prot\"]  = ['a<^']\n",
    "kwargs[\"isoforms_max\"] = 1024\n",
    "kwargs[\"n_modifications_max\"] = None\n",
    "\n",
    "generate_peptides('PEPTIDEM', **kwargs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mass Calculations\n",
    "\n",
    "Using the `mass_dict` from `constants` and being able to parse sequences with [`parse`](https://mannlabs.github.io/alphapept/fasta.html#parse), one can simply look up the masses for each modified or unmodified amino acid and add everything up.\n",
    "\n",
    "### Precursor\n",
    "\n",
    "To calculate the mass of the neutral precursor, we start with the mass of an $H_2O$ and add the masses of all amino acids of the sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L527){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_precmass\n",
       "\n",
       ">      get_precmass (parsed_pep:list, mass_dict:numba.typed.typeddict.Dict)\n",
       "\n",
       "Calculate the mass of the neutral precursor\n",
       "Args:\n",
       "    parsed_pep (list or numba.typed.List of str): the list of amino acids and modified amono acids.\n",
       "    mass_dict (numba.typed.Dict): key is the amino acid or the modified amino acid, and the value is the mass.\n",
       "Returns:\n",
       "    float: the peptide neutral mass."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L527){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_precmass\n",
       "\n",
       ">      get_precmass (parsed_pep:list, mass_dict:numba.typed.typeddict.Dict)\n",
       "\n",
       "Calculate the mass of the neutral precursor\n",
       "Args:\n",
       "    parsed_pep (list or numba.typed.List of str): the list of amino acids and modified amono acids.\n",
       "    mass_dict (numba.typed.Dict): key is the amino acid or the modified amino acid, and the value is the mass.\n",
       "Returns:\n",
       "    float: the peptide neutral mass."
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(get_precmass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "799.35996420346"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_precmass(parse('PEPTIDE'), constants.mass_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fragments\n",
    "\n",
    "Likewise, we can calculate the masses of the fragment ions. We employ two functions: [`get_fragmass`](https://mannlabs.github.io/alphapept/fasta.html#get_fragmass) and [`get_frag_dict`](https://mannlabs.github.io/alphapept/fasta.html#get_frag_dict). \n",
    "\n",
    "[`get_fragmass`](https://mannlabs.github.io/alphapept/fasta.html#get_fragmass) is a fast, `numba`-compatible function that calculates the fragment masses and returns an array indicating whether the ion-type was `b` or `y`. \n",
    "\n",
    "[`get_frag_dict`](https://mannlabs.github.io/alphapept/fasta.html#get_frag_dict) instead is not `numba`-compatible and hence a bit slower. It returns a dictionary with the respective ion and can be used for plotting theoretical spectra."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L546){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_fragmass\n",
       "\n",
       ">      get_fragmass (parsed_pep:list, mass_dict:numba.typed.typeddict.Dict)\n",
       "\n",
       "Calculate the masses of the fragment ions\n",
       "Args:\n",
       "    parsed_pep (numba.typed.List of str): the list of amino acids and modified amono acids.\n",
       "    mass_dict (numba.typed.Dict): key is the amino acid or the modified amino acid, and the value is the mass.\n",
       "Returns:\n",
       "    Tuple[np.ndarray(np.float64), np.ndarray(np.int8)]: the fragment masses and the fragment types (represented as np.int8). \n",
       "    For a fragment type, positive value means the b-ion, the value indicates the position (b1, b2, b3...); the negative value means\n",
       "    the y-ion, the absolute value indicates the position (y1, y2, ...)."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L546){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_fragmass\n",
       "\n",
       ">      get_fragmass (parsed_pep:list, mass_dict:numba.typed.typeddict.Dict)\n",
       "\n",
       "Calculate the masses of the fragment ions\n",
       "Args:\n",
       "    parsed_pep (numba.typed.List of str): the list of amino acids and modified amono acids.\n",
       "    mass_dict (numba.typed.Dict): key is the amino acid or the modified amino acid, and the value is the mass.\n",
       "Returns:\n",
       "    Tuple[np.ndarray(np.float64), np.ndarray(np.int8)]: the fragment masses and the fragment types (represented as np.int8). \n",
       "    For a fragment type, positive value means the b-ion, the value indicates the position (b1, b2, b3...); the negative value means\n",
       "    the y-ion, the absolute value indicates the position (y1, y2, ...)."
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(get_fragmass)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([ 98.06004033, 227.10263343, 324.15539729, 425.20307579,\n",
       "        538.28713979, 653.31408289, 148.06043425, 263.08737735,\n",
       "        376.17144135, 477.21911985, 574.27188371, 703.31447681]),\n",
       " array([ 1,  2,  3,  4,  5,  6, -1, -2, -3, -4, -5, -6], dtype=int8))"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_fragmass(parse('PEPTIDE'), constants.mass_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L583){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_frag_dict\n",
       "\n",
       ">      get_frag_dict (parsed_pep:list, mass_dict:dict)\n",
       "\n",
       "Calculate the masses of the fragment ions\n",
       "Args:\n",
       "    parsed_pep (list or numba.typed.List of str): the list of amino acids and modified amono acids.\n",
       "    mass_dict (numba.typed.Dict): key is the amino acid or the modified amino acid, and the value is the mass.\n",
       "Returns:\n",
       "    dict{str:float}: key is the fragment type (b1, b2, ..., y1, y2, ...), value is fragment mass."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L583){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_frag_dict\n",
       "\n",
       ">      get_frag_dict (parsed_pep:list, mass_dict:dict)\n",
       "\n",
       "Calculate the masses of the fragment ions\n",
       "Args:\n",
       "    parsed_pep (list or numba.typed.List of str): the list of amino acids and modified amono acids.\n",
       "    mass_dict (numba.typed.Dict): key is the amino acid or the modified amino acid, and the value is the mass.\n",
       "Returns:\n",
       "    dict{str:float}: key is the fragment type (b1, b2, ..., y1, y2, ...), value is fragment mass."
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(get_frag_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'b1': 98.06004032687,\n",
       " 'b2': 227.10263342687,\n",
       " 'b3': 324.15539728686997,\n",
       " 'b4': 425.20307578687,\n",
       " 'b5': 538.28713978687,\n",
       " 'b6': 653.31408288687,\n",
       " 'y1': 148.06043425033,\n",
       " 'y2': 263.08737735033,\n",
       " 'y3': 376.17144135033,\n",
       " 'y4': 477.21911985033,\n",
       " 'y5': 574.27188371033,\n",
       " 'y6': 703.31447681033}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_frag_dict(parse('PEPTIDE'), constants.mass_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This allows us also to generate the theoretical isotopes for a fragment:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA1IAAAHWCAYAAAB9mLjgAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/P9b71AAAACXBIWXMAAA9hAAAPYQGoP6dpAABNsElEQVR4nO3df3zN9f//8fuZsY3ZxuxnzOY3IULe8/vHWKISqbwlP4pokvSL96cUyq+8JSVS8qP8KETI75+pkJ9FIWl+ZDaKbYzN7Ly+f/juvB0b9ppt55y5XS+Xc7k4z9fznPN4nefzrHPv9Xo9j8UwDEMAAAAAgBxzc3QBAAAAAOBqCFIAAAAAYBJBCgAAAABMIkgBAAAAgEkEKQAAAAAwiSAFAAAAACYRpAAAAADAJIIUAAAAAJhEkAIAAAAAkwhSAPD/bdq0SRaLRQsXLnR0KXnOYrHorbfeypfnPnr0qCwWi2bOnJkvz49bW7VqlerUqSNPT09ZLBYlJiY6uiQAKPQIUgAKNYvFkqPbpk2bHF3qbVuxYkW+haW8cvToUfXq1UsVK1aUp6engoOD1axZM7355puOLk2SNGrUKC1ZssTRZZjyzz//6LHHHpOXl5cmT56szz//XCVKlMi315s5c6bdZ8fT01NVqlTRgAEDlJCQYOuX+T8mbnSbP3++rW94eLjdtsDAQDVt2lSLFy/O9jVvdAsPD5ckvfXWW7JYLPr7779tr9GzZ0+7vt7e3qpQoYIeffRRLVq0SFarNcu+tmjR4oavVa1atXx6hwG4CndHFwAA+enzzz+3uz979mytXbs2S3v16tV14MCBgiwtz61YsUKTJ0/ONkxdunRJ7u6O/ZP/xx9/qEGDBvLy8lLv3r0VHh6uU6dOaffu3Ro7dqyGDx/u0Pqkq0Hq0UcfVceOHR1dSo7t2LFD58+f18iRIxUVFVVgrztixAhFREQoNTVV33//vaZMmaIVK1Zo//79Kl68uK3fwIED1aBBgyyPj4yMtLtfp04dvfTSS5KkuLg4ffzxx+rUqZOmTJmitm3bZvnMPvPMM7rvvvvUt29fW5u3t/dNa/bw8NCnn34q6epn4tixY1q2bJkeffRRtWjRQt988418fHzsHlO2bFmNHj06y3P5+vre9LUAFH4EKQCF2pNPPml3f9u2bVq7dm2WdklOF6RSUlLy7MiCp6dnnjzP7Xjvvfd04cIF7d27V+XLl7fbdvr0aQdVlXt5OT63I/O98/Pzy7PnzMm+tWvXTvXr15d0NdT4+/trwoQJ+uabb9S1a1dbv6ZNm+rRRx+95Wvedddddp/Lp556SpUqVdJ7772nfv36qUKFCnb9M9uy+yzfiLu7e5b+b7/9tsaMGaOhQ4eqT58++vLLL+22+/r6mnoNAHcOTu0DgOtYrVa98847Klu2rDw9PdW6dWv98ccfWfpt375d999/v3x9fVW8eHE1b95cP/zwQ5Z+e/bsUbt27eTj4yNvb2+1bt1a27Zts+uTeerS5s2b9dxzzykwMFBly5a1bV+5cqWaNm2qEiVKqGTJkmrfvr1+/fVX2/aePXtq8uTJkuxPZ8yU3TVSJ0+e1NNPP63Q0FB5eHgoIiJC/fv31+XLlyVJZ8+e1csvv6xatWrJ29tbPj4+ateunX7++Wfzb6qkI0eOqGzZsllClCQFBgba3Q8PD1eHDh20Zs0a27U/NWrU0Ndff53lsYmJiRo0aJDKlSsnDw8PVapUSWPHjs1yqpbVatX777+vWrVqydPTUwEBAbr//vu1c+dO23uUkpKiWbNm2d6/nj17SvrfqWK//fab/v3vf6tUqVJq0qSJpKunf7Vo0SJLXT179rSdaib971qy8ePHa/LkyapQoYKKFy+utm3b6sSJEzIMQyNHjlTZsmXl5eWlhx9+WGfPnr3pe9qiRQv16NFDktSgQQO7miVpwYIFqlevnry8vFSmTBk9+eSTOnnyZJY6vb29deTIET3wwAMqWbKkunXrdtPXzU6rVq0kSbGxsaYfm53g4GBVr149z57vZoYMGaK2bdtqwYIF+v333/P99QAUDhyRAoDrjBkzRm5ubnr55ZeVlJSkcePGqVu3btq+fbutz4YNG9SuXTvVq1dPb775ptzc3DRjxgy1atVKW7Zs0X333SdJ+vXXX9W0aVP5+Pjo1VdfVdGiRfXxxx+rRYsW2rx5sxo2bGj32s8995wCAgI0bNgwpaSkSLp6emKPHj0UHR2tsWPH6uLFi5oyZYqaNGmiPXv2KDw8XM8++6zi4uKyPW0xO3FxcbrvvvuUmJiovn37qlq1ajp58qQWLlyoixcvqlixYvrzzz+1ZMkSdenSRREREUpISNDHH3+s5s2b67ffflNoaKip97V8+fJat26dNmzYYPvSfTOHDx/W448/rn79+qlHjx6aMWOGunTpolWrVqlNmzaSpIsXL6p58+Y6efKknn32WYWFhenHH3/U0KFDderUKU2cONH2fE8//bRmzpypdu3a6ZlnntGVK1e0ZcsWbdu2TfXr19fnn3+e5XSxihUr2tXUpUsXVa5cWaNGjZJhGKb2P9OcOXN0+fJlPf/88zp79qzGjRunxx57TK1atdKmTZv02muv6Y8//tAHH3ygl19+WZ999tkNn+v//u//VLVqVU2bNs12ql1mzTNnzlSvXr3UoEEDjR49WgkJCXr//ff1ww8/aM+ePXZHsK5cuaLo6Gg1adJE48ePtzs1L6eOHDkiSfL397drP3/+vN21Spn8/f3twv710tPTdeLEiSzPl1+6d++uNWvWaO3atapSpYqtPSMjI9v6vby8nOKIJAAHMgDgDhITE2Pc6E/fxo0bDUlG9erVjbS0NFv7+++/b0gy9u3bZxiGYVitVqNy5cpGdHS0YbVabf0uXrxoREREGG3atLG1dezY0ShWrJhx5MgRW1tcXJxRsmRJo1mzZra2GTNmGJKMJk2aGFeuXLG1nz9/3vDz8zP69OljV2t8fLzh6+tr136zfZNkvPnmm7b7Tz31lOHm5mbs2LEjS9/MfUpNTTUyMjLstsXGxhoeHh7GiBEj7NokGTNmzMj2tTPt37/f8PLyMiQZderUMV544QVjyZIlRkpKSpa+5cuXNyQZixYtsrUlJSUZISEhRt26dW1tI0eONEqUKGH8/vvvdo8fMmSIUaRIEeP48eOGYRjGhg0bDEnGwIEDb7i/hmEYJUqUMHr06JGlz5tvvmlIMrp27ZplW/PmzY3mzZtnae/Ro4dRvnx52/3M9ykgIMBITEy0tQ8dOtSQZNxzzz1Genq6rb1r165GsWLFjNTU1CzPfa3MuXPtWF6+fNkIDAw0atasaVy6dMnWvnz5ckOSMWzYMLs6JRlDhgy56etc/3rr1q0zzpw5Y5w4ccKYP3++4e/vb3h5eRl//fWXYRj/+zzd6Hbq1Cnbc5YvX95o27atcebMGePMmTPGzz//bDzxxBOGJOP555/Pto4bjZVh/G+8zpw5Y7efJUqUuOF+7dmzx5BkvPjii7a25s2b37D+Z599NkfvF4DCiyNSAHCdXr16qVixYrb7TZs2lST9+eefqlmzpvbu3avDhw/r9ddf1z///GP32NatW+vzzz+X1WqVYRhas2aNOnbsaHd9R0hIiP7973/rk08+UXJyst3F7X369FGRIkVs99euXavExER17drV7v+KFylSRA0bNtTGjRtN75/VatWSJUv04IMP2q5xuVbmUQIPDw9bW0ZGhhITE+Xt7a2qVatq9+7dpl/37rvv1t69ezVy5EgtX75ce/fu1fvvvy9vb29NmDBBffr0sesfGhqqRx55xHbfx8dHTz31lMaOHav4+HgFBwdrwYIFatq0qUqVKmX3/kRFRWnMmDH67rvv1K1bNy1atEgWiyXb1QFvdlTkev369TO939fr0qWL3UIFmUcln3zySbsFQRo2bKh58+bp5MmTWa4PupWdO3fq9OnTeuutt+yuj2vfvr2qVaumb7/9NsviHv379zf1GtcvbFG+fHnNmTNHd911l137sGHDbJ+ha5UuXdru/po1axQQEGC7X6RIEXXv3l1jx441VVduZS5Ucf78ebv28PBwffLJJ1n6X3vqLYA7E0EKAK4TFhZmd79UqVKSpHPnzkm6esqZJNu1KdlJSkpSWlqaLl68qKpVq2bZXr16dVmtVp04cUJ33323rT0iIsKuX+Zr3ehUuOtXGMuJM2fOKDk5WTVr1rxpv8xrij766CPFxsYqIyPDti23p1tVqVJFn3/+uTIyMvTbb79p+fLlGjdunPr27auIiAi7L+eVKlXKEnIyT7k6evSogoODdfjwYf3yyy92X8CvlbkQw5EjRxQaGprly7tZ149Pblw/vzJDVbly5bJtz5x3Zhw7dkySsp171apV0/fff2/X5u7ubjoYTJ48WVWqVJG7u7uCgoJUtWpVubllvfS6Vq1aOVpNsGHDhnr77bdlsVhUvHhxVa9ePU8X0LiVCxcuSJJKlixp116iRIkCXQ0RgOsgSAHAda49InQt4/9fE5O5iMG7776rOnXqZNvX29tbaWlppl/by8vL7n7ma33++ecKDg7O0j8/lzQfNWqU3njjDfXu3VsjR45U6dKl5ebmpkGDBmX7mztmFClSRLVq1VKtWrUUGRmpli1bas6cOaa/sFqtVrVp00avvvpqttuvvdYlL1w/PtLVI1pGNtdLXRs8r3Wj+XWreZefPDw8sg1BN3Pfffdle0Qzt8qUKePQwLJ//35JVwM8AOQEQQoATMq8mN/Hx+emX/wCAgJUvHhxHTp0KMu2gwcPys3NLctRiBu9VmBg4C2/ZOb0FLWAgAD5+PjYvjjeyMKFC9WyZUtNnz7drj0xMVFlypTJ0WvlROaX8VOnTtm1//HHHzIMw26/MldUy1wNr2LFirpw4cIt35uKFStq9erVOnv27E2PSpk5zS9TqVKl9Oeff2Zpzzwq5AiZKyMeOnQoy9HMQ4cOZbty4p3u888/l8VisS1kAgC3wvLnAGBSvXr1VLFiRY0fP952OtC1zpw5I+nqEYa2bdvqm2++0dGjR23bExISNHfuXDVp0uSWp+ZFR0fLx8dHo0aNUnp6+g1fS5JtBbHExMSbPqebm5s6duyoZcuW2Zb+vlbmEZAiRYpkORqyYMGCLMtn59SWLVuy3YcVK1ZIynoaWlxcnBYvXmy7n5ycrNmzZ6tOnTq2o3OPPfaYtm7dqtWrV2d53sTERF25ckWS1LlzZxmGke2P/l67jyVKlLjl+3e9ihUr6uDBg3Zj8fPPP2e7FH5BqV+/vgIDAzV16lS7I6MrV67UgQMH1L59e4fV5ozGjBmjNWvW6PHHH1flypUdXQ4AF8ERKQAwyc3NTZ9++qnatWunu+++W7169dJdd92lkydPauPGjfLx8dGyZcskXf2xz7Vr16pJkyZ67rnn5O7uro8//lhpaWkaN27cLV/Lx8dHU6ZMUffu3XXvvffqiSeeUEBAgI4fP65vv/1WjRs31ocffijpasCTpIEDByo6OlpFihTRE088ke3zjho1SmvWrFHz5s3Vt29fVa9eXadOndKCBQv0/fffy8/PTx06dNCIESPUq1cvNWrUSPv27dOcOXNML3yQaezYsdq1a5c6deqk2rVrS5J2796t2bNnq3Tp0ho0aJBd/ypVqujpp5/Wjh07FBQUpM8++0wJCQmaMWOGrc8rr7yipUuXqkOHDurZs6fq1aunlJQU7du3TwsXLtTRo0dVpkwZtWzZUt27d9ekSZN0+PBh3X///bJardqyZYtatmypAQMG2N7DdevWacKECQoNDVVERESWJeqv17t3b02YMEHR0dF6+umndfr0aU2dOlV33323kpOTc/Ve3a6iRYtq7Nix6tWrl5o3b66uXbvalj8PDw/Xiy++WGC1bNmyRampqVnaa9eubZsHBeXKlSv64osvJEmpqak6duyYli5dql9++UUtW7bUtGnTsjwmKSnJ9pjr8UO9wB3OYesFAoAD5GT58wULFti132h57z179hidOnUy/P39DQ8PD6N8+fLGY489Zqxfv96u3+7du43o6GjD29vbKF68uNGyZUvjxx9/tOuT3RLW19cWHR1t+Pr6Gp6enkbFihWNnj17Gjt37rT1uXLlivH8888bAQEBhsVisdtPXbf8uWEYxrFjx4ynnnrKCAgIMDw8PIwKFSoYMTExtqXfU1NTjZdeeskICQkxvLy8jMaNGxtbt27Nstx3Tpc//+GHH4yYmBijZs2ahq+vr1G0aFEjLCzM6Nmzp93y8IZxdTns9u3bG6tXrzZq165teHh4GNWqVcsyNoZxdYn4oUOHGpUqVTKKFStmlClTxmjUqJExfvx44/Lly3bvz7vvvmtUq1bNKFasmBEQEGC0a9fO2LVrl63PwYMHjWbNmtmWac9cXju75bSv9cUXXxgVKlQwihUrZtSpU8dYvXr1DZc/f/fdd+0ee6N5d6s5kZN+X375pVG3bl3Dw8PDKF26tNGtWzfb8uSZbrUsuJnXu9atlj+/dj5mjrcZuVn+/NrXL168uBEeHm507tzZWLhwYZal/g3j5suf8xUKgMUwCuAqVgAATAgPD1fNmjW1fPlyR5cCAEC2uEYKAAAAAEwiSAEAAACASQQpAAAAADCJa6QAAAAAwCSOSAEAAACASQQpAAAAADCJH+SVZLVaFRcXp5IlS8pisTi6HAAAAAAOYhiGzp8/r9DQULm53fi4E0FKUlxcnMqVK+foMgAAAAA4iRMnTqhs2bI33E6QklSyZElJV98sHx8fB1cDAAAAwFGSk5NVrlw5W0a4EYKUZDudz8fHhyAFAAAA4JaX/LDYBAAAAACYRJACAAAAAJMIUgAAAABgEkEKAAAAAEwiSDmpadOmqX79+kpMTHR0KShkmFuujzEsfBhTXIv5gOwwL5wPq/a5uG3btmnNmjXav3+/jh49qqCgIC1btszRZeW5v//+W/PmzdP+/ft14MABXbx4UR9//LHq1avn6NIKpdTUVC1dulSbN2/WH3/8oUuXLqls2bLq1KmTOnXqdNMfpytIP/30k1atWqW9e/cqISFB/v7+atCggfr3768yZco4ujyH++yzz/Tdd9/pr7/+0sWLFxUUFKQmTZqod+/eKlWqlKPLy1O7d+/WF198oYMHDyoxMVHe3t6qWrWqnnnmGd1zzz2OLi9fnD9/Xp06ddK5c+c0duxYtW7d2tEl5bu3335bS5YsUZMmTTRx4kRHl+Nwffv21e7du7O0R0ZG6oMPPnBARVktW7ZMw4cPz3bb6tWr5e/vX8AV3RnS09P1+eef69tvv9WpU6fk7e2t6tWr6//+7/8UGBjo6PLyzE8//aTPPvtMBw4ckGEYCgsL01NPPaW2bdsWyOsTpFzcqlWrtGbNGlWrVq1Qf3E8duyYZs2apbCwMFWqVEm//PKLo0sq1E6ePKl3331XDRo0ULdu3eTt7a2tW7dqzJgx2rdv3w3/o1jQPvjgAyUlJSkqKkphYWE6efKkvvzyS23ZskXz5s274/8DffDgQVWpUkVt27ZViRIlFBsbq8WLF+v777/X3Llz5eXl5egS88zx48dlsVj06KOPyt/fX8nJyVq5cqX69OmjiRMnqlGjRo4uMc9NnTpVqampji6jwPz2229atmyZihUr5uhSnEpgYKAGDBhg1xYQEOCgam6sX79+Cg0NtWu71W/0IHeuXLmiF154Qb/88oseeeQRVa5cWcnJydq/f78uXLhQaILU0qVLNXLkSDVs2FAxMTEqUqSIjh07poSEhAKrgSDl4mJiYvT666/L3d1dgwYN0pEjRxxdUr6oXr26NmzYIB8fH61fv54glc/8/f315ZdfqkKFCra2Tp06acSIEVq6dKmeeeYZlStXzoEVXvXiiy+qTp06dkfIIiMj1bdvX3355Zd67rnnHFid440bNy5LW+3atfXqq69qy5YtBfZ/7ApCx44d1bFjR7u2Ll266KGHHtK8efMKXZA6cuSIFi5cqD59+mjq1KmOLiffGYah8ePHq3379tqxY4ejy3Eq3t7eeuCBBxxdxi01atRINWrUcHQZd4S5c+dq9+7dmj59uu6++25Hl5Mv4uLiNHbsWD3++ON6+eWXHVaHc5yfgxtKTEzUkCFD1KxZM7Vq1Urjx4/X5cuXbdsDAgLk7u66eXjnzp2qX7++Nm7cmGXbqlWrVL9+ff3yyy8qXrw4P5acx242t/z8/OxCVKYWLVpIkmJjY/O1tpzOi3vvvTfLaYb33nuvfHx8dPTo0Xyt0Rnc6u9DdkJCQiRdPS3MFeR0LmTH09NTpUqVcpl9lXI+puPHj1fLli1Vt25dB1R5+8yO64oVK/THH38oJiamIMt0uJzOh4yMDF28eLFAa8vNZ/PixYuyWq0FVWKhdbN5YbVaNW/ePLVs2VJ33323MjIyXOrIdU7n1aJFi2S1WtWvXz9JV+eWYRgFXS5BytkNGTJEly9f1oABA9SkSRPNnz9fb7/9tqPLyjP16tVTUFCQVq5cmWXbypUrVbZsWdWuXdsBlRV+uZlb//zzj6SrQSs/3c68uHjxoi5dupTvNTqDnIyhYRhKTEzUP//8oz179ujdd9+Vm5uby1xfaHYupKSkKDExUUePHtXkyZN15MgR3XfffQVZ8m3JyZiuW7dOP//8swYOHOigKm+fmXG9ePGiJk2apN69e99xp+vmZD4cP35cTZo0UbNmzdS2bVtNmTJFV65cyffazH42+/Xrp2bNmqlx48YaPHiwjh8/nu81FlY3mxexsbE6c+aMKleurHfeeUdNmjRRkyZN9MQTT2jnzp0OrvzWcjqvfvrpJ4WHh+uHH37QAw88oGbNmql169aaMmVKgYZ11z2UcYcIDQ3VhAkTJEmPPfaYSpQooQULFqh79+6qXLmyg6u7fRaLRQ888IDmzJmjCxcuyNvbW5J07tw5bdu2TU8//bSDKyy8zM6t9PR0zZs3T6Ghofl+qsDtzIt58+YpPT1dbdq0ydcanUFOxvDs2bOKjo62PSYwMFDvvPOOwsPDHVGyaWbnwpAhQ7R161ZJUtGiRdWpUyc988wzBV53bt1qTNPS0jRx4kR169ZNoaGhOnXqlIMrzh0z4/rJJ5/Iw8ND//73vx1VrsPcaj6ULVtW9evXV6VKlZSamqp169Zp+vTpOn78uEaPHp2vteV0DD09PfXggw+qfv36KlGihA4cOKA5c+aod+/emjNnjoKCgvK1zsLoZvPir7/+kiTNmTNHvr6++s9//iNJmjFjhp5//nnNnj3bqb8/5nReHT9+XEWKFNHw4cP11FNPqUqVKtqwYYOmT5+ujIyMLNcN5heOSDm5xx57zO7+448/Lkn64YcfHFFOvmjfvr0uX76s9evX29rWrl2rjIwMtWvXzoGVFW5m59a4ceP0559/6rXXXlORIkXyvb7czIvdu3dr2rRpatOmjRo0aJDvNTpaTsbQx8dHH330kd577z3169dPfn5+unTpUoHWebvMzIXnn39ekydP1rBhw1SrVi1duXKlQP7vfF651ZjOnDlTV65cUa9evQq8tryWk3E9fvy45s2bp0GDBt2Ri0zcaj4MGzZMffv2VatWrfTAAw9owoQJeuSRR7R27Vrt27cv3+vLyRi2adNGb775ptq3b68WLVqof//++vDDD5WUlKTp06fne42F0c3mRebf94sXL2rKlCl68MEH9eCDD+qjjz6SYRiaPXt2gddrVk7m1aVLl5ScnKxnn31W/fr1U6tWrfT222+rUaNGmjdvXoGd6kqQcnJhYWF298uWLSs3NzfFxcU5qKK8Fx4erho1atgdxl25cqVq1arlFAsaFFZm5tbs2bO1ePFi9e/fX40bNy6Q+szOi6NHj+rll19WxYoV9cYbbxRIjY6WkzEsWrSo7rvvPjVt2lTPPPOMhgwZohEjRmjLli0FXW6umZkLVapUUcOGDfXQQw9p8uTJ2r9/v956660Crjj3bjamcXFxmj17tp577jkVL17cQRXmnZyM6/jx41W7dm21atXKUWU6VG6+Azz55JOSri4Lnd9y+9/vOnXqqGbNmgVSY2F0s3nh4eEhSbrnnnvsjvYFBwerTp06+vnnnwu01tzIybzK3M9rz7jIvJ+WlqaDBw8WSK0EKRdjsVgcXUK+6NChg3bv3q3Tp0/rr7/+0r59+zgaVcBuNLeWLVumDz74QJ07dy7wUy1zOi8SEhIUExMjb29vTZo0qVB8ycyNnPx9qF27tsqUKZPt+efOLDd/I4oWLarmzZtr48aNSktLK6BK89a1Yzp16lQFBgaqfv36tmCVed3iuXPnFBcX53IX8t9sXHfs2KEff/xRXbt2te1vXFycMjIylJaWpri4OKWkpDh4DwpWTj7jmV+ek5KS8rscSbn/73dQUJCSk5MLoMLC79p5kbn0fenSpbP0K126tMssvnOreXWj/cz8jcSC2k+ClJO7/mLMEydOyGq1ZvktBlfXtm1bubm5adWqVVq5cqXc3d0L1dLMzignc2vz5s0aOXKkWrZsqddee62gS8zRvEhKSlJMTIzS09P14YcfFurfU7tebv8+XL58WRcuXMjP0vJcbv9GpKWlyTAMl/nCfbMxjY+P14kTJ/TQQw/ZbpnXP4wZM0YPPfSQy+xnppuNa3x8vCTplVdesdvn06dPa8eOHXrooYf0zTffOLL8fJebz/jJkyclqcB+dDu3n82TJ0/eEYsC5YebzYtKlSrJ3d1dZ86cyfK4M2fOuMyPsd9qXlWrVk2Ssuzn33//Lang5j+LTTi5r776Sv/6179s97/88ktJKnS/ieLn56dGjRpp5cqVSktLU6NGjfgDm89uNbd2796toUOH6t5779Xbb7+dZZnxgnCreXHp0iUNHDhQp0+f1scff5zldIfC7mZjeOnSJVksFnl6eto9ZsOGDUpOTna533O51Vw4e/Zslv8zef78ea1fv15BQUHZ/t9ZZ3SzMa1Tp44SExPt+h85ckRTpkzRU089pdq1a7vcjyzfbFwbNGig8ePHZ3nMO++8o5CQEPXu3VuVKlUq4IoL1s3mQ0pKiooWLWp37ZhhGLbrjiIjIwukxlt9Ns+dO5flS+0PP/ygAwcO6IknniiQGgubm82L4sWLq3HjxtqyZYuOHj1qW1goNjZWP//8szp37uyIkk271bxq27at1qxZo2+++cb2m5FWq1VLly6Vj4+PLWjlN4KUk4uLi9PgwYMVGRmpffv2acWKFbr//vtVpUoVSdLhw4f13XffSbr6fyQuXLhg+yNauXJlNWvWzGG1m9WhQwe9+uqrkqT+/ftn2Z65X5k/Ovztt99q7969ksTqfrlws7l16tQpDR48WBaLRa1bt9a6devsHlu5cuUCW/XnZvPi9ddf16+//qqHHnpIsbGxdr9vVbx4cdvvXhVWNxvD33//Xf3791fbtm0VHh4ui8WiAwcOaMWKFQoNDVXXrl0dXb5pN5sLAwcOVGBgoGrWrKnSpUsrPj5eS5cu1d9//53vq5flpVv9zb9eyZIlJUl33323y873G41rcHCwgoODs/T/73//q9KlS7vs/ppxs/mwa9cu/ec//1F0dLTKlSuntLQ0bdy4UT///LM6depUYF8kpZt/Nnv37q2qVauqRo0a8vb21sGDB/XNN98oKChIvXv3LrAaC5Nb/Z2IiYnRjh071K9fP1tYnT9/vnx9fV1qoZqbzavmzZvrvvvu04wZM5SYmKjKlStr06ZN2rt3r/7zn/8U2OI0BCknN3r0aE2dOlUffPCB3N3d9dhjj2nQoEG27QcPHtSUKVPsHpN5v0OHDi4VpJo2bSofHx9ZrVY1b948y/br93Pp0qW2fxOkzLvZ3IqLi7Od+jV27Ngsj+3bt2+BBambzYvff/9d0tW5cO18kK7+8Gxh/6J1szEMDAxU69attWPHDi1fvlxXrlxRSEiIHn/8cfXu3Vu+vr6OLT4XbjYXHnroIa1Zs0Zz587V+fPn5ePjo1q1aql79+4u9aO1t/qbXxjd6m//nexm8yEkJER169bVpk2b9Pfff8vNzU0RERH6z3/+o0ceeaRA67zZGLZp00bff/+9tm3bptTUVJUpU0aPPPKI+vbt6zJHip3Nrf5OVKhQQdOmTdOkSZM0ffp0WSwWNWjQQC+88IICAwMdV7hJN5tXFotF48eP15QpU7RmzRotW7ZM5cuX18iRIwv0GnuL4YifAXYyycnJ8vX1VVJSknx8fBxdzh0rIyND999/v5o2baphw4Y5uhw4CeYFMjEXCifG1fUxhsgPjpxXOc0GLDYBp7Fp0yadO3dOHTp0cHQpcCLMC2RiLhROjKvrYwyRH1xhXnFqHxxu//79Onz4sD799FNVrVpV9957r6NLghNgXiATc6FwYlxdH2OI/OBK84ogBYdbuHChVqxYoapVq7rUD2cifzEvkIm5UDgxrq6PMUR+cKV5xTVS4hopAAAAAFdxjRQAAAAA5BOCFAAAAACYRJACAAAAAJMIUgAAAABgEkEKAAAAAExyaJD67rvv9OCDDyo0NFQWi0VLliyx224YhoYNG6aQkBB5eXkpKipKhw8ftutz9uxZdevWTT4+PvLz89PTTz+tCxcuFOBeAAAAALjTODRIpaSk6J577tHkyZOz3T5u3DhNmjRJU6dO1fbt21WiRAlFR0crNTXV1qdbt2769ddftXbtWi1fvlzfffed+vbtW1C7AAAAAOAO5DS/I2WxWLR48WJ17NhR0tWjUaGhoXrppZf08ssvS5KSkpIUFBSkmTNn6oknntCBAwdUo0YN7dixQ/Xr15ckrVq1Sg888ID++usvhYaG5ui1+R0pAAAAAFIh+B2p2NhYxcfHKyoqytbm6+urhg0bauvWrZKkrVu3ys/PzxaiJCkqKkpubm7avn37DZ87LS1NycnJdjcAAAAAyCmnDVLx8fGSpKCgILv2oKAg27b4+HgFBgbabXd3d1fp0qVtfbIzevRo+fr62m7lypXL4+pzx2q1ateuXdq1a5esVqujy8k3d8p+OgtXeb9dpU5HuBPfmzthn++EfbzWnba/ZrjKe+MqdRYWd8r77cr76bRBKj8NHTpUSUlJttuJEyccXRIAAAAAF+K0QSo4OFiSlJCQYNeekJBg2xYcHKzTp0/bbb9y5YrOnj1r65MdDw8P+fj42N0AAAAAIKecNkhFREQoODhY69evt7UlJydr+/btioyMlCRFRkYqMTFRu3btsvXZsGGDrFarGjZsWOA1AwAAALgzuDvyxS9cuKA//vjDdj82NlZ79+5V6dKlFRYWpkGDBuntt99W5cqVFRERoTfeeEOhoaG2lf2qV6+u+++/X3369NHUqVOVnp6uAQMG6Iknnsjxin0AAAAAYJZDg9TOnTvVsmVL2/3BgwdLknr06KGZM2fq1VdfVUpKivr27avExEQ1adJEq1atkqenp+0xc+bM0YABA9S6dWu5ubmpc+fOmjRpUoHvCwAAAIA7h0ODVIsWLXSzn7GyWCwaMWKERowYccM+pUuX1ty5c/OjPAAAAADIltNeIwUAAAAAzoogBQAAAAAmEaQAAAAAwCSCFAAAAACYRJACAAAAAJMIUgAAAABgEkEKAAAAAEwiSAEAAACASQQpAAAAADCJIAUAAAAAJhGkAAAAAMAkghQAAAAAmESQAgAAAACTCFIAAAAAYBJBCgAAAABMIkgBAAAAgEkEKQAAAAAwiSAFAAAAACYRpAAAAADAJIIUAAAAAJhEkAIAAAAAkwhSAAAAAGASQQoAAAAATCJIAQAAAIBJBCkAAAAAMIkgBQAAAAAmEaQAAAAAwCSCFAAAAACYRJACAAAAAJMIUgAAAABgEkEKAAAAAEwiSAEAAACASQQpAAAAADCJIAUAAAAAJhGkAAAAAMAkghQAAAAAmESQAgAAAACTCFIAAAAAYBJBCgAAAABMIkgBAAAAgEkEKQAAAAAwiSAFAAAAACYRpAAAAADAJIIUAAAAAJhEkAIAAAAAkwhSAAAAAGASQQoAAAAATCJIAQAAAIBJBCkAAAAAMIkgBQAAAAAmEaQAAAAAwCSCFAAAAACYRJACAAAAAJMIUgAAAABgEkEKAAAAAEwiSAEAAACASQQpAAAAADCJIAUAAAAAJhGkAAAAAMAkpw5SGRkZeuONNxQRESEvLy9VrFhRI0eOlGEYtj6GYWjYsGEKCQmRl5eXoqKidPjwYQdWDQAAAKCwc+ogNXbsWE2ZMkUffvihDhw4oLFjx2rcuHH64IMPbH3GjRunSZMmaerUqdq+fbtKlCih6OhopaamOrByAAAAAIWZu6MLuJkff/xRDz/8sNq3by9JCg8P17x58/TTTz9Juno0auLEiXr99df18MMPS5Jmz56toKAgLVmyRE888YTDagcAAABQeDn1EalGjRpp/fr1+v333yVJP//8s77//nu1a9dOkhQbG6v4+HhFRUXZHuPr66uGDRtq69atN3zetLQ0JScn290AAAAAIKec+ojUkCFDlJycrGrVqqlIkSLKyMjQO++8o27dukmS4uPjJUlBQUF2jwsKCrJty87o0aM1fPjw/CscAAAAQKHm1EekvvrqK82ZM0dz587V7t27NWvWLI0fP16zZs26recdOnSokpKSbLcTJ07kUcUAAAAA7gROfUTqlVde0ZAhQ2zXOtWqVUvHjh3T6NGj1aNHDwUHB0uSEhISFBISYntcQkKC6tSpc8Pn9fDwkIeHR77WDgAAAKDwcuojUhcvXpSbm32JRYoUkdVqlSRFREQoODhY69evt21PTk7W9u3bFRkZWaC1AgAAALhzOPURqQcffFDvvPOOwsLCdPfdd2vPnj2aMGGCevfuLUmyWCwaNGiQ3n77bVWuXFkRERF64403FBoaqo4dOzq2eAAAAACFllMHqQ8++EBvvPGGnnvuOZ0+fVqhoaF69tlnNWzYMFufV199VSkpKerbt68SExPVpEkTrVq1Sp6eng6sHAAAAEBh5tRBqmTJkpo4caImTpx4wz4Wi0UjRozQiBEjCq4wAAAAAHc0p75GCgAAAACcEUEKAAAAAEwiSAEAAACASQQpAAAAADCJIAUAAAAAJhGkAAAAAMAkghQAAAAAmESQAgAAAACTCFIAAAAAYBJBCgAAAABMIkgBAAAAgEkEKQAAAAAwiSAFAAAAACYRpAAAAADAJIIUAAAAAJhEkAIAAAAAkwhSAAAAAGASQQoAAAAATCJIAQAAAIBJBCkAAAAAMIkgBQAAAAAmEaQAAAAAwCSCFAAAAACYRJACAAAAAJMIUgAAAABgEkEKAAAAAEwiSAEAAACASQQpAAAAADCJIAUAAAAAJhGkAAAAAMAkghQAAAAAmESQAgAAAACTCFIAAAAAYBJBCgAAAABMIkgBAAAAgEkEKQAAAAAwiSAFAAAAACYRpAAAAADAJIIUAAAAAJhEkAIAAAAAkwhSAAAAAGASQQoAAAAATCJIAQAAAIBJBCkAAAAAMIkgBQAAAAAmEaQAAAAAwCSCFAAAAACYRJACAAAAAJMIUgAAAABgEkEKAAAAAEwiSAEAAACASQQpAAAAADCJIAUAAAAAJhGkAAAAAMCkXAWpHj166LvvvsvrWgAAAADAJeQqSCUlJSkqKkqVK1fWqFGjdPLkybyuCwAAAACcVq6C1JIlS3Ty5En1799fX375pcLDw9WuXTstXLhQ6enpeV0jAAAAADiVXF8jFRAQoMGDB+vnn3/W9u3bValSJXXv3l2hoaF68cUXdfjw4bysEwAAAACcxm0vNnHq1CmtXbtWa9euVZEiRfTAAw9o3759qlGjht577728qBEAAAAAnEquglR6eroWLVqkDh06qHz58lqwYIEGDRqkuLg4zZo1S+vWrdNXX32lESNG3HaBJ0+e1JNPPil/f395eXmpVq1a2rlzp227YRgaNmyYQkJC5OXlpaioKI6GAQAAAMhX7rl5UEhIiKxWq7p27aqffvpJderUydKnZcuW8vPzu63izp07p8aNG6tly5ZauXKlAgICdPjwYZUqVcrWZ9y4cZo0aZJmzZqliIgIvfHGG4qOjtZvv/0mT0/P23p9AAAAAMhOroLUe++9py5dutw0qPj5+Sk2NjbXhUnS2LFjVa5cOc2YMcPWFhERYfu3YRiaOHGiXn/9dT388MOSpNmzZysoKEhLlizRE088cVuvDwAAAADZydWpfRs3bsx2db6UlBT17t37tovKtHTpUtWvX19dunRRYGCg6tatq08++cS2PTY2VvHx8YqKirK1+fr6qmHDhtq6desNnzctLU3Jycl2NwAAAADIqVwFqVmzZunSpUtZ2i9duqTZs2ffdlGZ/vzzT02ZMkWVK1fW6tWr1b9/fw0cOFCzZs2SJMXHx0uSgoKC7B4XFBRk25ad0aNHy9fX13YrV65cntUMAAAAoPAzdWpfcnKyDMOQYRg6f/683al9GRkZWrFihQIDA/OsOKvVqvr162vUqFGSpLp162r//v2aOnWqevTokevnHTp0qAYPHmy7n5ycTJgCAAAAkGOmgpSfn58sFossFouqVKmSZbvFYtHw4cPzrLiQkBDVqFHDrq169epatGiRJCk4OFiSlJCQoJCQEFufhISEbBfAyOTh4SEPD488qxMAAADAncVUkNq4caMMw1CrVq20aNEilS5d2ratWLFiKl++vEJDQ/OsuMaNG+vQoUN2bb///rvKly8v6erCE8HBwVq/fr0tOCUnJ2v79u3q379/ntUBAAAAANcyFaSaN28u6eoiD2FhYbJYLPlSVKYXX3xRjRo10qhRo/TYY4/pp59+0rRp0zRt2jRJV4+ADRo0SG+//bYqV65sW/48NDRUHTt2zNfaAAAAANy5chykfvnlF9WsWVNubm5KSkrSvn37bti3du3aeVJcgwYNtHjxYg0dOlQjRoxQRESEJk6cqG7dutn6vPrqq0pJSVHfvn2VmJioJk2aaNWqVfyGFAAAAIB8k+MgVadOHcXHxyswMFB16tSRxWKRYRhZ+lksFmVkZORZgR06dFCHDh1uuN1isWjEiBEaMWJEnr0mAAAAANxMjoNUbGysAgICbP8GAAAAgDtVjoNU5gIP1/8bAAAAAO40uf5B3m+//dZ2/9VXX5Wfn58aNWqkY8eO5VlxAAAAAOCMchWkRo0aJS8vL0nS1q1b9eGHH2rcuHEqU6aMXnzxxTwtEAAAAACcjanlzzOdOHFClSpVkiQtWbJEjz76qPr27avGjRurRYsWeVkfAAAAADidXB2R8vb21j///CNJWrNmjdq0aSNJ8vT01KVLl/KuOgAAAABwQrk6ItWmTRs988wzqlu3rn7//Xc98MADkqRff/1V4eHheVkfAAAAADidXB2Rmjx5siIjI3XmzBktWrRI/v7+kqRdu3apa9eueVogAAAAADibXB2R8vPz04cffpilffjw4bddEAAAAAA4u1wFKUlKTEzUTz/9pNOnT8tqtdraLRaLunfvnifFAQAAAIAzylWQWrZsmbp166YLFy7Ix8dHFovFto0gBQAAAKCwy9U1Ui+99JJ69+6tCxcuKDExUefOnbPdzp49m9c1AgAAAIBTyVWQOnnypAYOHKjixYvndT0AAAAA4PRyFaSio6O1c+fOvK4FAAAAAFxCrq6Rat++vV555RX99ttvqlWrlooWLWq3/aGHHsqT4gAAAADAGeUqSPXp00eSNGLEiCzbLBaLMjIybq8qAAAAAHBiuQpS1y53DgAAAAB3mlxdI3Wt1NTUvKgDAAAAAFxGroJURkaGRo4cqbvuukve3t76888/JUlvvPGGpk+fnqcFAgAAAICzyVWQeueddzRz5kyNGzdOxYoVs7XXrFlTn376aZ4VBwAAAADOKFdBavbs2Zo2bZq6deumIkWK2NrvueceHTx4MM+KAwAAAABnlOsf5K1UqVKWdqvVqvT09NsuCgAAAACcWa6CVI0aNbRly5Ys7QsXLlTdunVvuygAAAAAcGa5Wv582LBh6tGjh06ePCmr1aqvv/5ahw4d0uzZs7V8+fK8rhEAAAAAnEqujkg9/PDDWrZsmdatW6cSJUpo2LBhOnDggJYtW6Y2bdrkdY0AAAAA4FRydURKkpo2baq1a9fmZS0AAAAA4BJydUSqQoUK+ueff7K0JyYmqkKFCrddFAAAAAA4s1wFqaNHjyojIyNLe1pamk6ePHnbRQEAAACAMzN1at/SpUtt/169erV8fX1t9zMyMrR+/XqFh4fnWXEAAAAA4IxMBamOHTtKkiwWi3r06GG3rWjRogoPD9d///vfPCsOAAAAAJyRqSBltVolSREREdqxY4fKlCmTL0UBAAAAgDPL1ap9sbGxeV0HAAAAALiMXC9/vn79eq1fv16nT5+2HanK9Nlnn912YQAAAADgrHIVpIYPH64RI0aofv36CgkJkcViyeu6AAAAAMBp5SpITZ06VTNnzlT37t3zuh4AAAAAcHq5+h2py5cvq1GjRnldCwAAAAC4hFwFqWeeeUZz587N61oAAAAAwCXk6tS+1NRUTZs2TevWrVPt2rVVtGhRu+0TJkzIk+IAAAAAwBnlKkj98ssvqlOnjiRp//79eVkPAAAAADi9XAWpjRs35nUdAAAAAOAyTAWpTp063bKPxWLRokWLcl0QAAAAADg7U0HK19c3v+oAAAAAAJdhKkjNmDEjv+oAAAAAAJeRq+XPAQAAAOBORpACAAAAAJMIUgAAAABgEkEKAAAAAEwiSAEAAACASQQpAAAAADCJIAUAAAAAJhGkAAAAAMAkghQAAAAAmESQAgAAAACTCFIAAAAAYBJBCgAAAABMIkgBAAAAgEkEKQAAAAAwiSAFAAAAACYRpAAAAADAJJcKUmPGjJHFYtGgQYNsbampqYqJiZG/v7+8vb3VuXNnJSQkOK5IAAAAAIWeywSpHTt26OOPP1bt2rXt2l988UUtW7ZMCxYs0ObNmxUXF6dOnTo5qEoAAAAAdwKXCFIXLlxQt27d9Mknn6hUqVK29qSkJE2fPl0TJkxQq1atVK9ePc2YMUM//vijtm3b5sCKAQAAABRmLhGkYmJi1L59e0VFRdm179q1S+np6Xbt1apVU1hYmLZu3XrD50tLS1NycrLdDQAAAAByyt3RBdzK/PnztXv3bu3YsSPLtvj4eBUrVkx+fn527UFBQYqPj7/hc44ePVrDhw/P61IBAAAA3CGc+ojUiRMn9MILL2jOnDny9PTMs+cdOnSokpKSbLcTJ07k2XMDAAAAKPycOkjt2rVLp0+f1r333it3d3e5u7tr8+bNmjRpktzd3RUUFKTLly8rMTHR7nEJCQkKDg6+4fN6eHjIx8fH7gYAAAAAOeXUp/a1bt1a+/bts2vr1auXqlWrptdee03lypVT0aJFtX79enXu3FmSdOjQIR0/flyRkZGOKBkAAADAHcCpg1TJkiVVs2ZNu7YSJUrI39/f1v70009r8ODBKl26tHx8fPT8888rMjJS//rXvxxRMgAAAIA7gFMHqZx477335Obmps6dOystLU3R0dH66KOPHF0WAAAAgELM5YLUpk2b7O57enpq8uTJmjx5smMKAgAAAHDHcerFJgAAAADAGRGkAAAAAMAkghQAAAAAmESQAgAAAACTCFIAAAAAYBJBCgAAAABMIkgBAAAAgEkEKQAAAAAwiSAFAAAAACYRpAAAAADAJIIUAAAAAJhEkAIAAAAAkwhSAAAAAGASQQoAAAAATCJIAQAAAIBJBCkAAAAAMIkgBQAAAAAmEaQAAAAAwCSCFAAAAACYRJACAAAAAJMIUgAAAABgEkEKAAAAAEwiSAEAAACASQQpAAAAADCJIAUAAAAAJhGkAAAAAMAkghQAAAAAmESQAgAAAACTCFIAAAAAYBJBCgAAAABMIkgBAAAAgEkEKQAAAAAwiSAFAAAAACYRpAAAAADAJIIUAAAAAJhEkAIAAAAAkwhSAAAAAGASQQoAAAAATCJIAQAAAIBJBCkAAAAAMIkgBQAAAAAmEaQAAAAAwCSCFAAAAACYRJACAAAAAJMIUgAAAABgEkEKAAAAAEwiSAEAAACASQQpAAAAADCJIAUAAAAAJhGkAAAAAMAkghQAAAAAmESQAgAAAACTCFIAAAAAYBJBCgAAAABMIkgBAAAAgEkEKQAAAAAwiSAFAAAAACYRpAAAAADAJIIUAAAAAJjk1EFq9OjRatCggUqWLKnAwEB17NhRhw4dsuuTmpqqmJgY+fv7y9vbW507d1ZCQoKDKgYAAABwJ3DqILV582bFxMRo27ZtWrt2rdLT09W2bVulpKTY+rz44otatmyZFixYoM2bNysuLk6dOnVyYNUAAAAACjt3RxdwM6tWrbK7P3PmTAUGBmrXrl1q1qyZkpKSNH36dM2dO1etWrWSJM2YMUPVq1fXtm3b9K9//csRZQMAAAAo5Jz6iNT1kpKSJEmlS5eWJO3atUvp6emKioqy9alWrZrCwsK0devWGz5PWlqakpOT7W4AAAAAkFMuE6SsVqsGDRqkxo0bq2bNmpKk+Ph4FStWTH5+fnZ9g4KCFB8ff8PnGj16tHx9fW23cuXK5WfpAAAAAAoZlwlSMTEx2r9/v+bPn3/bzzV06FAlJSXZbidOnMiDCgEAAADcKZz6GqlMAwYM0PLly/Xdd9+pbNmytvbg4GBdvnxZiYmJdkelEhISFBwcfMPn8/DwkIeHR36WDAAAAKAQc+ojUoZhaMCAAVq8eLE2bNigiIgIu+316tVT0aJFtX79elvboUOHdPz4cUVGRhZ0uQAAAADuEE59RComJkZz587VN998o5IlS9que/L19ZWXl5d8fX319NNPa/DgwSpdurR8fHz0/PPPKzIykhX7AAAAAOQbpw5SU6ZMkSS1aNHCrn3GjBnq2bOnJOm9996Tm5ubOnfurLS0NEVHR+ujjz4q4EoBAAAA3EmcOkgZhnHLPp6enpo8ebImT55cABUBAAAAgJNfIwUAAAAAzoggBQAAAAAmEaQAAAAAwCSCFAAAAACYRJACAAAAAJMIUgAAAABgEkEKAAAAAEwiSAEAAACASQQpAAAAADCJIAUAAAAAJhGkAAAAAMAkghQAAAAAmESQAgAAAACTCFIAAAAAYBJBCgAAAABMIkgBAAAAgEkEKQAAAAAwiSAFAAAAACYRpAAAAADAJIIUAAAAAJhEkAIAAAAAkwhSAAAAAGASQQoAAAAATCJIAQAAAIBJBCkAAAAAMIkgBQAAAAAmEaQAAAAAwCSCFAAAAACYRJACAAAAAJMIUgAAAABgEkEKAAAAAEwiSAEAAACASQQpAAAAADCJIAUAAAAAJhGkAAAAAMAkghQAAAAAmESQAgAAAACTCFIAAAAAYBJBCgAAAABMIkgBAAAAgEkEKQAAAAAwiSAFAAAAACYRpAAAAADAJIIUAAAAAJhEkAIAAAAAkwhSAAAAAGASQQoAAAAATCJIAQAAAIBJBCkAAAAAMIkgBQAAAAAmEaQAAAAAwCSCFAAAAACYRJACAAAAAJMIUgAAAABgEkEKAAAAAEwiSAEAAACASQQpAAAAADCJIAUAAAAAJhGkAAAAAMCkQhOkJk+erPDwcHl6eqphw4b66aefHF0SAAAAgEKqUASpL7/8UoMHD9abb76p3bt365577lF0dLROnz7t6NIAAAAAFEKFIkhNmDBBffr0Ua9evVSjRg1NnTpVxYsX12effebo0gAAAAAUQu6OLuB2Xb58Wbt27dLQoUNtbW5uboqKitLWrVuzfUxaWprS0tJs95OSkiRJycnJ+VvsLVitVl24cMFWi5tboci5Wdwp++ksXOX9dpU6HeFOfG/uhH2+E/bxWnfa/prhKu+Nq9RZWNwp77cz7mdmJjAM46b9XD5I/f3338rIyFBQUJBde1BQkA4ePJjtY0aPHq3hw4dnaS9Xrly+1AgAAADAtZw/f16+vr433O7yQSo3hg4dqsGDB9vuW61WnT17Vv7+/rJYLA6s7H+Sk5NVrlw5nThxQj4+Po4uB3mAMS1cGM/ChfEsfBjTwoXxLHyceUwNw9D58+cVGhp6034uH6TKlCmjIkWKKCEhwa49ISFBwcHB2T7Gw8NDHh4edm1+fn75VeJt8fHxcbrJhdvDmBYujGfhwngWPoxp4cJ4Fj7OOqY3OxKVyfEnId6mYsWKqV69elq/fr2tzWq1av369YqMjHRgZQAAAAAKK5c/IiVJgwcPVo8ePVS/fn3dd999mjhxolJSUtSrVy9HlwYAAACgECoUQerxxx/XmTNnNGzYMMXHx6tOnTpatWpVlgUoXImHh4fefPPNLKcgwnUxpoUL41m4MJ6FD2NauDCehU9hGFOLcat1/QAAAAAAdlz+GikAAAAAKGgEKQAAAAAwiSAFAAAAACYRpAAAAADAJIJUAfvuu+/04IMPKjQ0VBaLRUuWLLHbbhiGhg0bppCQEHl5eSkqKkqHDx+263P27Fl169ZNPj4+8vPz09NPP60LFy4U4F5AkkaPHq0GDRqoZMmSCgwMVMeOHXXo0CG7PqmpqYqJiZG/v7+8vb3VuXPnLD8effz4cbVv317FixdXYGCgXnnlFV25cqUgdwX/35QpU1S7dm3bjwNGRkZq5cqVtu2Mp2sbM2aMLBaLBg0aZGtjTF3LW2+9JYvFYnerVq2abTvj6XpOnjypJ598Uv7+/vLy8lKtWrW0c+dO23a+F7mW8PDwLJ9Ri8WimJgYSYXvM0qQKmApKSm65557NHny5Gy3jxs3TpMmTdLUqVO1fft2lShRQtHR0UpNTbX16datm3799VetXbtWy5cv13fffae+ffsW1C7g/9u8ebNiYmK0bds2rV27Vunp6Wrbtq1SUlJsfV588UUtW7ZMCxYs0ObNmxUXF6dOnTrZtmdkZKh9+/a6fPmyfvzxR82aNUszZ87UsGHDHLFLd7yyZctqzJgx2rVrl3bu3KlWrVrp4Ycf1q+//iqJ8XRlO3bs0Mcff6zatWvbtTOmrufuu+/WqVOnbLfvv//eto3xdC3nzp1T48aNVbRoUa1cuVK//fab/vvf/6pUqVK2Pnwvci07duyw+3yuXbtWktSlSxdJhfAzasBhJBmLFy+23bdarUZwcLDx7rvv2toSExMNDw8PY968eYZhGMZvv/1mSDJ27Nhh67Ny5UrDYrEYJ0+eLLDakdXp06cNScbmzZsNw7g6dkWLFjUWLFhg63PgwAFDkrF161bDMAxjxYoVhpubmxEfH2/rM2XKFMPHx8dIS0sr2B1AtkqVKmV8+umnjKcLO3/+vFG5cmVj7dq1RvPmzY0XXnjBMAw+o67ozTffNO65555stzGerue1114zmjRpcsPtfC9yfS+88IJRsWJFw2q1FsrPKEeknEhsbKzi4+MVFRVla/P19VXDhg21detWSdLWrVvl5+en+vXr2/pERUXJzc1N27dvL/Ca8T9JSUmSpNKlS0uSdu3apfT0dLvxrFatmsLCwuzGs1atWnY/Hh0dHa3k5GTbURA4RkZGhubPn6+UlBRFRkYyni4sJiZG7du3txs7ic+oqzp8+LBCQ0NVoUIFdevWTcePH5fEeLqipUuXqn79+urSpYsCAwNVt25dffLJJ7btfC9ybZcvX9YXX3yh3r17y2KxFMrPKEHKicTHx0uS3eTJvJ+5LT4+XoGBgXbb3d3dVbp0aVsfFDyr1apBgwapcePGqlmzpqSrY1WsWDH5+fnZ9b1+PLMb78xtKHj79u2Tt7e3PDw81K9fPy1evFg1atRgPF3U/PnztXv3bo0ePTrLNsbU9TRs2FAzZ87UqlWrNGXKFMXGxqpp06Y6f/484+mC/vzzT02ZMkWVK1fW6tWr1b9/fw0cOFCzZs2SxPciV7dkyRIlJiaqZ8+ekgrn31x3RxcAFAYxMTHav3+/3bn6cE1Vq1bV3r17lZSUpIULF6pHjx7avHmzo8tCLpw4cUIvvPCC1q5dK09PT0eXgzzQrl07279r166thg0bqnz58vrqq6/k5eXlwMqQG1arVfXr19eoUaMkSXXr1tX+/fs1depU9ejRw8HV4XZNnz5d7dq1U2hoqKNLyTcckXIiwcHBkpRl9ZKEhATbtuDgYJ0+fdpu+5UrV3T27FlbHxSsAQMGaPny5dq4caPKli1raw8ODtbly5eVmJho1//68cxuvDO3oeAVK1ZMlSpVUr169TR69Gjdc889ev/99xlPF7Rr1y6dPn1a9957r9zd3eXu7q7Nmzdr0qRJcnd3V1BQEGPq4vz8/FSlShX98ccffEZdUEhIiGrUqGHXVr16ddvpmnwvcl3Hjh3TunXr9Mwzz9jaCuNnlCDlRCIiIhQcHKz169fb2pKTk7V9+3ZFRkZKkiIjI5WYmKhdu3bZ+mzYsEFWq1UNGzYs8JrvZIZhaMCAAVq8eLE2bNigiIgIu+316tVT0aJF7cbz0KFDOn78uN147tu3z+4/AmvXrpWPj0+W/7jAMaxWq9LS0hhPF9S6dWvt27dPe/futd3q16+vbt262f7NmLq2Cxcu6MiRIwoJCeEz6oIaN26c5WdDfv/9d5UvX14S34tc2YwZMxQYGKj27dvb2grlZ9TRq13cac6fP2/s2bPH2LNnjyHJmDBhgrFnzx7j2LFjhmEYxpgxYww/Pz/jm2++MX755Rfj4YcfNiIiIoxLly7ZnuP+++836tata2zfvt34/vvvjcqVKxtdu3Z11C7dsfr372/4+voamzZtMk6dOmW7Xbx40danX79+RlhYmLFhwwZj586dRmRkpBEZGWnbfuXKFaNmzZpG27Ztjb179xqrVq0yAgICjKFDhzpil+54Q4YMMTZv3mzExsYav/zyizFkyBDDYrEYa9asMQyD8SwMrl21zzAYU1fz0ksvGZs2bTJiY2ONH374wYiKijLKlCljnD592jAMxtPV/PTTT4a7u7vxzjvvGIcPHzbmzJljFC9e3Pjiiy9sffhe5HoyMjKMsLAw47XXXsuyrbB9RglSBWzjxo2GpCy3Hj16GIZxdanPN954wwgKCjI8PDyM1q1bG4cOHbJ7jn/++cfo2rWr4e3tbfj4+Bi9evUyzp8/74C9ubNlN46SjBkzZtj6XLp0yXjuueeMUqVKGcWLFzceeeQR49SpU3bPc/ToUaNdu3aGl5eXUaZMGeOll14y0tPTC3hvYBiG0bt3b6N8+fJGsWLFjICAAKN169a2EGUYjGdhcH2QYkxdy+OPP26EhIQYxYoVM+666y7j8ccfN/744w/bdsbT9SxbtsyoWbOm4eHhYVSrVs2YNm2a3Xa+F7me1atXG5KyjJNhFL7PqMUwDMMhh8IAAAAAwEVxjRQAAAAAmESQAgAAAACTCFIAAAAAYBJBCgAAAABMIkgBAAAAgEkEKQAAAAAwiSAFAAAAACYRpAAAAADAJIIUAAAAAJhEkAIAFBo9e/aUxWJRv379smyLiYmRxWJRz549C74wAEChQ5ACABQq5cqV0/z583Xp0iVbW2pqqubOnauwsDAHVgYAKEwIUgCAQuXee+9VuXLl9PXXX9vavv76a4WFhalu3bq2tlWrVqlJkyby8/OTv7+/OnTooCNHjti2X758WQMGDFBISIg8PT1Vvnx5jR49WpJkGIbeeusthYWFycPDQ6GhoRo4cGDB7SQAwOEIUgCAQqd3796aMWOG7f5nn32mXr162fVJSUnR4MGDtXPnTq1fv15ubm565JFHZLVaJUmTJk3S0qVL9dVXX+nQoUOaM2eOwsPDJUmLFi3Se++9p48//liHDx/WkiVLVKtWrQLbPwCA47k7ugAAAPLak08+qaFDh+rYsWOSpB9++EHz58/Xpk2bbH06d+5s95jPPvtMAQEB+u2331SzZk0dP35clStXVpMmTWSxWFS+fHlb3+PHjys4OFhRUVEqWrSowsLCdN999xXIvgEAnANHpAAAhU5AQIDat2+vmTNnasaMGWrfvr3KlClj1+fw4cPq2rWrKlSoIB8fH9vRpuPHj0u6unDF3r17VbVqVQ0cOFBr1qyxPbZLly66dOmSKlSooD59+mjx4sW6cuVKge0fAMDxCFIAgEKpd+/emjlzpmbNmqXevXtn2f7ggw/q7Nmz+uSTT7R9+3Zt375d0tVro6Sr11rFxsZq5MiRunTpkh577DE9+uijkq4uaHHo0CF99NFH8vLy0nPPPadmzZopPT294HYQAOBQBCkAQKF0//336/Lly0pPT1d0dLTdtn/++UeHDh3S66+/rtatW6t69eo6d+5clufw8fHR448/rk8++URffvmlFi1apLNnz0qSvLy89OCDD2rSpEnatGmTtm7dqn379hXIvgEAHI9rpAAAhVKRIkV04MAB27+vVapUKfn7+2vatGkKCQnR8ePHNWTIELs+EyZMUEhIiOrWrSs3NzctWLBAwcHB8vPz08yZM5WRkaGGDRuqePHi+uKLL+Tl5WV3HRUAoHAjSAEACi0fH59s293c3DR//nwNHDhQNWvWVNWqVTVp0iS1aNHC1qdkyZIaN26cDh8+rCJFiqhBgwZasWKF3Nzc5OfnpzFjxmjw4MHKyMhQrVq1tGzZMvn7+xfQngEAHM1iGIbh6CIAAAAAwJVwjRQAAAAAmESQAgAAAACTCFIAAAAAYBJBCgAAAABMIkgBAAAAgEkEKQAAAAAwiSAFAAAAACYRpAAAAADAJIIUAAAAAJhEkAIAAAAAkwhSAAAAAGDS/wM4JL9F4ZiiUwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1000x500 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "peptide = 'PEPTIDE'\n",
    "\n",
    "frag_dict = get_frag_dict(parse(peptide), constants.mass_dict)\n",
    "\n",
    "db_frag = list(frag_dict.values())\n",
    "db_int = [100 for _ in db_frag]\n",
    "\n",
    "plt.figure(figsize=(10,5))\n",
    "plt.vlines(db_frag, 0, db_int, \"k\", label=\"DB\", alpha=0.2)\n",
    "\n",
    "for _ in frag_dict.keys():\n",
    "    plt.text(frag_dict[_], 104, _, fontsize=12, alpha = 0.8)\n",
    "    \n",
    "plt.title('Theoretical Spectrum for {}'.format(peptide))\n",
    "plt.xlabel('Mass')\n",
    "plt.ylabel('Intensity')\n",
    "plt.ylim([0,110])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Spectra\n",
    "\n",
    "The function [`get_spectrum`](https://mannlabs.github.io/alphapept/fasta.html#get_spectrum) returns a tuple with the following content:\n",
    "\n",
    "* precursor mass\n",
    "* peptide sequence\n",
    "* fragmasses\n",
    "* fragtypes\n",
    "\n",
    "Likewise, [`get_spectra`](https://mannlabs.github.io/alphapept/fasta.html#get_spectra) returns a list of tuples. We employ a list of tuples here as this way, we can sort them easily by precursor mass."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L630){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_spectra\n",
       "\n",
       ">      get_spectra (peptides:numba.typed.typedlist.List,\n",
       ">                   mass_dict:numba.typed.typeddict.Dict)\n",
       "\n",
       "Get neutral peptide mass, fragment masses and fragment types for a list of peptides\n",
       "Args:\n",
       "    peptides (list of str): the (modified) peptide list.\n",
       "    mass_dict (numba.typed.Dict): key is the amino acid or modified amino acid, and the value is the mass.\n",
       "Raises:\n",
       "    Unknown exception and pass.\n",
       "Returns:\n",
       "    list of Tuple[float, str, np.ndarray(np.float64), np.ndarray(np.int8)]: See get_spectrum."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L630){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_spectra\n",
       "\n",
       ">      get_spectra (peptides:numba.typed.typedlist.List,\n",
       ">                   mass_dict:numba.typed.typeddict.Dict)\n",
       "\n",
       "Get neutral peptide mass, fragment masses and fragment types for a list of peptides\n",
       "Args:\n",
       "    peptides (list of str): the (modified) peptide list.\n",
       "    mass_dict (numba.typed.Dict): key is the amino acid or modified amino acid, and the value is the mass.\n",
       "Raises:\n",
       "    Unknown exception and pass.\n",
       "Returns:\n",
       "    list of Tuple[float, str, np.ndarray(np.float64), np.ndarray(np.int8)]: See get_spectrum."
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(get_spectra)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L609){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_spectrum\n",
       "\n",
       ">      get_spectrum (peptide:str, mass_dict:numba.typed.typeddict.Dict)\n",
       "\n",
       "Get neutral peptide mass, fragment masses and fragment types for a peptide\n",
       "Args:\n",
       "    peptide (str): the (modified) peptide.\n",
       "    mass_dict (numba.typed.Dict): key is the amino acid or modified amino acid, and the value is the mass.\n",
       "Returns:\n",
       "    Tuple[float, str, np.ndarray(np.float64), np.ndarray(np.int8)]: (peptide mass, peptide, fragment masses, fragment_types), for fragment types, see get_fragmass."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L609){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_spectrum\n",
       "\n",
       ">      get_spectrum (peptide:str, mass_dict:numba.typed.typeddict.Dict)\n",
       "\n",
       "Get neutral peptide mass, fragment masses and fragment types for a peptide\n",
       "Args:\n",
       "    peptide (str): the (modified) peptide.\n",
       "    mass_dict (numba.typed.Dict): key is the amino acid or modified amino acid, and the value is the mass.\n",
       "Returns:\n",
       "    Tuple[float, str, np.ndarray(np.float64), np.ndarray(np.int8)]: (peptide mass, peptide, fragment masses, fragment_types), for fragment types, see get_fragmass."
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(get_spectrum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(799.35996420346, 'PEPTIDE', array([ 98.06004033, 148.06043425, 227.10263343, 263.08737735,\n",
      "       324.15539729, 376.17144135, 425.20307579, 477.21911985,\n",
      "       538.28713979, 574.27188371, 653.31408289, 703.31447681]), array([ 1, -1,  2, -2,  3, -3,  4, -4,  5, -5,  6, -6], dtype=int8)), ...]\n"
     ]
    }
   ],
   "source": [
    "print(get_spectra(List(['PEPTIDE']), constants.mass_dict))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading FASTA\n",
    "\n",
    "To read FASTA files, we use the `SeqIO` module from the `Biopython` library. This is a generator expression so that we read one FASTA entry after another until the `StopIteration` is reached, which is implemented in [`read_fasta_file`](https://mannlabs.github.io/alphapept/fasta.html#read_fasta_file). Additionally, we define the function [`read_fasta_file_entries`](https://mannlabs.github.io/alphapept/fasta.html#read_fasta_file_entries) that simply counts the number of FASTA entries.\n",
    "\n",
    "All FASTA entries that contain AAs which are not in the mass_dict can be checked with [`check_sequence`](https://mannlabs.github.io/alphapept/fasta.html#check_sequence) and will be ignored."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L709){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### check_sequence\n",
       "\n",
       ">      check_sequence (element:dict, AAs:set, verbose:bool=False)\n",
       "\n",
       "Checks wheter a sequence from a FASTA entry contains valid AAs\n",
       "Args:\n",
       "    element (dict): fasta entry of the protein information.\n",
       "    AAs (set): a set of amino acid letters.\n",
       "    verbose (bool): logging the invalid amino acids.\n",
       "Returns:\n",
       "    bool: False if the protein sequence contains non-AA letters, otherwise True."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L709){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### check_sequence\n",
       "\n",
       ">      check_sequence (element:dict, AAs:set, verbose:bool=False)\n",
       "\n",
       "Checks wheter a sequence from a FASTA entry contains valid AAs\n",
       "Args:\n",
       "    element (dict): fasta entry of the protein information.\n",
       "    AAs (set): a set of amino acid letters.\n",
       "    verbose (bool): logging the invalid amino acids.\n",
       "Returns:\n",
       "    bool: False if the protein sequence contains non-AA letters, otherwise True."
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(check_sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L688){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### read_fasta_file_entries\n",
       "\n",
       ">      read_fasta_file_entries (fasta_filename='')\n",
       "\n",
       "Function to count entries in fasta file\n",
       "Args:\n",
       "    fasta_filename (str): fasta.\n",
       "Returns:\n",
       "    int: number of entries."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L688){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### read_fasta_file_entries\n",
       "\n",
       ">      read_fasta_file_entries (fasta_filename='')\n",
       "\n",
       "Function to count entries in fasta file\n",
       "Args:\n",
       "    fasta_filename (str): fasta.\n",
       "Returns:\n",
       "    int: number of entries."
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(read_fasta_file_entries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L657){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### read_fasta_file\n",
       "\n",
       ">      read_fasta_file (fasta_filename:str='')\n",
       "\n",
       "Read a FASTA file line by line\n",
       "Args:\n",
       "    fasta_filename (str): fasta.\n",
       "Yields:\n",
       "    dict {id:str, name:str, description:str, sequence:str}: protein information."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L657){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### read_fasta_file\n",
       "\n",
       ">      read_fasta_file (fasta_filename:str='')\n",
       "\n",
       "Read a FASTA file line by line\n",
       "Args:\n",
       "    fasta_filename (str): fasta.\n",
       "Yields:\n",
       "    dict {id:str, name:str, description:str, sequence:str}: protein information."
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(read_fasta_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'id': 'A0PJZ0',\n",
       " 'name': 'sp|A0PJZ0|A20A5_HUMAN',\n",
       " 'description': 'sp|A0PJZ0|A20A5_HUMAN Putative ankyrin repeat domain-containing protein 20A5 OS=Homo sapiens OX=9606 GN=ANKRD20A5P PE=5 SV=1',\n",
       " 'sequence': 'MKLFGFRSRRGQTVLGSIDHLYTGSGYRIRYSELQKIHKAAVKGDAAEMERCLARRSGDLDALDKQHRTALHLACASGHVKVVTLLVNRKCQIDIYDKENRTPLIQAVHCQEEACAVILLEHGANPNLKDIYGNTALHYAVYSESTSLAEKLLFHGENIEALDKV'}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load example fasta file\n",
    "\n",
    "fasta_path = '../testfiles/test.fasta'\n",
    "\n",
    "list(read_fasta_file(fasta_path))[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Peptide Dictionary\n",
    "\n",
    "In order to efficiently store peptides, we rely on the Python dictionary. The idea is to have a dictionary with peptides as keys and indices to proteins as values. This way, one can quickly look up to which protein a peptide belongs to. The function [`add_to_pept_dict`](https://mannlabs.github.io/alphapept/fasta.html#add_to_pept_dict) uses a regular python dictionary and allows to add peptides and stores indices to the originating proteins as a list. If a peptide is already present in the dictionary, the list is appended. The function returns a list of `added_peptides`, which were not present in the dictionary yet. One can use the function [`merge_pept_dicts`](https://mannlabs.github.io/alphapept/fasta.html#merge_pept_dicts) to merge multiple peptide dicts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L730){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_to_pept_dict\n",
       "\n",
       ">      add_to_pept_dict (pept_dict:dict, new_peptides:list, i:int)\n",
       "\n",
       "Add peptides to the peptide dictionary\n",
       "Args:\n",
       "    pept_dict (dict): the key is peptide sequence, and the value is protein id list indicating where the peptide is from.\n",
       "    new_peptides (list): the list of peptides to be added to pept_dict.\n",
       "    i (int): the protein id where new_peptides are from.\n",
       "Returns:\n",
       "    dict: same as the pept_dict in the arguments.\n",
       "    list (of str): the peptides from new_peptides that not in the pept_dict."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L730){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### add_to_pept_dict\n",
       "\n",
       ">      add_to_pept_dict (pept_dict:dict, new_peptides:list, i:int)\n",
       "\n",
       "Add peptides to the peptide dictionary\n",
       "Args:\n",
       "    pept_dict (dict): the key is peptide sequence, and the value is protein id list indicating where the peptide is from.\n",
       "    new_peptides (list): the list of peptides to be added to pept_dict.\n",
       "    i (int): the protein id where new_peptides are from.\n",
       "Returns:\n",
       "    dict: same as the pept_dict in the arguments.\n",
       "    list (of str): the peptides from new_peptides that not in the pept_dict."
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(add_to_pept_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ABC': [0], 'DEF': [0, 1], 'GHI': [1]}\n"
     ]
    }
   ],
   "source": [
    "pept_dict = {}\n",
    "new_peptides = ['ABC','DEF']\n",
    "\n",
    "pept_dict, added_peptides = add_to_pept_dict(pept_dict, new_peptides, 0)\n",
    "\n",
    "new_peptides = ['DEF','GHI']\n",
    "\n",
    "pept_dict, added_peptides = add_to_pept_dict(pept_dict, new_peptides, 1)\n",
    "\n",
    "print(pept_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L753){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### merge_pept_dicts\n",
       "\n",
       ">      merge_pept_dicts (list_of_pept_dicts:list)\n",
       "\n",
       "Merge a list of peptide dict into a single dict.\n",
       "Args:\n",
       "    list_of_pept_dicts (list of dict): the key of the pept_dict is peptide sequence, and the value is protein id list indicating where the peptide is from.\n",
       "Returns:\n",
       "    dict: the key is peptide sequence, and the value is protein id list indicating where the peptide is from."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L753){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### merge_pept_dicts\n",
       "\n",
       ">      merge_pept_dicts (list_of_pept_dicts:list)\n",
       "\n",
       "Merge a list of peptide dict into a single dict.\n",
       "Args:\n",
       "    list_of_pept_dicts (list of dict): the key of the pept_dict is peptide sequence, and the value is protein id list indicating where the peptide is from.\n",
       "Returns:\n",
       "    dict: the key is peptide sequence, and the value is protein id list indicating where the peptide is from."
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(merge_pept_dicts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ABC': [0, 3, 4], 'DEF': [0, 1], 'GHI': [1], 'JKL': [5, 6], 'MNO': [7]}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pept_dict_1 = {'ABC': [0], 'DEF': [0, 1], 'GHI': [1]}\n",
    "pept_dict_2 = {'ABC': [3,4], 'JKL': [5, 6], 'MNO': [7]}\n",
    "\n",
    "merge_pept_dicts([pept_dict_1, pept_dict_2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating a database\n",
    "\n",
    "To wrap everything up, we employ two functions, [`generate_database`](https://mannlabs.github.io/alphapept/fasta.html#generate_database) and [`generate_spectra`](https://mannlabs.github.io/alphapept/fasta.html#generate_spectra). The first one reads a FASTA file and generates a list of peptides, as well as the peptide dictionary and an ordered FASTA dictionary to be able to look up the protein indices later. For the `callback` we first read the whole FASTA file to determine the total number of entries in the FASTA file.  For a typical FASTA file of 30 Mb with 40k entries, this should take less than a second. The progress of the digestion is monitored by processing the FASTA file one by one.\n",
    "The function [`generate_spectra`](https://mannlabs.github.io/alphapept/fasta.html#generate_spectra) then calculates precursor masses and fragment ions. Here, we split the total_number of sequences in `1000` steps to be able to track progress with the `callback`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L780){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### generate_fasta_list\n",
       "\n",
       ">      generate_fasta_list (fasta_paths:list, callback=None, **kwargs)\n",
       "\n",
       "Function to generate a database from a fasta file\n",
       "Args:\n",
       "    fasta_paths (str or list of str): fasta path or a list of fasta paths.\n",
       "    callback (function, optional): callback function.\n",
       "Returns:\n",
       "    fasta_list (list of dict): list of protein entry dict {id:str, name:str, description:str, sequence:str}.\n",
       "    fasta_dict (dict{int:dict}): the key is the protein id, the value is the protein entry dict."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L780){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### generate_fasta_list\n",
       "\n",
       ">      generate_fasta_list (fasta_paths:list, callback=None, **kwargs)\n",
       "\n",
       "Function to generate a database from a fasta file\n",
       "Args:\n",
       "    fasta_paths (str or list of str): fasta path or a list of fasta paths.\n",
       "    callback (function, optional): callback function.\n",
       "Returns:\n",
       "    fasta_list (list of dict): list of protein entry dict {id:str, name:str, description:str, sequence:str}.\n",
       "    fasta_dict (dict{int:dict}): the key is the protein id, the value is the protein entry dict."
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(generate_fasta_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L820){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### generate_database\n",
       "\n",
       ">      generate_database (mass_dict:dict, fasta_paths:list, callback=None,\n",
       ">                         **kwargs)\n",
       "\n",
       "Function to generate a database from a fasta file\n",
       "Args:\n",
       "    mass_dict (dict): not used, will be removed in the future.\n",
       "    fasta_paths (str or list of str): fasta path or a list of fasta paths.\n",
       "    callback (function, optional): callback function.\n",
       "Returns:\n",
       "    to_add (list of str): non-redundant (modified) peptides to be added.\n",
       "    pept_dict (dict{str:list of int}): the key is peptide sequence, and the value is protein id list indicating where the peptide is from.\n",
       "    fasta_dict (dict{int:dict}): the key is the protein id, the value is the protein entry dict {id:str, name:str, description:str, sequence:str}."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L820){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### generate_database\n",
       "\n",
       ">      generate_database (mass_dict:dict, fasta_paths:list, callback=None,\n",
       ">                         **kwargs)\n",
       "\n",
       "Function to generate a database from a fasta file\n",
       "Args:\n",
       "    mass_dict (dict): not used, will be removed in the future.\n",
       "    fasta_paths (str or list of str): fasta path or a list of fasta paths.\n",
       "    callback (function, optional): callback function.\n",
       "Returns:\n",
       "    to_add (list of str): non-redundant (modified) peptides to be added.\n",
       "    pept_dict (dict{str:list of int}): the key is peptide sequence, and the value is protein id list indicating where the peptide is from.\n",
       "    fasta_dict (dict{int:dict}): the key is the protein id, the value is the protein entry dict {id:str, name:str, description:str, sequence:str}."
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(generate_database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L866){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### generate_spectra\n",
       "\n",
       ">      generate_spectra (to_add:list, mass_dict:dict, callback=None)\n",
       "\n",
       "Function to generate spectra list database from a fasta file\n",
       "Args:\n",
       "    to_add (list): \n",
       "    mass_dict (dict{str:float}): amino acid mass dict.\n",
       "    callback (function, optional): callback function. (Default: None)\n",
       "Returns:\n",
       "    list (of tuple): list of (peptide mass, peptide, fragment masses, fragment_types), see get_fragmass."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L866){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### generate_spectra\n",
       "\n",
       ">      generate_spectra (to_add:list, mass_dict:dict, callback=None)\n",
       "\n",
       "Function to generate spectra list database from a fasta file\n",
       "Args:\n",
       "    to_add (list): \n",
       "    mass_dict (dict{str:float}): amino acid mass dict.\n",
       "    callback (function, optional): callback function. (Default: None)\n",
       "Returns:\n",
       "    list (of tuple): list of (peptide mass, peptide, fragment masses, fragment_types), see get_fragmass."
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(generate_spectra)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parallelized version\n",
    "\n",
    "To speed up spectra generated, one can use the parallelized version. The function [`generate_database_parallel`](https://mannlabs.github.io/alphapept/fasta.html#generate_database_parallel) reads an entire FASTA file and splits it into multiple blocks. Each block will be processed, and the generated pept_dicts will be merged."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L919){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### blocks\n",
       "\n",
       ">      blocks (l:int, n:int)\n",
       "\n",
       "Helper function to create blocks from a given list\n",
       "Args:\n",
       "    l (list): the list\n",
       "    n (int): size per block\n",
       "Returns:\n",
       "    Generator: List with splitted elements"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L919){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### blocks\n",
       "\n",
       ">      blocks (l:int, n:int)\n",
       "\n",
       "Helper function to create blocks from a given list\n",
       "Args:\n",
       "    l (list): the list\n",
       "    n (int): size per block\n",
       "Returns:\n",
       "    Generator: List with splitted elements"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(blocks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L899){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### block_idx\n",
       "\n",
       ">      block_idx (len_list:int, block_size:int=1000)\n",
       "\n",
       "Helper function to split length into blocks\n",
       "Args:\n",
       "    len_list (int): list length.\n",
       "    block_size (int, optional, default 1000): size per block.\n",
       "Returns:\n",
       "    list[(int, int)]: list of (start, end) positions of blocks."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L899){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### block_idx\n",
       "\n",
       ">      block_idx (len_list:int, block_size:int=1000)\n",
       "\n",
       "Helper function to split length into blocks\n",
       "Args:\n",
       "    len_list (int): list length.\n",
       "    block_size (int, optional, default 1000): size per block.\n",
       "Returns:\n",
       "    list[(int, int)]: list of (start, end) positions of blocks."
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(block_idx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L967){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### generate_database_parallel\n",
       "\n",
       ">      generate_database_parallel (settings:dict, callback=None)\n",
       "\n",
       "Function to generate a database from a fasta file in parallel.\n",
       "Args:\n",
       "    settings: alphapept settings.\n",
       "Returns:\n",
       "    list: theoretical spectra. See generate_spectra()\n",
       "    dict: peptide dict. See add_to_pept_dict()\n",
       "    dict: fasta_dict. See generate_fasta_list()"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L967){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### generate_database_parallel\n",
       "\n",
       ">      generate_database_parallel (settings:dict, callback=None)\n",
       "\n",
       "Function to generate a database from a fasta file in parallel.\n",
       "Args:\n",
       "    settings: alphapept settings.\n",
       "Returns:\n",
       "    list: theoretical spectra. See generate_spectra()\n",
       "    dict: peptide dict. See add_to_pept_dict()\n",
       "    dict: fasta_dict. See generate_fasta_list()"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(generate_database_parallel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L937){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### digest_fasta_block\n",
       "\n",
       ">      digest_fasta_block (to_process:tuple)\n",
       "\n",
       "Digest and create spectra for a whole fasta_block for multiprocessing. See generate_database_parallel."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L937){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### digest_fasta_block\n",
       "\n",
       ">      digest_fasta_block (to_process:tuple)\n",
       "\n",
       "Digest and create spectra for a whole fasta_block for multiprocessing. See generate_database_parallel."
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(digest_fasta_block)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parallel search on large files\n",
    "\n",
    "In some cases (e.g., a lot of modifications or very large FASTA files), it will not be useful to save the database as it will consume too much memory. Here, we use the function [`search_parallel`](https://mannlabs.github.io/alphapept/search.html#search_parallel) from search. It creates theoretical spectra on the fly and directly searches against them. As we cannot create a pept_dict here, we need to create one from the search results. For this, we group peptides by their FASTA index and generate a lookup dictionary that can be used as a pept_dict.\n",
    "\n",
    "> Note that we are passing the settings argument here. Search results should be stored in the corresponding path in the `*.hdf` file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L1011){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### pept_dict_from_search\n",
       "\n",
       ">      pept_dict_from_search (settings:dict)\n",
       "\n",
       "Generates a peptide dict from a large search."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L1011){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### pept_dict_from_search\n",
       "\n",
       ">      pept_dict_from_search (settings:dict)\n",
       "\n",
       "Generates a peptide dict from a large search."
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(pept_dict_from_search)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Saving\n",
    "\n",
    "To save the generated spectra, we rely on the HDF format. For this, we create a dictionary and save all the generated elements. The container will contain the following elements:\n",
    "\n",
    "* `precursors`: An array containing the precursor masses\n",
    "* `seqs`: An array containing the peptide sequences for the precursor masses\n",
    "* `pept_dict`: A peptide dictionary to look up the peptides and return their FASTA index\n",
    "* `fasta_dict`: A FASTA dictionary to look up the FASTA entry based on a pept_dict index\n",
    "* `fragmasses`: An array containing the fragment masses. Unoccupied cells are filled with -1\n",
    "* `fragtypes:`: An array containing the fragment types. 0 equals b-ions, and 1 equals y-ions. Unoccupied cells are filled with -1\n",
    "* `bounds`: An integer array containing the upper bounds for the fragment masses/types array. This is needed to quickly slice the data.\n",
    "\n",
    "All arrays are sorted according to the precursor mass.\n",
    "\n",
    ":::{.callout-note}\n",
    "\n",
    "To access the dictionaries such as `pept_dict` or `fasta_dict`, one needs to extract them using the `.item()` method like so: `container[\"pept_dict\"].item()`.\n",
    "\n",
    ":::"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L1056){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### save_database\n",
       "\n",
       ">      save_database (spectra:list, pept_dict:dict, fasta_dict:dict,\n",
       ">                     database_path:str, **kwargs)\n",
       "\n",
       "Function to save a database to the *.hdf format. Write the database into hdf.\n",
       "\n",
       "Args:\n",
       "    spectra (list): list: theoretical spectra. See generate_spectra().\n",
       "    pept_dict (dict): peptide dict. See add_to_pept_dict().\n",
       "    fasta_dict (dict): fasta_dict. See generate_fasta_list().\n",
       "    database_path (str): Path to database."
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L1056){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### save_database\n",
       "\n",
       ">      save_database (spectra:list, pept_dict:dict, fasta_dict:dict,\n",
       ">                     database_path:str, **kwargs)\n",
       "\n",
       "Function to save a database to the *.hdf format. Write the database into hdf.\n",
       "\n",
       "Args:\n",
       "    spectra (list): list: theoretical spectra. See generate_spectra().\n",
       "    pept_dict (dict): peptide dict. See add_to_pept_dict().\n",
       "    fasta_dict (dict): fasta_dict. See generate_fasta_list().\n",
       "    database_path (str): Path to database."
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(save_database)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L1132){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### read_database\n",
       "\n",
       ">      read_database (database_path:str, array_name:str=None)\n",
       "\n",
       "Read database from hdf file.\n",
       "Args:\n",
       "    database_path (str): hdf database file generate by alphapept.\n",
       "    array_name (str): the dataset name to read\n",
       "return:\n",
       "    dict: key is the dataset_name in hdf file, value is the python object read from the dataset_name"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/mannlabs/alphapept/blob/master/alphapept/fasta.py#L1132){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### read_database\n",
       "\n",
       ">      read_database (database_path:str, array_name:str=None)\n",
       "\n",
       "Read database from hdf file.\n",
       "Args:\n",
       "    database_path (str): hdf database file generate by alphapept.\n",
       "    array_name (str): the dataset name to read\n",
       "return:\n",
       "    dict: key is the dataset_name in hdf file, value is the python object read from the dataset_name"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(read_database)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
