{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp settings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Settings\n",
    "\n",
    "> A template for settings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Things that could be added:\n",
    "* Check to make sure that settings are not contradicting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from nbdev.showdoc import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Settings\n",
    "\n",
    "### Saving and Loading\n",
    "\n",
    "The default scheme for saving settings are yaml file. This allows easy editing via a text editor."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#export\n",
    "import yaml\n",
    "\n",
    "def print_settings(settings):\n",
    "    \"\"\"\n",
    "    Print a yaml settings file\n",
    "    \"\"\"\n",
    "    print(yaml.dump(settings, default_flow_style=False))\n",
    "\n",
    "\n",
    "def load_settings(path):\n",
    "    \"\"\"\n",
    "    Load a yaml settings file\n",
    "    \"\"\"\n",
    "    with open(path, \"r\") as settings_file:\n",
    "        SETTINGS_LOADED = yaml.load(settings_file, Loader=yaml.FullLoader)\n",
    "        return SETTINGS_LOADED\n",
    "\n",
    "def save_settings(settings, path):\n",
    "    \"\"\"\n",
    "    Save a yaml settings file to path\n",
    "    \"\"\"\n",
    "    with open(path, \"w\") as file:\n",
    "        yaml.dump(settings, file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Settings Template\n",
    "\n",
    "The settings template defines individual settings. The idea is to provide a template so that a graphical user interface can be automatically generated. \n",
    "\n",
    "Each entry has a type, default values, and a description.\n",
    "\n",
    "* spinbox -> QSpinBox with minimum and maximum values (int)\n",
    "* doublespinbox -> QDoubleSpinBox with minimum and maximum values (float)\n",
    "* path -> Clickable button to select a path to save / load files.\n",
    "* combobox -> QComboBox, dropdown menu with values to choose from\n",
    "* checkbox -> QCheckBox, checkbox that can be selected\n",
    "* checkgroup -> Creates a list of QCheckBox options that can be selected\n",
    "* list -> Creates a list that is displayed\n",
    "* placeholder -> This just prints the parameter and cannot be changed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "calibration:\n",
      "  method:\n",
      "    default: linear\n",
      "    description: Interpolation method.\n",
      "    type: combobox\n",
      "    value:\n",
      "    - linear\n",
      "    - nearest\n",
      "    - cubic\n",
      "  min_mz_step:\n",
      "    default: 80\n",
      "    description: Minimum number of datapoints in for mz interpolation.\n",
      "    max: 999\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  min_rt_step:\n",
      "    default: 50\n",
      "    description: Minimum number of datapoints in for rt interpolation.\n",
      "    max: 999\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  minimum_score:\n",
      "    default: 20\n",
      "    description: Minimum score for psms used for calibration.\n",
      "    max: 999\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  outlier_std:\n",
      "    default: 3\n",
      "    description: Number of std. deviations to filter outliers in psms.\n",
      "    max: 5\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "experiment:\n",
      "  file_paths:\n",
      "    default: []\n",
      "    description: Filepaths of the experiments.\n",
      "    type: list\n",
      "  fractions:\n",
      "    default: []\n",
      "    description: List of fractions for the raw files.\n",
      "    type: list\n",
      "  results_path:\n",
      "    default: null\n",
      "    description: Path where the results should be stored.\n",
      "    filetype:\n",
      "    - hdf\n",
      "    folder: false\n",
      "    type: path\n",
      "  shortnames:\n",
      "    default: []\n",
      "    description: List of shortnames for the raw files.\n",
      "    type: list\n",
      "fasta:\n",
      "  AL_swap:\n",
      "    default: false\n",
      "    description: Swap A and L for decoy generation.\n",
      "    type: checkbox\n",
      "  KR_swap:\n",
      "    default: false\n",
      "    description: Swap K and R (only if terminal) for decoy generation.\n",
      "    type: checkbox\n",
      "  database_path:\n",
      "    default: null\n",
      "    description: Path to library file (.hdf).\n",
      "    filetype:\n",
      "    - hdf\n",
      "    folder: false\n",
      "    type: path\n",
      "  fasta_block:\n",
      "    default: 1000\n",
      "    description: Number of fasta entries to be processed in one block.\n",
      "    max: 10000\n",
      "    min: 100\n",
      "    type: spinbox\n",
      "  fasta_paths:\n",
      "    default: []\n",
      "    description: List of paths for FASTA files.\n",
      "    type: list\n",
      "  max_isoforms:\n",
      "    default: 1024\n",
      "    description: Maximum number of isoforms per peptide.\n",
      "    max: 4096\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  max_length:\n",
      "    default: 27\n",
      "    description: Maximum peptide length.\n",
      "    max: 99\n",
      "    min: 6\n",
      "    type: spinbox\n",
      "  min_length:\n",
      "    default: 6\n",
      "    description: Minimum peptide length.\n",
      "    max: 99\n",
      "    min: 6\n",
      "    type: spinbox\n",
      "  mods_fixed:\n",
      "    default:\n",
      "    - cC\n",
      "    description: Fixed modifications.\n",
      "    type: checkgroup\n",
      "    value: &id001\n",
      "      aK: acetylation of lysine\n",
      "      cC: carbamidomethylation of C\n",
      "      deamN: deamidation of N\n",
      "      deamNQ: deamidation of Q\n",
      "      eK: EASItag 6-plex on K\n",
      "      itraq4K: iTRAQ 4-plex on K\n",
      "      itraq4Y: iTRAQ 4-plex on Y\n",
      "      itraq8K: iTRAQ 8-plex on K\n",
      "      itraq8Y: iTRAQ 8-plex on Y\n",
      "      oxM: oxidation of M\n",
      "      pS: phosphorylation of S\n",
      "      pT: phosphorylation of T\n",
      "      pY: phosphorylation of Y\n",
      "      tmt0K: TMT zero on K\n",
      "      tmt0Y: TMT zero on Y\n",
      "      tmt2K: TMT duplex on K\n",
      "      tmt2Y: TMT duplex on Y\n",
      "      tmt6K: TMT sixplex/tenplex on K\n",
      "      tmt6Y: TMT sixplex/tenplex on Y\n",
      "  mods_fixed_terminal:\n",
      "    default: []\n",
      "    description: Fixed terminal modifications.\n",
      "    type: checkgroup\n",
      "    value: &id002\n",
      "      cmC<: pyro-cmC\n",
      "      e<^: EASItag 6-plex on peptide N-terminus\n",
      "      itraq4K<^: iTRAQ 4-plex on peptide N-terminus\n",
      "      itraq8K<^: iTRAQ 8-plex on peptide N-terminus\n",
      "      pgE<: pyro-E\n",
      "      pgQ<: pyro-Q\n",
      "      tmt0<^: TMT zero on peptide N-terminus\n",
      "      tmt2<^: TMT duplex on peptide N-terminus\n",
      "      tmt6<^: TMT sixplex/tenplex on peptide N-terminus\n",
      "  mods_fixed_terminal_prot:\n",
      "    default: []\n",
      "    description: Fixed terminal modifications on proteins.\n",
      "    type: checkgroup\n",
      "    value: &id003\n",
      "      a<^: acetylation of protein N-terminus\n",
      "      am>^: amidation of protein C-terminus\n",
      "  mods_variable:\n",
      "    default:\n",
      "    - oxM\n",
      "    description: Fixed variable modifications.\n",
      "    type: checkgroup\n",
      "    value: *id001\n",
      "  mods_variable_terminal:\n",
      "    default: []\n",
      "    description: Fixed varibale terminal modifications.\n",
      "    type: checkgroup\n",
      "    value: *id002\n",
      "  mods_variable_terminal_prot:\n",
      "    default: []\n",
      "    description: Fixed varibale terminal modifications  on proteins.\n",
      "    type: checkgroup\n",
      "    value: *id003\n",
      "  num_missed_cleavages:\n",
      "    default: 2\n",
      "    description: Number of missed cleavages.\n",
      "    max: 99\n",
      "    min: 0\n",
      "    type: spinbox\n",
      "  protease:\n",
      "    default: trypsin\n",
      "    description: Protease for digestions.\n",
      "    type: combobox\n",
      "    value:\n",
      "    - arg-c\n",
      "    - asp-n\n",
      "    - bnps-skatole\n",
      "    - caspase 1\n",
      "    - caspase 2\n",
      "    - caspase 3\n",
      "    - caspase 4\n",
      "    - caspase 5\n",
      "    - caspase 6\n",
      "    - caspase 7\n",
      "    - caspase 8\n",
      "    - caspase 9\n",
      "    - caspase 10\n",
      "    - chymotrypsin high specificity\n",
      "    - chymotrypsin low specificity\n",
      "    - clostripain\n",
      "    - cnbr\n",
      "    - enterokinase\n",
      "    - factor xa\n",
      "    - formic acid\n",
      "    - glutamyl endopeptidase\n",
      "    - granzyme b\n",
      "    - hydroxylamine\n",
      "    - iodosobenzoic acid\n",
      "    - lysc\n",
      "    - ntcb\n",
      "    - pepsin ph1.3\n",
      "    - pepsin ph2.0\n",
      "    - proline endopeptidase\n",
      "    - proteinase k\n",
      "    - staphylococcal peptidase i\n",
      "    - thermolysin\n",
      "    - thrombin\n",
      "    - trypsin_full\n",
      "    - trypsin_exception\n",
      "    - non-specific\n",
      "    - trypsin\n",
      "  pseudo_reverse:\n",
      "    default: true\n",
      "    description: Use pseudo-reverse strategy instead of reverse.\n",
      "    type: checkbox\n",
      "  save_db:\n",
      "    default: true\n",
      "    description: Save DB or create on the fly.\n",
      "    type: checkbox\n",
      "  spectra_block:\n",
      "    default: 100000\n",
      "    description: Maximum number of sequences to be collected before theoretical spectra\n",
      "      are generated.\n",
      "    max: 1000000\n",
      "    min: 1000\n",
      "    type: spinbox\n",
      "features:\n",
      "  mass_importance:\n",
      "    default: 100\n",
      "    max: 1000\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  max_distance:\n",
      "    default: 0.4\n",
      "    max: 1.0\n",
      "    min: 0.0\n",
      "    type: doublespinbox\n",
      "  max_gap:\n",
      "    default: 2\n",
      "    max: 99\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  max_neighbors:\n",
      "    default: 4\n",
      "    max: 10\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  min_hill_length:\n",
      "    default: 3\n",
      "    max: 10\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  ppm_tol:\n",
      "    default: 8\n",
      "    max: 99\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  smoothing:\n",
      "    default: 1\n",
      "    max: 10\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "general:\n",
      "  find_features:\n",
      "    default: true\n",
      "    description: Flag to perform feature finding.\n",
      "    type: checkbox\n",
      "  n_processes:\n",
      "    default: 128\n",
      "    description: Maximum number of processes for multiprocessing. If larger than number\n",
      "      of processors it will be capped.\n",
      "    max: 128\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  ppm:\n",
      "    default: true\n",
      "    description: Use ppm instead of Dalton.\n",
      "    type: checkbox\n",
      "  score:\n",
      "    default: random_forest\n",
      "    description: Scoring method.\n",
      "    type: combobox\n",
      "    value:\n",
      "    - x_tandem\n",
      "    - random_forest\n",
      "matching:\n",
      "  match_between_runs:\n",
      "    default: true\n",
      "    description: Perform match between runs type quantification.\n",
      "    type: checkbox\n",
      "  min_match_d:\n",
      "    default: 3\n",
      "    description: Minimum distance cutoff for matching\n",
      "    max: 10.0\n",
      "    min: 0.001\n",
      "    type: doublespinbox\n",
      "  min_match_p:\n",
      "    default: 0.05\n",
      "    description: Minimum probability cutoff for matching\n",
      "    max: 1.0\n",
      "    min: 0.001\n",
      "    type: doublespinbox\n",
      "quantification:\n",
      "  max_lfq:\n",
      "    default: true\n",
      "    description: Perform max lfq type quantification.\n",
      "    type: checkbox\n",
      "  mode:\n",
      "    default: int_sum\n",
      "    description: Column to perform quantification on.\n",
      "    type: combobox\n",
      "    value:\n",
      "    - int_sum\n",
      "raw:\n",
      "  most_abundant:\n",
      "    default: 400\n",
      "    description: Number of most abundant peaks to be isolated from raw spectra.\n",
      "    max: 1000\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  use_profile_ms1:\n",
      "    default: false\n",
      "    description: Use profile data for MS1 and perform own centroiding.\n",
      "    type: checkbox\n",
      "search:\n",
      "  calibrate:\n",
      "    default: true\n",
      "    description: Recalibrate masses.\n",
      "    type: checkbox\n",
      "  calibration_std:\n",
      "    default: 3\n",
      "    description: Std range for search after calibration.\n",
      "    max: 5\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  m_offset:\n",
      "    default: 10\n",
      "    description: Precursor mass offset.\n",
      "    max: 99\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  m_tol:\n",
      "    default: 20\n",
      "    description: MSMS mass offset.\n",
      "    max: 99\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  min_frag_hits:\n",
      "    default: 7\n",
      "    description: Minimum number of fragment hits.\n",
      "    max: 99\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  parallel:\n",
      "    default: true\n",
      "    description: Use parallel processing.\n",
      "    type: checkbox\n",
      "  peptide_fdr:\n",
      "    default: 0.01\n",
      "    description: FDR level for peptides.\n",
      "    max: 1.0\n",
      "    min: 0.0\n",
      "    type: doublespinbox\n",
      "  ppm:\n",
      "    default: true\n",
      "    description: Use ppm instead of Dalton.\n",
      "    type: checkbox\n",
      "  protein_fdr:\n",
      "    default: 0.01\n",
      "    description: FDR level for proteins.\n",
      "    max: 1.0\n",
      "    min: 0.0\n",
      "    type: doublespinbox\n",
      "  recalibration_min:\n",
      "    default: 100\n",
      "    description: Minimum number of datapoints to perform calibration.\n",
      "    max: 10000\n",
      "    min: 100\n",
      "    type: spinbox\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#hide\n",
    "import pandas as pd\n",
    "from alphapept.constants import protease_dict\n",
    "\n",
    "SETTINGS_TEMPLATE = {}\n",
    "\n",
    "# General\n",
    "general = {}\n",
    "SETTINGS_TEMPLATE[\"general\"] = general\n",
    "general[\"ppm\"] = {'type':'checkbox', 'default':True, 'description':\"Use ppm instead of Dalton.\"}\n",
    "general[\"find_features\"] = {'type':'checkbox', 'default':True, 'description':\"Flag to perform feature finding.\"}\n",
    "general[\"score\"] = {'type':'combobox', 'value':['x_tandem','random_forest'], 'default':'random_forest', 'description':\"Scoring method.\"}\n",
    "general['n_processes'] = {'type':'spinbox', 'min':1, 'max':128, 'default':128, 'description':\"Maximum number of processes for multiprocessing. If larger than number of processors it will be capped.\"}\n",
    "\n",
    "SETTINGS_TEMPLATE[\"general\"] = general\n",
    "\n",
    "experiment = {}\n",
    "\n",
    "experiment[\"results_path\"] = {'type':'path','default': None, 'filetype':['hdf'], 'folder':False, 'description':\"Path where the results should be stored.\"}\n",
    "experiment[\"shortnames\"] = {'type':'list','default':[], 'description':\"List of shortnames for the raw files.\"}\n",
    "experiment[\"file_paths\"] = {'type':'list','default':[], 'description':\"Filepaths of the experiments.\"}\n",
    "experiment[\"fractions\"] = {'type':'list','default':[], 'description':\"List of fractions for the raw files.\"}\n",
    "\n",
    "SETTINGS_TEMPLATE[\"experiment\"] = experiment\n",
    "\n",
    "# Raw\n",
    "raw = {}\n",
    "SETTINGS_TEMPLATE[\"raw\"] = raw\n",
    "\n",
    "raw[\"most_abundant\"] = {'type':'spinbox', 'min':1, 'max':1000, 'default':400, 'description':\"Number of most abundant peaks to be isolated from raw spectra.\"}\n",
    "raw[\"use_profile_ms1\"] = {'type':'checkbox', 'default':False, 'description':\"Use profile data for MS1 and perform own centroiding.\"}\n",
    "\n",
    "# Fasta\n",
    "fasta = {}\n",
    "SETTINGS_TEMPLATE[\"fasta\"] = fasta\n",
    "\n",
    "## Read modifications from modifications file\n",
    "mod_db = pd.read_csv('../modifications.tsv', sep='\\t')\n",
    "\n",
    "mods = {}\n",
    "mods_terminal = {}\n",
    "mods_protein = {}\n",
    "\n",
    "for i in range(len(mod_db)):\n",
    "    mod = mod_db.iloc[i]\n",
    "    if 'terminus' in mod['Type']:\n",
    "        if 'peptide' in mod['Type']:\n",
    "            mods_terminal[mod['Identifier']] = mod['Description']\n",
    "        elif 'protein' in mod['Type']:\n",
    "            mods_protein[mod['Identifier']] = mod['Description']\n",
    "        else:\n",
    "            print('Not understood')\n",
    "            print(mod['Type'])\n",
    "    else:\n",
    "        mods[mod['Identifier']] = mod['Description']\n",
    "\n",
    "fasta[\"mods_fixed\"] = {'type':'checkgroup', 'value':mods, 'default':['cC'],'description':\"Fixed modifications.\"}\n",
    "fasta[\"mods_fixed_terminal\"] = {'type':'checkgroup', 'value':mods_terminal, 'default':[],'description':\"Fixed terminal modifications.\"}\n",
    "fasta[\"mods_variable\"] = {'type':'checkgroup', 'value':mods, 'default':['oxM'],'description':\"Fixed variable modifications.\"}\n",
    "fasta[\"mods_variable_terminal\"]  = {'type':'checkgroup', 'value':mods_terminal, 'default':[], 'description':\"Fixed varibale terminal modifications.\"}\n",
    "\n",
    "fasta[\"mods_fixed_terminal_prot\"] = {'type':'checkgroup', 'value':mods_protein, 'default':[],'description':\"Fixed terminal modifications on proteins.\"}\n",
    "fasta[\"mods_variable_terminal_prot\"]  = {'type':'checkgroup', 'value':mods_protein, 'default':[], 'description':\"Fixed varibale terminal modifications  on proteins.\"}\n",
    "\n",
    "fasta[\"num_missed_cleavages\"] = {'type':'spinbox', 'min':0, 'max':99, 'default':2, 'description':\"Number of missed cleavages.\"}\n",
    "fasta[\"min_length\"] = {'type':'spinbox', 'min':6, 'max':99, 'default':6, 'description':\"Minimum peptide length.\"}\n",
    "fasta[\"max_length\"] = {'type':'spinbox', 'min':6, 'max':99, 'default':27, 'description':\"Maximum peptide length.\"}\n",
    "fasta[\"max_isoforms\"] = {'type':'spinbox', 'min':1, 'max':4096, 'default':1024, 'description':\"Maximum number of isoforms per peptide.\"}\n",
    "\n",
    "fasta[\"pseudo_reverse\"] = {'type':'checkbox', 'default':True, 'description':\"Use pseudo-reverse strategy instead of reverse.\"}\n",
    "fasta[\"AL_swap\"] = {'type':'checkbox', 'default':False, 'description':\"Swap A and L for decoy generation.\"}\n",
    "fasta[\"KR_swap\"] = {'type':'checkbox', 'default':False, 'description':\"Swap K and R (only if terminal) for decoy generation.\"}\n",
    "\n",
    "proteases = [_ for _ in protease_dict.keys()]\n",
    "fasta[\"protease\"] = {'type':'combobox', 'value':proteases, 'default':'trypsin', 'description':\"Protease for digestions.\"}\n",
    "\n",
    "fasta[\"spectra_block\"] = {'type':'spinbox', 'min':1000, 'max':1000000, 'default':100000, 'description':\"Maximum number of sequences to be collected before theoretical spectra are generated.\"}\n",
    "fasta[\"fasta_block\"] = {'type':'spinbox', 'min':100, 'max':10000, 'default':1000, 'description':\"Number of fasta entries to be processed in one block.\"}\n",
    "fasta[\"save_db\"] = {'type':'checkbox', 'default':True, 'description':\"Save DB or create on the fly.\"}\n",
    "\n",
    "fasta[\"database_path\"] = {'type':'path','default':None, 'filetype':['hdf'], 'folder':False, 'description':\"Path to library file (.hdf).\"}\n",
    "fasta[\"fasta_paths\"] = {'type':'list','default':[], 'description':\"List of paths for FASTA files.\"}\n",
    "\n",
    "\n",
    "# Search Settings\n",
    "search = {}\n",
    "\n",
    "SETTINGS_TEMPLATE[\"search\"] = search\n",
    "\n",
    "search[\"m_offset\"] = {'type':'spinbox', 'min':1, 'max':99, 'default':30, 'description':\"Precursor mass offset.\"}\n",
    "search[\"m_tol\"] = {'type':'spinbox', 'min':1, 'max':99, 'default':30, 'description':\"MSMS mass offset.\"}\n",
    "search[\"min_frag_hits\"] = {'type':'spinbox', 'min':1, 'max':99, 'default':7, 'description':\"Minimum number of fragment hits.\"}\n",
    "search[\"ppm\"] = {'type':'checkbox', 'default':True, 'description':\"Use ppm instead of Dalton.\"}\n",
    "search[\"calibrate\"] = {'type':'checkbox', 'default':True, 'description':\"Recalibrate masses.\"}\n",
    "search[\"calibration_std\"] = {'type':'spinbox', 'min':1, 'max':5, 'default':3, 'description':\"Std range for search after calibration.\"}\n",
    "search[\"parallel\"] = {'type':'checkbox', 'default':True, 'description':\"Use parallel processing.\"}\n",
    "search[\"peptide_fdr\"] = {'type':'doublespinbox', 'min':0.0, 'max':1.0, 'default':0.01, 'description':\"FDR level for peptides.\"}\n",
    "search[\"protein_fdr\"] = {'type':'doublespinbox', 'min':0.0, 'max':1.0, 'default':0.01, 'description':\"FDR level for proteins.\"}\n",
    "search['recalibration_min'] = {'type':'spinbox', 'min':100, 'max':10000, 'default':100, 'description':\"Minimum number of datapoints to perform calibration.\"}\n",
    "\n",
    "\n",
    "# Feature Settings\n",
    "features = {}\n",
    "SETTINGS_TEMPLATE[\"features\"] = features\n",
    "\n",
    "features[\"min_hill_length\"] = {'type':'spinbox', 'min':1, 'max':10, 'default':3}\n",
    "features[\"max_gap\"] = {'type':'spinbox', 'min':1, 'max':99, 'default':2}\n",
    "features[\"ppm_tol\"] = {'type':'spinbox', 'min':1, 'max':99, 'default':8}\n",
    "features[\"smoothing\"] = {'type':'spinbox', 'min':1, 'max':10, 'default':1}\n",
    "\n",
    "features[\"max_neighbors\"] = {'type':'spinbox', 'min':1, 'max':10, 'default':4}\n",
    "features[\"max_distance\"] = {'type':'doublespinbox', 'min':0.0, 'max':1.0, 'default':0.4}\n",
    "features[\"mass_importance\"] = {'type':'spinbox', 'min':1, 'max':1000, 'default':100}\n",
    "\n",
    "# Calibration\n",
    "\n",
    "calibration = {}\n",
    "SETTINGS_TEMPLATE[\"calibration\"] = calibration\n",
    "\n",
    "calibration[\"min_mz_step\"] = {'type':'spinbox', 'min':1, 'max':999, 'default':80, 'description':\"Minimum number of datapoints in for mz interpolation.\"}\n",
    "calibration[\"min_rt_step\"] = {'type':'spinbox', 'min':1, 'max':999, 'default':50, 'description':\"Minimum number of datapoints in for rt interpolation.\"}\n",
    "calibration[\"minimum_score\"] = {'type':'spinbox', 'min':1, 'max':999, 'default':20, 'description':\"Minimum score for psms used for calibration.\"}\n",
    "calibration[\"outlier_std\"] = {'type':'spinbox', 'min':1, 'max':5, 'default':3, 'description':\"Number of std. deviations to filter outliers in psms.\"}\n",
    "calibration[\"method\"] = {'type':'combobox', 'value':['linear', 'nearest', 'cubic'], 'default':'linear', 'description':\"Interpolation method.\"}\n",
    "\n",
    "\n",
    "# Matching\n",
    "\n",
    "matching = {}\n",
    "SETTINGS_TEMPLATE[\"matching\"] = matching\n",
    "\n",
    "matching['match_between_runs'] = {'type':'checkbox', 'default':True, 'description':\"Perform match between runs type quantification.\"}\n",
    "matching[\"min_match_p\"] = {'type':'doublespinbox', 'min':0.001, 'max':1.0, 'default':0.05, 'description':\"Minimum probability cutoff for matching\"}\n",
    "matching[\"min_match_d\"] = {'type':'doublespinbox', 'min':0.001, 'max':10.0, 'default':3, 'description': \"Minimum distance cutoff for matching\"}\n",
    "\n",
    "# Quantification\n",
    "\n",
    "quantification = {}\n",
    "SETTINGS_TEMPLATE[\"quantification\"] = quantification\n",
    "quantification[\"max_lfq\"] = {'type':'checkbox', 'default':True, 'description':\"Perform max lfq type quantification.\"}\n",
    "quantification[\"mode\"] = {'type':'combobox', 'value':['int_sum'], 'default':'int_sum', 'description':\"Column to perform quantification on.\"}\n",
    "\n",
    "# Save everything\n",
    "\n",
    "path = \"../alphapept/settings_template.yaml\"\n",
    "\n",
    "with open(path, \"w\") as file:\n",
    "    yaml.dump(SETTINGS_TEMPLATE, file)\n",
    "    \n",
    "save_settings(SETTINGS_TEMPLATE, path)\n",
    "print_settings(load_settings(path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "settings = {}\n",
    "\n",
    "for category in SETTINGS_TEMPLATE.keys():\n",
    "    \n",
    "    temp_settings = {}\n",
    "    \n",
    "    for key in SETTINGS_TEMPLATE[category].keys():\n",
    "        temp_settings[key] = SETTINGS_TEMPLATE[category][key]['default']\n",
    "        \n",
    "    settings[category] = temp_settings\n",
    "    \n",
    "path = \"../default_settings.yaml\"\n",
    "\n",
    "save_settings(settings, path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converted 00_settings.ipynb.\n",
      "Converted 01_chem.ipynb.\n",
      "Converted 02_io.ipynb.\n",
      "Converted 03_fasta.ipynb.\n",
      "Converted 04_feature_finding.ipynb.\n",
      "Converted 05_search.ipynb.\n",
      "Converted 06_score.ipynb.\n",
      "Converted 07_recalibration.ipynb.\n",
      "Converted 08_quantification.ipynb.\n",
      "Converted 09_matching.ipynb.\n",
      "Converted 10_constants.ipynb.\n",
      "Converted 11_interface.ipynb.\n",
      "Converted index.ipynb.\n",
      "Converted Untitled.ipynb.\n"
     ]
    }
   ],
   "source": [
    "#hide\n",
    "from nbdev.export import *\n",
    "notebook2script()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-alphapept] *",
   "language": "python",
   "name": "conda-env-.conda-alphapept-py"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
