{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# default_exp settings_template"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Settings Template\n",
    "\n",
    "> A template for settings"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ToDo:\n",
    "Make checks that settings are not contradicting.\n",
    "E.g. if a fasta file is provided check to make the library conversion.\n",
    "If path to library is provided don't do this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from nbdev.showdoc import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Settings\n",
    "\n",
    "Each entry has a type, default values and a description.\n",
    "\n",
    "* spinbox -> QSpinBox with minimum and maximum values (int)\n",
    "* doublespinbox -> QDoubleSpinBox with minimum and maximum values (float)\n",
    "* path -> Clickable button to select a path\n",
    "* combobox -> QComboBox, dropdown menu with values to choose from\n",
    "* checkbox -> QCheckBox, checkbox that can be selected\n",
    "* checkgroup -> Creates a list of QCheckBox options that can be selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "calibration:\n",
      "  method:\n",
      "    default: linear\n",
      "    description: Interpolation method.\n",
      "    type: combobox\n",
      "    value:\n",
      "    - linear\n",
      "    - nearest\n",
      "    - cubic\n",
      "  min_mz_step:\n",
      "    default: 80\n",
      "    description: Minimum number of datapoints in for mz interpolation.\n",
      "    max: 999\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  min_rt_step:\n",
      "    default: 50\n",
      "    description: Minimum number of datapoints in for rt interpolation.\n",
      "    max: 999\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  minimum_score:\n",
      "    default: 20\n",
      "    description: Minimum score for psms used for calibration.\n",
      "    max: 999\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  outlier_std:\n",
      "    default: 3\n",
      "    description: Number of std. deviations to filter outliers in psms.\n",
      "    max: 5\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "fasta:\n",
      "  fasta_folder:\n",
      "    default: '...'\n",
      "    description: Path to folder with fasta files.\n",
      "    filetype:\n",
      "    - fasta\n",
      "    folder: true\n",
      "    type: path\n",
      "  fasta_path:\n",
      "    default: '...'\n",
      "    description: Path to fasta file.\n",
      "    filetype:\n",
      "    - fasta\n",
      "    folder: false\n",
      "    type: path\n",
      "  library_path:\n",
      "    default: '...'\n",
      "    description: Path to library file (.npz).\n",
      "    filetype:\n",
      "    - npz\n",
      "    folder: false\n",
      "    type: path\n",
      "  max_isoforms:\n",
      "    default: 1024\n",
      "    description: Maximum number of isoforms per peptide.\n",
      "    max: 4096\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  max_length:\n",
      "    default: 27\n",
      "    description: Maximum peptide length.\n",
      "    max: 99\n",
      "    min: 6\n",
      "    type: spinbox\n",
      "  min_length:\n",
      "    default: 6\n",
      "    description: Minimum peptide length.\n",
      "    max: 99\n",
      "    min: 6\n",
      "    type: spinbox\n",
      "  mods_fixed:\n",
      "    default:\n",
      "    - cC\n",
      "    description: Fixed modifications.\n",
      "    type: checkgroup\n",
      "    value: &id001\n",
      "      aK: acetylation of lysine\n",
      "      cC: carbamidomethylation of C\n",
      "      deamN: deamidation of N\n",
      "      deamNQ: deamidation of Q\n",
      "      eK: EASItag 6-plex on K\n",
      "      itraq4K: iTRAQ 4-plex on K\n",
      "      itraq4Y: iTRAQ 4-plex on Y\n",
      "      itraq8K: iTRAQ 8-plex on K\n",
      "      itraq8Y: iTRAQ 8-plex on Y\n",
      "      oxM: oxidation of M\n",
      "      pS: phosphorylation of S\n",
      "      pT: phosphorylation of T\n",
      "      pY: phosphorylation of Y\n",
      "      tmt0K: TMT zero on K\n",
      "      tmt0Y: TMT zero on Y\n",
      "      tmt2K: TMT duplex on K\n",
      "      tmt2Y: TMT duplex on Y\n",
      "      tmt6K: TMT sixplex/tenplex on K\n",
      "      tmt6Y: TMT sixplex/tenplex on Y\n",
      "  mods_fixed_terminal:\n",
      "    default: []\n",
      "    description: Fixed terminal modifications.\n",
      "    type: checkgroup\n",
      "    value: &id002\n",
      "      cmC<: pyro-cmC\n",
      "      e<^: EASItag 6-plex on peptide N-terminus\n",
      "      itraq4K<^: iTRAQ 4-plex on peptide N-terminus\n",
      "      itraq8K<^: iTRAQ 8-plex on peptide N-terminus\n",
      "      pgE<: pyro-E\n",
      "      pgQ<: pyro-Q\n",
      "      tmt0<^: TMT zero on peptide N-terminus\n",
      "      tmt2<^: TMT duplex on peptide N-terminus\n",
      "      tmt6<^: TMT sixplex/tenplex on peptide N-terminus\n",
      "  mods_fixed_terminal_prot:\n",
      "    default: []\n",
      "    description: Fixed terminal modifications on proteins.\n",
      "    type: checkgroup\n",
      "    value: &id003\n",
      "      a<^: acetylation of protein N-terminus\n",
      "      am>^: amidation of protein C-terminus\n",
      "  mods_variable:\n",
      "    default:\n",
      "    - oxM\n",
      "    description: Fixed variable modifications.\n",
      "    type: checkgroup\n",
      "    value: *id001\n",
      "  mods_variable_terminal:\n",
      "    default: []\n",
      "    description: Fixed varibale terminal modifications.\n",
      "    type: checkgroup\n",
      "    value: *id002\n",
      "  mods_variable_terminal_prot:\n",
      "    default: []\n",
      "    description: Fixed varibale terminal modifications  on proteins.\n",
      "    type: checkgroup\n",
      "    value: *id003\n",
      "  num_missed_cleavages:\n",
      "    default: 2\n",
      "    description: Number of missed cleavages.\n",
      "    max: 99\n",
      "    min: 0\n",
      "    type: spinbox\n",
      "  protease:\n",
      "    default: trypsin\n",
      "    description: Protease for digestions.\n",
      "    type: combobox\n",
      "    value:\n",
      "    - arg-c\n",
      "    - asp-n\n",
      "    - bnps-skatole\n",
      "    - caspase 1\n",
      "    - caspase 2\n",
      "    - caspase 3\n",
      "    - caspase 4\n",
      "    - caspase 5\n",
      "    - caspase 6\n",
      "    - caspase 7\n",
      "    - caspase 8\n",
      "    - caspase 9\n",
      "    - caspase 10\n",
      "    - chymotrypsin high specificity\n",
      "    - chymotrypsin low specificity\n",
      "    - clostripain\n",
      "    - cnbr\n",
      "    - enterokinase\n",
      "    - factor xa\n",
      "    - formic acid\n",
      "    - glutamyl endopeptidase\n",
      "    - granzyme b\n",
      "    - hydroxylamine\n",
      "    - iodosobenzoic acid\n",
      "    - lysc\n",
      "    - ntcb\n",
      "    - pepsin ph1.3\n",
      "    - pepsin ph2.0\n",
      "    - proline endopeptidase\n",
      "    - proteinase k\n",
      "    - staphylococcal peptidase i\n",
      "    - thermolysin\n",
      "    - thrombin\n",
      "    - trypsin\n",
      "    - trypsin_exception\n",
      "    - non-specific\n",
      "    - morpheus trypsin\n",
      "features:\n",
      "  mass_importance:\n",
      "    default: 100\n",
      "    max: 1000\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  max_distance:\n",
      "    default: 0.4\n",
      "    max: 1.0\n",
      "    min: 0.0\n",
      "    type: doublespinbox\n",
      "  max_gap:\n",
      "    default: 2\n",
      "    max: 99\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  max_neighbors:\n",
      "    default: 4\n",
      "    max: 10\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  min_hill_length:\n",
      "    default: 3\n",
      "    max: 10\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  ppm_tol:\n",
      "    default: 8\n",
      "    max: 99\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  smoothing:\n",
      "    default: 1\n",
      "    max: 10\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "general:\n",
      "  convert_raw:\n",
      "    default: true\n",
      "    description: Flag to re-create the npz container from raw.\n",
      "    type: checkbox\n",
      "  create_library:\n",
      "    default: true\n",
      "    description: Flag to re-create the database from fasta.\n",
      "    type: checkbox\n",
      "  parallel:\n",
      "    default: true\n",
      "    description: Use parallel processing.\n",
      "    type: checkbox\n",
      "  ppm:\n",
      "    default: true\n",
      "    description: Use ppm instead of Dalton.\n",
      "    type: checkbox\n",
      "  settings_path:\n",
      "    default: '...'\n",
      "    description: Path to settings file (.yaml).\n",
      "    filetype:\n",
      "    - yaml\n",
      "    folder: false\n",
      "    type: path\n",
      "raw:\n",
      "  most_abundant:\n",
      "    default: 400\n",
      "    description: Number of most abundant peaks to be isolated from raw spectra.\n",
      "    max: 1000\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  raw_folder:\n",
      "    default: '...'\n",
      "    description: Path to folder with raw files.\n",
      "    filetype:\n",
      "    - raw\n",
      "    folder: true\n",
      "    type: path\n",
      "  raw_path:\n",
      "    default: '...'\n",
      "    description: Path to raw file.\n",
      "    filetype:\n",
      "    - raw\n",
      "    folder: false\n",
      "    type: path\n",
      "  raw_path_npz:\n",
      "    default: '...'\n",
      "    description: Path to npz file, converted from raw.\n",
      "    filetype:\n",
      "    - npz\n",
      "    folder: false\n",
      "    type: path\n",
      "search:\n",
      "  calibrate:\n",
      "    default: true\n",
      "    description: Recalibrate masses.\n",
      "    type: checkbox\n",
      "  calibration_std:\n",
      "    default: 3\n",
      "    description: Std range for search after calibration.\n",
      "    max: 5\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  m_offset:\n",
      "    default: 10\n",
      "    description: Precursor mass offset.\n",
      "    max: 99\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  m_tol:\n",
      "    default: 20\n",
      "    description: MSMS mass offset.\n",
      "    max: 99\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  min_frag_hits:\n",
      "    default: 7\n",
      "    description: Minimum number of fragment hits.\n",
      "    max: 99\n",
      "    min: 1\n",
      "    type: spinbox\n",
      "  parallel:\n",
      "    default: true\n",
      "    description: Use parallel processing.\n",
      "    type: checkbox\n",
      "  peptide_fdr:\n",
      "    default: 0.01\n",
      "    description: FDR level for peptides.\n",
      "    max: 1.0\n",
      "    min: 0.0\n",
      "    type: doublespinbox\n",
      "  ppm:\n",
      "    default: true\n",
      "    description: Use ppm instead of Dalton.\n",
      "    type: checkbox\n",
      "  protein_fdr:\n",
      "    default: 0.01\n",
      "    description: FDR level for proteins.\n",
      "    max: 1.0\n",
      "    min: 0.0\n",
      "    type: doublespinbox\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import yaml\n",
    "from alphapept.constants import protease_dict\n",
    "\n",
    "SETTINGS_TEMPLATE = {}\n",
    "\n",
    "# General\n",
    "general = {}\n",
    "SETTINGS_TEMPLATE[\"general\"] = general\n",
    "general[\"settings_path\"] = {'type':'path','default':'...', 'filetype':['yaml'], 'folder':False, 'description':\"Path to settings file (.yaml).\"}\n",
    "general[\"ppm\"] = {'type':'checkbox', 'default':True, 'description':\"Use ppm instead of Dalton.\"}\n",
    "general[\"parallel\"] = {'type':'checkbox', 'default':True, 'description':\"Use parallel processing.\"}\n",
    "general[\"create_library\"] = {'type':'checkbox', 'default':True, 'description':\"Flag to re-create the database from fasta.\"}\n",
    "general[\"convert_raw\"] = {'type':'checkbox', 'default':True, 'description':\"Flag to re-create the npz container from raw.\"}\n",
    "\n",
    "# Raw\n",
    "raw = {}\n",
    "SETTINGS_TEMPLATE[\"raw\"] = raw\n",
    "\n",
    "raw[\"most_abundant\"] = {'type':'spinbox', 'min':1, 'max':1000, 'default':400, 'description':\"Number of most abundant peaks to be isolated from raw spectra.\"}\n",
    "raw[\"raw_path\"] = {'type':'path','default':'...', 'filetype':['raw'], 'folder':False, 'description':\"Path to raw file.\"}\n",
    "raw[\"raw_folder\"] = {'type':'path','default':'...', 'filetype':['raw'], 'folder':True, 'description':\"Path to folder with raw files.\"}\n",
    "raw[\"raw_path_npz\"] = {'type':'path','default':'...', 'filetype':['npz'], 'folder':False, 'description':\"Path to npz file, converted from raw.\"}\n",
    "\n",
    "\n",
    "# Fasta\n",
    "fasta = {}\n",
    "SETTINGS_TEMPLATE[\"fasta\"] = fasta\n",
    "\n",
    "## Read modifications from modifications file\n",
    "mod_db = pd.read_csv('./modifications.tsv', sep='\\t')\n",
    "\n",
    "mods = {}\n",
    "mods_terminal = {}\n",
    "mods_protein = {}\n",
    "\n",
    "for i in range(len(mod_db)):\n",
    "    mod = mod_db.iloc[i]\n",
    "    if 'terminus' in mod['Type']:\n",
    "        if 'peptide' in mod['Type']:\n",
    "            mods_terminal[mod['Identifier']] = mod['Description']\n",
    "        elif 'protein' in mod['Type']:\n",
    "            mods_protein[mod['Identifier']] = mod['Description']\n",
    "        else:\n",
    "            print('Not understood')\n",
    "            print(mod['Type'])\n",
    "    else:\n",
    "        mods[mod['Identifier']] = mod['Description']\n",
    "\n",
    "fasta[\"mods_fixed\"] = {'type':'checkgroup', 'value':mods, 'default':['cC'],'description':\"Fixed modifications.\"}\n",
    "fasta[\"mods_fixed_terminal\"] = {'type':'checkgroup', 'value':mods_terminal, 'default':[],'description':\"Fixed terminal modifications.\"}\n",
    "fasta[\"mods_variable\"] = {'type':'checkgroup', 'value':mods, 'default':['oxM'],'description':\"Fixed variable modifications.\"}\n",
    "fasta[\"mods_variable_terminal\"]  = {'type':'checkgroup', 'value':mods_terminal, 'default':[], 'description':\"Fixed varibale terminal modifications.\"}\n",
    "\n",
    "fasta[\"mods_fixed_terminal_prot\"] = {'type':'checkgroup', 'value':mods_protein, 'default':[],'description':\"Fixed terminal modifications on proteins.\"}\n",
    "fasta[\"mods_variable_terminal_prot\"]  = {'type':'checkgroup', 'value':mods_protein, 'default':[], 'description':\"Fixed varibale terminal modifications  on proteins.\"}\n",
    "\n",
    "fasta[\"num_missed_cleavages\"] = {'type':'spinbox', 'min':0, 'max':99, 'default':2, 'description':\"Number of missed cleavages.\"}\n",
    "fasta[\"min_length\"] = {'type':'spinbox', 'min':6, 'max':99, 'default':6, 'description':\"Minimum peptide length.\"}\n",
    "fasta[\"max_length\"] = {'type':'spinbox', 'min':6, 'max':99, 'default':27, 'description':\"Maximum peptide length.\"}\n",
    "fasta[\"max_isoforms\"] = {'type':'spinbox', 'min':1, 'max':4096, 'default':1024, 'description':\"Maximum number of isoforms per peptide.\"}\n",
    "\n",
    "fasta[\"fasta_path\"] = {'type':'path','default':'...', 'filetype':['fasta'], 'folder':False, 'description':\"Path to fasta file.\"}\n",
    "fasta[\"fasta_folder\"] = {'type':'path','default':'...', 'filetype':['fasta'], 'folder':True, 'description':\"Path to folder with fasta files.\"}\n",
    "fasta[\"library_path\"] = {'type':'path','default':'...', 'filetype':['npz'], 'folder':False, 'description':\"Path to library file (.npz).\"}\n",
    "\n",
    "fasta[\"contaminants_path\"] = {'type':'path','default':'...', 'filetype':['fasta'], 'folder':False, 'description':\"Path to conaminants fasta file.\"}\n",
    "\n",
    "proteases = [_ for _ in protease_dict.keys()]\n",
    "fasta[\"protease\"] = {'type':'combobox', 'value':proteases, 'default':'trypsin', 'description':\"Protease for digestions.\"}\n",
    "\n",
    "# Search Settings\n",
    "search = {}\n",
    "\n",
    "SETTINGS_TEMPLATE[\"search\"] = search\n",
    "\n",
    "search[\"m_offset\"] = {'type':'spinbox', 'min':1, 'max':99, 'default':10, 'description':\"Precursor mass offset.\"}\n",
    "search[\"m_tol\"] = {'type':'spinbox', 'min':1, 'max':99, 'default':20, 'description':\"MSMS mass offset.\"}\n",
    "search[\"min_frag_hits\"] = {'type':'spinbox', 'min':1, 'max':99, 'default':7, 'description':\"Minimum number of fragment hits.\"}\n",
    "search[\"ppm\"] = {'type':'checkbox', 'default':True, 'description':\"Use ppm instead of Dalton.\"}\n",
    "search[\"calibrate\"] = {'type':'checkbox', 'default':True, 'description':\"Recalibrate masses.\"}\n",
    "search[\"calibration_std\"] = {'type':'spinbox', 'min':1, 'max':5, 'default':3, 'description':\"Std range for search after calibration.\"}\n",
    "search[\"parallel\"] = {'type':'checkbox', 'default':True, 'description':\"Use parallel processing.\"}\n",
    "search[\"peptide_fdr\"] = {'type':'doublespinbox', 'min':0.0, 'max':1.0, 'default':0.01, 'description':\"FDR level for peptides.\"}\n",
    "search[\"protein_fdr\"] = {'type':'doublespinbox', 'min':0.0, 'max':1.0, 'default':0.01, 'description':\"FDR level for proteins.\"}\n",
    "\n",
    "# Feature Settings\n",
    "features = {}\n",
    "SETTINGS_TEMPLATE[\"features\"] = features\n",
    "\n",
    "features[\"min_hill_length\"] = {'type':'spinbox', 'min':1, 'max':10, 'default':3}\n",
    "features[\"max_gap\"] = {'type':'spinbox', 'min':1, 'max':99, 'default':2}\n",
    "features[\"ppm_tol\"] = {'type':'spinbox', 'min':1, 'max':99, 'default':8}\n",
    "features[\"smoothing\"] = {'type':'spinbox', 'min':1, 'max':10, 'default':1}\n",
    "\n",
    "features[\"max_neighbors\"] = {'type':'spinbox', 'min':1, 'max':10, 'default':4}\n",
    "features[\"max_distance\"] = {'type':'doublespinbox', 'min':0.0, 'max':1.0, 'default':0.4}\n",
    "features[\"mass_importance\"] = {'type':'spinbox', 'min':1, 'max':1000, 'default':100}\n",
    "\n",
    "# Calibration\n",
    "\n",
    "calibration = {}\n",
    "SETTINGS_TEMPLATE[\"calibration\"] = calibration\n",
    "\n",
    "calibration[\"min_mz_step\"] = {'type':'spinbox', 'min':1, 'max':999, 'default':80, 'description':\"Minimum number of datapoints in for mz interpolation.\"}\n",
    "calibration[\"min_rt_step\"] = {'type':'spinbox', 'min':1, 'max':999, 'default':50, 'description':\"Minimum number of datapoints in for rt interpolation.\"}\n",
    "calibration[\"minimum_score\"] = {'type':'spinbox', 'min':1, 'max':999, 'default':20, 'description':\"Minimum score for psms used for calibration.\"}\n",
    "calibration[\"outlier_std\"] = {'type':'spinbox', 'min':1, 'max':5, 'default':3, 'description':\"Number of std. deviations to filter outliers in psms.\"}\n",
    "calibration[\"method\"] = {'type':'combobox', 'value':['linear', 'nearest', 'cubic'], 'default':'linear', 'description':\"Interpolation method.\"}\n",
    "\n",
    "# Save everything\n",
    "\n",
    "path = \"settings_template.yaml\"\n",
    "\n",
    "with open(path, \"w\") as file:\n",
    "    yaml.dump(SETTINGS_TEMPLATE, file)\n",
    "\n",
    "# for debugging\n",
    "with open(path, \"r\") as settings_file:\n",
    "    SETTINGS_LOADED = yaml.load(settings_file, Loader=yaml.FullLoader)\n",
    "    print(yaml.dump(SETTINGS_LOADED))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#hide\n",
    "from nbdev.export import *\n",
    "notebook2script()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
