{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a472eeef",
   "metadata": {},
   "outputs": [],
   "source": [
    "#| default_exp proteomics_analysis\n",
    "from nbdev import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a70d5b9",
   "metadata": {},
   "source": [
    "# AlphaPept: Proteomic Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef4eb39a",
   "metadata": {},
   "source": [
    "This tutorial will cover the basic analysis of a mixed species dataset.\n",
    "We will analyze files from Puyvelde et al.: [\"A comprehensive LFQ benchmark dataset on modern day acquisition strategies in proteomics\"](10.1038/s41597-022-01216-6).\n",
    "\n",
    "The data is available at ProteomeXchange as  [PXD028735](https://proteomecentral.proteomexchange.org/cgi/GetDataset?ID=PXD028735).\n",
    "\n",
    "## Prerequistes\n",
    "\n",
    "Download and install AlphaPept according to the Readme.\n",
    "\n",
    "## Files\n",
    "\n",
    "We will download the files from the DDA dataset for Thermo. This is the following six files:\n",
    "    \n",
    "- LFQ_Orbitrap_DDA_Condition_A_Sample_Alpha_01.raw\n",
    "- LFQ_Orbitrap_DDA_Condition_A_Sample_Alpha_02.raw\n",
    "- LFQ_Orbitrap_DDA_Condition_A_Sample_Alpha_03.raw\n",
    "- LFQ_Orbitrap_DDA_Condition_B_Sample_Alpha_01.raw\n",
    "- LFQ_Orbitrap_DDA_Condition_B_Sample_Alpha_02.raw\n",
    "- LFQ_Orbitrap_DDA_Condition_B_Sample_Alpha_03.raw\n",
    "\n",
    "## Organize Files\n",
    "\n",
    "To process the data, we will place all files in one folder like so:\n",
    "\n",
    "![](https://i.imgur.com/Nj7BCOQ.png)\n",
    "\n",
    "## FASTA files\n",
    "You can either place the `.fasta`-files directly in the folder or you can use the FASTA management in the GUI. AlphaPept can directly download FASTA files from UniProt and stores them locally in the user folder (i.e., that is `C:\\Users\\admin\\.alphapept\\fasta`. \n",
    "\n",
    "We download all four FASTA files by clicking on the respective button, waiting for the download, and refreshing the page afterward. \n",
    "![](https://i.imgur.com/m1nq6VI.png)\n",
    "\n",
    "![](https://i.imgur.com/j2vRaAI.png)\n",
    "\n",
    "## Creating a new experiment\n",
    "\n",
    "###  Files\n",
    "Next, we switch to the `New experiment`-Tab and enter the file path. The page should refresh, and all files should be listed.\n",
    "\n",
    "![](https://i.imgur.com/U0Ww4Ke.png)\n",
    "\n",
    "ℹ️ Note that you can deselect images by moving the slider to the right and deselecting `Use`.\n",
    "\n",
    "![](https://i.imgur.com/HuXv8qk.png)\n",
    "\n",
    "Next, select the FASTA files you want to use for this run. Click on the `Select FASTA files`-dropdown button. If a FASTA file was in your data folder, it would be selectable here; otherwise, it will list everything alphapept user folder.\n",
    "For the example, we will select all FASTA files.\n",
    "\n",
    "![](https://i.imgur.com/OAHEWvA.png)\n",
    "\n",
    "\n",
    "ℹ️ Note that you can hover over your selection to display the full filename.\n",
    "\n",
    "###  Settings\n",
    "By default, settings in AlphaPept are collapsed; you will not need to change them. You can adjust the Processing steps (e.g. if you only want to do feature finding and no search) in `Processing steps` and change settings in `Modify settings`).\n",
    "Suppose you would want to add a fixed modification; you would select fasta. This enables the FASTA settings, and you can use the menus to add and change the settings:\n",
    "\n",
    "![](https://i.imgur.com/qlzTOYN.png)\n",
    "\n",
    "For the tutorial, we will not change any settings. Give your experiment a meaningful name, and if no warnings appear, you can click on `Submit`. Once the job has been submitted you will be prompted to go to the Status page.\n",
    "\n",
    "![](https://i.imgur.com/vPsz5xX.png)\n",
    "\n",
    "ℹ️ AlphaPept has an internal processing queue - so you can submit multiple experiments, and they will be processed one after another. \n",
    "\n",
    "###  Status\n",
    "\n",
    "The Status page will periodically update and show you the process. In the green bar, you can see how much time has been spent on the respective job. It can take 1-2 minutes until you will see the process once a job has started.\n",
    "\n",
    "![](https://i.imgur.com/HmRnIEC.png)\n",
    "\n",
    "ℹ️ You can expand `Full log` and `Queue` to see a detailed log of your processing job or to inspect the queue of future jobs that will be processed.\n",
    "\n",
    "ℹ️ Once the experiment is done, it will show `No files to process`.\n",
    "\n",
    "\n",
    "###  Results\n",
    "Once an experiment is processed, we can do some basic data exploration within AlphaPept. For this, click on the `Results`-Tab.\n",
    "\n",
    "![](https://i.imgur.com/AwCTCDf.png)\n",
    "\n",
    "All previous experiments that were run in the GUI will be selectable here. You can also select a `*.hdf` file. Once loaded, you will see a `Run File Summary`, showing the number of MS1-Features, Peptides, and Protein Groups per run. You can hover over each column to see the filename. Additionally, you can read the `Run summary` with information about the settings and timings and the `Run log`.\n",
    "\n",
    "#### Exploring tables from Experiment\n",
    "AlphaPept provides basic plotting functions to explore results from an experiment. \n",
    "\n",
    "**Volcano Plot:**\n",
    "In `Volcano plot`, you can create a volcano plot to explore differentially expressed proteins. For this, select the groups. You will have the filenames available, and if LFQ-based processing is present can choose between LFQ-intensity or normal intensity. For the example, we will divide in Condition A and B LFQ intensities according to the filenames. If you can't read the filenames as they are too long, hover over the file and wait to see the full name.\n",
    "The plot will automatically update while selecting the groups. You can hover over each point in the volcano plot to see the protein id. Additionally, you can use `Highlight proteins` to annotate specific proteins in the plot. For the sake of demonstration, we choose `P46683` and `P36578` for highlighting. The field is interactive so you can type to search for specific proteins.\n",
    "\n",
    "![](https://i.imgur.com/DLFMzz6.png)\n",
    "\n",
    "**Correlation heatmap:**\n",
    "\n",
    "In `Correlation heatmap`, you can create a correlation heatmap based on the protein intensities. In the example case, you can nicely see how the conditions correlate.\n",
    "\n",
    "![](https://i.imgur.com/vcs9gEj.png)\n",
    "\n",
    "**Scatterplot:**\n",
    "\n",
    "`Scatterplot` shows you a scatterplot of protein intensities in one sample against another. You can choose the two filenames via the Group1 and Group2 selector. You can only select one file per group but can select the same file for each group. For the tutorial case, we will create a scatter plot between two samples of condition A (01 and 02. We can see that they are nicely correlating. The output also provides the results of an OLS Regression. In our case, we have a `R-squared` of `0.979` highlighting the high correlation (as expected) of samples of the same condition.\n",
    "\n",
    "ℹ️ Hovering over a data point will show you the respective protein. This can be useful to identify outliers.\n",
    "\n",
    "![](https://i.imgur.com/zdjhUYb.png)\n",
    "\n",
    "**PCA:**\n",
    "\n",
    "`PCA` will perform a PCA analysis of the data and show the first two components as a scatterplot. This can be very useful for finding files that cluster together and e.g., identify batch effects. For our example, we see that condition A and condition B are in close proximity of each other and form clusters, as expected.\n",
    "\n",
    "![](https://i.imgur.com/7VzsJQe.png)\n",
    "\n",
    "\n",
    "**Sequence coverage map:**\n",
    "\n",
    "`Sequence coverage map` can show you which peptides were identified for a specific protein. Click on a protein of interest, select the filename, and see which part of the protein sequence is covered by identified peptides. \n",
    "\n",
    "![](https://i.imgur.com/iASIXoo.png)\n",
    "\n",
    "**Select group:**\n",
    "\n",
    "This allows to display the tabular data, which is stored in the `*.hdf`-files. As tables can be very large, you can select the data range via the slider. Clicking `Create download link` allows you to download the table. \n",
    "\n",
    "![](https://i.imgur.com/1gEAl31.png)\n",
    "\n",
    "ℹ️ By default, you explore the `results.hdf` of an experiment. By using `Select file from experiment`, you can select individual experimental files and explore tables only from the respective run."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
