---

title: Chem


keywords: fastai
sidebar: home_sidebar

summary: "Chemistry related functions"
description: "Chemistry related functions"
nb_path: "nbs/01_chem.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbs/01_chem.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This notebook contains all chemistry-related functionality.  Here, a major part is functionality to generate isotope patterns and the averagine model. This is required for feature finding, where an isotope pattern is validated by being compared to its averagine model. We use the data structure <code>Isotopes</code> from constants to handle Isotopes. Next, we define the class <a href="/chem.html#IsotopeDistribution"><code>IsotopeDistribution</code></a> to calculate isotope distributions for a given Isotope. The core of the calculation is the function <a href="/chem.html#fast_add"><code>fast_add</code></a>, which allows the fast estimation of isotope distributions. <br>
To calculate the isotope distribution for the averagine model, we define the function <a href="/chem.html#get_average_formula"><code>get_average_formula</code></a>, which calculates the amino acid composition of the averagine molecule for the given mass.
{% include note.html content='Ideally, we would like to have constants such as isotopes as global variables to not overload functions with parameters. However, currently, numba is not able to handle typed dictionaries or jitclasses as globals. We, therefore, pass them as arguments.*' %}</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="IsotopeDistributions">IsotopeDistributions<a class="anchor-link" href="#IsotopeDistributions"> </a></h2><p>The calculation of isotope distributions is based on the algorithm introduced in the paper <em>Calculation of isotope distributions in mass spectrometry. A trivial solution for a non-trivial problem</em> by <em>Hugo Kubinyi</em>. 
A more detailed description of the algorithm can be found in the <a href="https://doi.org/10.1016/S0003-2670(00">paper</a>83059-7).</p>
<p>The implementation is a Python port of the Java version of the <a href="https://github.com/fickludd/proteomicore/blob/c5e311c21c3191fc007798391ce77066531f1932/Proteins/src/main/java/se/lth/immun/chem/IsotopeDistribution.java">proteomicore - implementation</a> from Johan Teleman</p>
<p>In brief,  the approach avoids expanding polynomial expressions by combining precalculated patterns of hypothetical atom clusters and pruning low-intensity peaks. <br>
To calculate the isotope distribution for a given isotope, we define the function <code>dict_to_dist,</code> which accepts a dictionary of amino acids and returns the isotope distribution.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h2 id="IsotopeDistribution" class="doc_header"><code>class</code> <code>IsotopeDistribution</code><a href="" class="source_link" style="float:right">[source]</a></h2><blockquote><p><code>IsotopeDistribution</code>(<strong>*<code>args</code></strong>, <strong>**<code>kwargs</code></strong>) :: <a href="/chem.html#IsotopeDistribution"><code>IsotopeDistribution</code></a></p>
</blockquote>
<p>Class to represent isotope distributions.</p>
<p>Attributes:
    m0 (int): base mass.
    dm (int): number of isotopes.
    intensities (np.ndarray): isotope intensities.</p>
<p>Metnods:
    add: dd another isotope distribution.
    copy: create a copy of the current isotope distribution.
    mult: multiply the current isotope distribution.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="fast_add" class="doc_header"><code>fast_add</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/chem.py#L96" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>fast_add</code>(<strong><code>m0</code></strong>:<code>int</code>, <strong><code>dm0</code></strong>:<code>int</code>, <strong><code>int0</code></strong>:<code>ndarray</code>, <strong><code>m1</code></strong>:<code>int</code>, <strong><code>dm1</code></strong>:<code>int</code>, <strong><code>int1</code></strong>:<code>ndarray</code>, <strong><code>prune_level</code></strong>:<code>float</code>=<em><code>1e-06</code></em>)</p>
</blockquote>
<p>Helper function to quickly add isotope patterns.</p>
<p>Args:
    m0 (int): Mass of pattern 1.
    dm0 (int): dm of pattern 1 (number of isotopes).
    int0 (float): Intensities of pattern 1.
    m1 (int): Mass of pattern 2.
    dm1 (int): dm of pattern 2 (number of isotopes).
    int1 (float): Intensities of pattern 2.
    prune_level (float): Precision threshold. Defaults to 0.000001.</p>
<p>Returns:
    int: Mass of new pattern.
    int: Number of isotopes in new pattern.
    np.ndarray: Intensity of new pattern.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="numba_bin" class="doc_header"><code>numba_bin</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/chem.py#L132" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>numba_bin</code>(<strong><code>decimal</code></strong>:<code>int</code>)</p>
</blockquote>
<p>Numba compatible function to convert a decimal number to a binary (list).</p>
<p>Args:
    decimal (int): Decimal number.</p>
<p>Returns:
    list: Number in binary.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="dict_to_dist" class="doc_header"><code>dict_to_dist</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/chem.py#L152" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>dict_to_dist</code>(<strong><code>counted_AA</code></strong>:<code>Dict</code>, <strong><code>isotopes</code></strong>:<code>Dict</code>)</p>
</blockquote>
<p>Function to convert a dictionary with counts of atoms to an isotope distribution.</p>
<p>Args:
    counted_AA (Dict): Numba-typed dict with counts of atoms.
    isotopes (Dict): Numba-typed lookup dict with isotopes.</p>
<p>Returns:
    IsotopeDistribution: The calculated sotope distribution for the chemical compound.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>One can compare the implementation to the algorithm with plotting values from the paper against the own results. Example $K_{23}I_{22}$ is Table 2 from the paper.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="o">%</span><span class="k">matplotlib</span> inline 
<span class="kn">from</span> <span class="nn">alphapept.constants</span> <span class="kn">import</span> <span class="n">isotopes</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">numba.typed</span> <span class="kn">import</span> <span class="n">Dict</span>
<span class="kn">from</span> <span class="nn">numba</span> <span class="kn">import</span> <span class="n">types</span>

<span class="c1"># Example from paper:</span>
<span class="n">mass</span> <span class="o">=</span> <span class="mf">3688.1</span>
<span class="n">abundances</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">58.83</span><span class="p">,</span><span class="mf">0.18</span><span class="p">,</span><span class="mi">100</span><span class="p">,</span><span class="mf">0.29</span><span class="p">,</span><span class="mf">81.29</span><span class="p">,</span><span class="mf">0.22</span><span class="p">,</span><span class="mf">42.05</span><span class="p">,</span><span class="mf">0.11</span><span class="p">,</span><span class="mf">15.54</span><span class="p">,</span><span class="mf">0.04</span><span class="p">,</span><span class="mf">4.36</span><span class="p">,</span><span class="mf">0.01</span><span class="p">,</span><span class="mf">0.97</span><span class="p">,</span><span class="mf">0.17</span><span class="p">,</span><span class="mf">0.03</span><span class="p">])</span><span class="o">/</span><span class="mi">100</span>
<span class="n">masses</span> <span class="o">=</span> <span class="p">[</span><span class="n">mass</span> <span class="o">+</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">abundances</span><span class="p">))]</span>

<span class="c1"># Calculation with own function</span>
<span class="n">counted_AA</span> <span class="o">=</span> <span class="n">Dict</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="n">key_type</span><span class="o">=</span><span class="n">types</span><span class="o">.</span><span class="n">unicode_type</span><span class="p">,</span> <span class="n">value_type</span><span class="o">=</span><span class="n">types</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
<span class="n">counted_AA</span><span class="p">[</span><span class="s2">&quot;K&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">23</span>
<span class="n">counted_AA</span><span class="p">[</span><span class="s2">&quot;I&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">22</span>
<span class="n">dist</span> <span class="o">=</span> <span class="n">dict_to_dist</span><span class="p">(</span><span class="n">counted_AA</span><span class="p">,</span> <span class="n">isotopes</span><span class="p">)</span>
<span class="n">masses_</span> <span class="o">=</span> <span class="p">[</span><span class="n">dist</span><span class="o">.</span><span class="n">m0</span> <span class="o">+</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">dist</span><span class="o">.</span><span class="n">intensities</span><span class="p">))]</span>

<span class="c1"># Plot </span>
<span class="n">masses</span> <span class="o">=</span> <span class="p">[</span><span class="n">mass</span> <span class="o">+</span> <span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">abundances</span><span class="p">))]</span>
<span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">masses</span><span class="p">,</span> <span class="n">abundances</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Paper&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">masses_</span><span class="p">,</span> <span class="n">dist</span><span class="o">.</span><span class="n">intensities</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s1">&#39;Implementation&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Mass&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Relative Intensity (%)&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">legend</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Averagine">Averagine<a class="anchor-link" href="#Averagine"> </a></h2><p>The averagine model is based on <a href="https://www.sciencedirect.com/science/article/pii/1044030595000178">Senko et al.</a>. We define the function <a href="/chem.html#get_average_formula"><code>get_average_formula</code></a> to calculate a dictionary with averagine masses.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_average_formula" class="doc_header"><code>get_average_formula</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/chem.py#L180" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_average_formula</code>(<strong><code>molecule_mass</code></strong>:<code>float</code>, <strong><code>averagine_aa</code></strong>:<code>Dict</code>, <strong><code>isotopes</code></strong>:<code>Dict</code>, <strong><code>sulphur</code></strong>:<code>bool</code>=<em><code>True</code></em>)</p>
</blockquote>
<p>Function to calculate the averagine formula for a molecule mass.</p>
<p>Args:
    molecule_mass (float): Input molecule mass for which the averagine model should be calculated.
    averagine_aa (Dict): Numba-typed dictionary with averagine masses.
    isotopes (Dict): Numba-typed lookup dict with isotopes.
    sulphur (bool, optional): Flag to consider sulphur. Defaults to True.</p>
<p>Raises:
    NotImplementedError: If mode w/o sulphur is selected.</p>
<p>Returns:
    Dict: Numba-typed dict with averagine composition.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">alphapept.constants</span> <span class="kn">import</span> <span class="n">averagine_aa</span><span class="p">,</span> <span class="n">averagine_avg</span>
<span class="n">molecule_mass</span> <span class="o">=</span> <span class="mi">300</span>
<span class="nb">print</span><span class="p">(</span><span class="n">get_average_formula</span><span class="p">(</span><span class="n">molecule_mass</span><span class="p">,</span> <span class="n">averagine_aa</span><span class="p">,</span> <span class="n">isotopes</span><span class="p">,</span> <span class="n">sulphur</span><span class="o">=</span><span class="kc">True</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>To directly calculate the isotope distribution of a molecule mass based on the averagine model, we define the wrapper function <a href="/chem.html#mass_to_dist"><code>mass_to_dist</code></a>.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="mass_to_dist" class="doc_header"><code>mass_to_dist</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/chem.py#L219" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>mass_to_dist</code>(<strong><code>molecule_mass</code></strong>:<code>float</code>, <strong><code>averagine_aa</code></strong>:<code>Dict</code>, <strong><code>isotopes</code></strong>:<code>Dict</code>)</p>
</blockquote>
<p>Function to calculate an isotope distribution from a molecule mass using the averagine model.</p>
<p>Args:
    molecule_mass (float, averagine_aa): input molecule mass.
    averagine_aa (Dict): Numba-typed dictionary with averagine masses.
    isotopes (Dict): Numba-typed lookup dict with isotopes.</p>
<p>Returns:
    np.ndarray: isotope masses.
    np.ndarray: isotope intensity.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">mass_to_dist</span><span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="n">averagine_aa</span><span class="p">,</span> <span class="n">isotopes</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>We can use this to plot isotope distributions for a given mass.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">def</span> <span class="nf">plot_averagine</span><span class="p">(</span><span class="n">mass</span><span class="p">,</span> <span class="n">averagine_aa</span><span class="p">,</span> <span class="n">isotopes</span><span class="p">):</span>

    <span class="n">masses</span><span class="p">,</span> <span class="n">intensity</span> <span class="o">=</span> <span class="n">mass_to_dist</span><span class="p">(</span><span class="n">mass</span><span class="p">,</span> <span class="n">averagine_aa</span><span class="p">,</span> <span class="n">isotopes</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s2">&quot;Averagine Isotopic Distribution for Mass </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">mass</span><span class="p">))</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">stem</span><span class="p">(</span><span class="n">masses</span><span class="p">,</span> <span class="n">intensity</span><span class="o">*</span><span class="mi">100</span><span class="p">,</span> <span class="n">use_line_collection</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;Relative Intensity (%)&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;Molecular Weight&quot;</span><span class="p">)</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>

<span class="n">plot_averagine</span><span class="p">(</span><span class="mi">500</span><span class="p">,</span> <span class="n">averagine_aa</span><span class="p">,</span> <span class="n">isotopes</span><span class="p">)</span>
<span class="n">plot_averagine</span><span class="p">(</span><span class="mi">1000</span><span class="p">,</span> <span class="n">averagine_aa</span><span class="p">,</span> <span class="n">isotopes</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Mass-Calculations">Mass Calculations<a class="anchor-link" href="#Mass-Calculations"> </a></h2><p><a href="/chem.html#calculate_mass"><code>calculate_mass</code></a>: This function allows to calculate the precursor mass from the monoisotopic m/z and the charge.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="calculate_mass" class="doc_header"><code>calculate_mass</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/chem.py#L246" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>calculate_mass</code>(<strong><code>mono_mz</code></strong>:<code>float</code>, <strong><code>charge</code></strong>:<code>int</code>)</p>
</blockquote>
<p>Calculate the precursor mass from mono mz and charge.</p>
<p>Args:
    mono_mz (float): mono m/z.
    charge (int): charge.</p>
<p>Returns:
    float: precursor mass.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">mz</span> <span class="o">=</span> <span class="mi">300</span>
<span class="n">charge</span> <span class="o">=</span> <span class="mi">2</span>

<span class="nb">print</span><span class="p">(</span><span class="n">calculate_mass</span><span class="p">(</span><span class="n">mz</span><span class="p">,</span> <span class="n">charge</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

</div>
 

