---

title: FASTA

keywords: fastai
sidebar: home_sidebar

summary: "Functions related to generating spectra from FASTA files"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 03_fasta.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
    {% raw %}
        
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This notebook contains all functions related to creating spectra from FASTA files. In brief, what we are doing is the following:</p>
<ol>
<li>Read a FASTA file and digest the sequences</li>
<li>For each peptide, calculate a synthetic spectrum and precursor mass</li>
<li>Save spectra</li>
</ol>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Cleaving">Cleaving<a class="anchor-link" href="#Cleaving">&#182;</a></h2><p>For cleaving, we rely on the <code>parser</code> from the <code>pyteomics</code> package and write the wrapper <a href="/alphapept/fasta#cleave_sequence"><code>cleave_sequence</code></a> to use it.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="cleave_sequence" class="doc_header"><code>cleave_sequence</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L14" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>cleave_sequence</code>(<strong><code>sequence</code></strong>=<em><code>''</code></em>, <strong><code>num_missed_cleavages</code></strong>=<em><code>0</code></em>, <strong><code>protease</code></strong>=<em><code>'trypsin'</code></em>, <strong><code>min_length</code></strong>=<em><code>6</code></em>, <strong><code>max_length</code></strong>=<em><code>65</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Cleave a peptide with a a given rule</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">protease</span> <span class="o">=</span> <span class="s2">&quot;trypsin&quot;</span>
<span class="n">num_missed_cleavages</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">min_length</span><span class="p">,</span> <span class="n">max_length</span> <span class="o">=</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">65</span>

<span class="n">cleave_sequence</span><span class="p">(</span><span class="s1">&#39;ABCDEFGHIJKLMNOPQRST&#39;</span><span class="p">,</span> <span class="n">num_missed_cleavages</span><span class="p">,</span> <span class="n">protease</span><span class="p">,</span> <span class="n">min_length</span><span class="p">,</span> <span class="n">max_length</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;ABCDEFGHIJK&#39;, &#39;LMNOPQR&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Parsing">Parsing<a class="anchor-link" href="#Parsing">&#182;</a></h2><p>Peptides are composed out of amino acids that are written in capital letters - <code>PEPTIDE</code>. to distinguish modifications, they are written in lowercase such as <code>PEPTIoxDE</code> and can be of arbitrary length. For a modified amino acid, the modification preceds the letter of the amino acid. Decoys are indicated with a underscore, hence the <a href="/alphapept/fasta#parse"><code>parse</code></a> function splits after <code>_</code>. When parsing, the peptide string is converted into a numba-compatible list so that each element can be determined with the <a href="/alphapept/constants#mass_dict"><code>mass_dict</code></a> from <code>alphapept.constants</code>.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="parse" class="doc_header"><code>parse</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L40" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>parse</code>(<strong><code>peptide</code></strong>)</p>
</blockquote>
<p>Parser to parse peptide strings</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="list_to_numba" class="doc_header"><code>list_to_numba</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L58" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>list_to_numba</code>(<strong><code>a_list</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPoxTIDE&#39;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>[P, E, P, T, I, D, E]
[P, E, P, oxT, I, D, E]
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Decoy">Decoy<a class="anchor-link" href="#Decoy">&#182;</a></h2><p>The decoy strategy that is employed is a reversal of the sequence. Additionally, we can call the functions <a href="/alphapept/fasta#swap_KR"><code>swap_KR</code></a> and and <a href="/alphapept/fasta#swap_AL"><code>swap_AL</code></a> that will swap the respective AAs. The function <a href="/alphapept/fasta#swap_KR"><code>swap_KR</code></a> will only swap terminal AAs. The swapping functions only work if the AA is not modified.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_decoy_sequence" class="doc_header"><code>get_decoy_sequence</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L67" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_decoy_sequence</code>(<strong><code>peptide</code></strong>, <strong><code>AL_swap</code></strong>=<em><code>True</code></em>, <strong><code>KR_swap</code></strong>=<em><code>True</code></em>)</p>
</blockquote>
<p>Reverses a sequence and adds the '_decoy' tag.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="swap_KR" class="doc_header"><code>swap_KR</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L87" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>swap_KR</code>(<strong><code>peptide</code></strong>)</p>
</blockquote>
<p>Swaps a terminal K or R. Note: Only if AA is not modified.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="swap_AL" class="doc_header"><code>swap_AL</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L100" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>swap_AL</code>(<strong><code>peptide</code></strong>)</p>
</blockquote>
<p>Swaps a A with L. Note: Only if AA is not modified.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_decoys" class="doc_header"><code>get_decoys</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L119" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_decoys</code>(<strong><code>peptide_list</code></strong>)</p>
</blockquote>
<p>Wrapper to get decoys for lists of peptides</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_decoy_tag" class="doc_header"><code>add_decoy_tag</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L127" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_decoy_tag</code>(<strong><code>peptides</code></strong>)</p>
</blockquote>
<p>Adds a _decoy tag to a list of peptides</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">swap_AL</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;KKKALKKK&#39;</span><span class="p">)))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">swap_KR</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;AAAKRAAA&#39;</span><span class="p">)))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>[K, K, K, L, A, K, K, K]
[A, A, A, K, R, A, A, A]
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">get_decoy_sequence</span><span class="p">(</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>EDITPEP
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">get_decoys</span><span class="p">([</span><span class="s1">&#39;ABC&#39;</span><span class="p">,</span><span class="s1">&#39;DEF&#39;</span><span class="p">,</span><span class="s1">&#39;GHI&#39;</span><span class="p">]))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>[&#39;CBA&#39;, &#39;FED&#39;, &#39;IHG&#39;]
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Modifications">Modifications<a class="anchor-link" href="#Modifications">&#182;</a></h2><p>To add modifications to the peptides we distinguish fixed and variable modifications. Additionally, we make a distinciton between whether the modification is only terminal or not.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Fixed-Modifications">Fixed Modifications<a class="anchor-link" href="#Fixed-Modifications">&#182;</a></h3><p>Fixed modifications are implemented by passing a list with modified AAs that should be replaced. As we only have one letter AAs the remainder is the modification.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_fixed_mods" class="doc_header"><code>add_fixed_mods</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L134" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_fixed_mods</code>(<strong><code>seqs</code></strong>, <strong><code>mods_fixed</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Adds fixed modifications to sequences.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">mods_fixed</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;cC&#39;</span><span class="p">,</span><span class="s1">&#39;bB&#39;</span><span class="p">]</span>
<span class="n">peptide_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ABCDEF&#39;</span><span class="p">]</span>

<span class="n">add_fixed_mods</span><span class="p">(</span><span class="n">peptide_list</span><span class="p">,</span> <span class="n">mods_fixed</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;AbBcCDEF&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Variable-Modifications">Variable Modifications<a class="anchor-link" href="#Variable-Modifications">&#182;</a></h3><p>To employ variable modifications, we use the function <a href="/alphapept/fasta#get_mod_pos"><code>get_mod_pos</code></a> that returns a list of tuples with all possible modifications when giving a dicitionary with variable modifications.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_mod_pos" class="doc_header"><code>get_mod_pos</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L146" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_mod_pos</code>(<strong><code>variable_mods_r</code></strong>, <strong><code>sequence</code></strong>)</p>
</blockquote>
<p>Returns a list with of tuples with all possibilities for modified an unmodified AAs.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">mods_variable_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;M&#39;</span><span class="p">:</span><span class="s1">&#39;oxM&#39;</span><span class="p">}</span>
<span class="n">peptide</span> <span class="o">=</span> <span class="s1">&#39;AMAMA&#39;</span>
<span class="n">get_mod_pos</span><span class="p">(</span><span class="n">mods_variable_dict</span><span class="p">,</span> <span class="n">peptide</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[(&#39;A&#39;,), (&#39;M&#39;, &#39;oxM&#39;), (&#39;A&#39;,), (&#39;M&#39;, &#39;oxM&#39;), (&#39;A&#39;,)]</pre>
</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>To now generate all isoforms, we employ the function <a href="/alphapept/fasta#get_isoforms"><code>get_isoforms</code></a> that generates all isoforms for a given peptide. As the number of isoforms can become large, we restrict it with the parameter <code>max_isoforms</code>.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_isoforms" class="doc_header"><code>get_isoforms</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L162" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_isoforms</code>(<strong><code>variable_mods_r</code></strong>, <strong><code>sequence</code></strong>, <strong><code>max_isoforms</code></strong>)</p>
</blockquote>
<p>Function to generate isoforms for a given peptide - returns a list of isoforms.
The original sequence is included in the list</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">mods_variable_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;M&#39;</span><span class="p">:</span><span class="s1">&#39;oxM&#39;</span><span class="p">}</span>
<span class="n">peptide</span> <span class="o">=</span> <span class="s1">&#39;AMAMA&#39;</span>
<span class="n">max_isoforms</span> <span class="o">=</span> <span class="mi">1024</span>
<span class="n">get_isoforms</span><span class="p">(</span><span class="n">mods_variable_dict</span><span class="p">,</span> <span class="n">peptide</span><span class="p">,</span> <span class="n">max_isoforms</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Lastly, we define the wrapper <a href="/alphapept/fasta#add_variable_mods"><code>add_variable_mods</code></a> so that the functions can be called for lists of peptides and a list of variable modifications.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_variable_mods" class="doc_header"><code>add_variable_mods</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L183" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_variable_mods</code>(<strong><code>peptide_list</code></strong>, <strong><code>mods_variable</code></strong>, <strong><code>max_isoforms</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">peptide_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AMA&#39;</span><span class="p">,</span> <span class="s1">&#39;AAC&#39;</span><span class="p">]</span>
<span class="n">mods_variable</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;oxM&#39;</span><span class="p">,</span><span class="s1">&#39;amC&#39;</span><span class="p">]</span>
<span class="n">max_isoforms</span> <span class="o">=</span> <span class="mi">1024</span>

<span class="n">add_variable_mods</span><span class="p">(</span><span class="n">peptide_list</span><span class="p">,</span> <span class="n">mods_variable</span><span class="p">,</span> <span class="n">max_isoforms</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Terminal-Modifications---Fixed">Terminal Modifications - Fixed<a class="anchor-link" href="#Terminal-Modifications---Fixed">&#182;</a></h3><p>To handle terminal modifications, we use the following convention:</p>
<ul>
<li><code>&lt;</code> for the left side (N-terminal)</li>
<li><code>&gt;</code> for the right side (C-Terminal)</li>
</ul>
<p>Additionally, if we want to have a terminal modification on any AA we indicate this <code>^</code>.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_fixed_mod_terminal" class="doc_header"><code>add_fixed_mod_terminal</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L195" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_fixed_mod_terminal</code>(<strong><code>peptides</code></strong>, <strong><code>mod</code></strong>)</p>
</blockquote>
<p>Adds fixed terminal modifications</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_fixed_mods_terminal" class="doc_header"><code>add_fixed_mods_terminal</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L213" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_fixed_mods_terminal</code>(<strong><code>peptides</code></strong>, <strong><code>mods_fixed_terminal</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Wrapper to add fixed mods on sequences and lists of mods</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">peptide</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AMAMA&#39;</span><span class="p">]</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Any n-term modified with x (x&lt;^):&#39;</span><span class="p">,</span> <span class="n">add_fixed_mods_terminal</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&lt;^&#39;</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Any c-term modified with x (x&gt;^):&#39;</span><span class="p">,</span> <span class="n">add_fixed_mods_terminal</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&gt;^&#39;</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Only A on n-term modified with x (x&lt;A):&#39;</span><span class="p">,</span> <span class="n">add_fixed_mods_terminal</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&lt;A&#39;</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Only A on c-term modified with x (x&lt;A):&#39;</span><span class="p">,</span> <span class="n">add_fixed_mods_terminal</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&gt;A&#39;</span><span class="p">]))</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Terminal-Modifications---Variable">Terminal Modifications - Variable<a class="anchor-link" href="#Terminal-Modifications---Variable">&#182;</a></h3><p>Lastly, to handle terminal variable modifications we use the function <a href="/alphapept/fasta#add_variable_mods_terminal"><code>add_variable_mods_terminal</code></a>. As the modifcation can only be at the terminal end this function only adds a peptide where the terminal end is modified.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_variable_mods_terminal" class="doc_header"><code>add_variable_mods_terminal</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L226" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_variable_mods_terminal</code>(<strong><code>peptides</code></strong>, <strong><code>mods_variable_terminal</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Function to add variable terminal modifications</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_unique_peptides" class="doc_header"><code>get_unique_peptides</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L251" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_unique_peptides</code>(<strong><code>peptides</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">peptide_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AMAMA&#39;</span><span class="p">]</span>
<span class="n">add_variable_mods_terminal</span><span class="p">(</span><span class="n">peptide_list</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&lt;^&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Generating-Peptides">Generating Peptides<a class="anchor-link" href="#Generating-Peptides">&#182;</a></h3><p>Lastly we put all the functions into a wrapper <a href="/alphapept/fasta#generate_peptides"><code>generate_peptides</code></a>. It will accept a peptide and a dictionary with settings so that we can get all modified peptides</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="generate_peptides" class="doc_header"><code>generate_peptides</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L256" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>generate_peptides</code>(<strong><code>peptide</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Wrapper to get modified peptides from a peptide</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">kwargs</span> <span class="o">=</span> <span class="p">{}</span>

<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;protease&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;trypsin&quot;</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;num_missed_cleavages&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;min_length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">6</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;max_length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">27</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_variable&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;oxM&quot;</span><span class="p">]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_variable_terminal&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_fixed&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;cC&quot;</span><span class="p">]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_fixed_terminal&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_fixed_terminal_prot&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_variable_terminal_prot&quot;</span><span class="p">]</span>  <span class="o">=</span> <span class="p">[]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;max_isoforms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1024</span>

<span class="n">generate_peptides</span><span class="p">(</span><span class="s1">&#39;PEPTIDEM&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Mass-Calculations">Mass Calculations<a class="anchor-link" href="#Mass-Calculations">&#182;</a></h2><p>Using the <a href="/alphapept/constants#mass_dict"><code>mass_dict</code></a> from <code>constants</code> and being able to parse sequences with <a href="/alphapept/fasta#parse"><code>parse</code></a> we can simply look up the masses for each modified or unmodified amino acid and add everything up.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Precursor">Precursor<a class="anchor-link" href="#Precursor">&#182;</a></h3><p>To calculate the mass of the neutral precursor we start with the mass of an $H_2O$ and add the masses of all amino acids of the sequence.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_precmass" class="doc_header"><code>get_precmass</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L291" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_precmass</code>(<strong><code>parsed_pep</code></strong>, <strong><code>mass_dict</code></strong>)</p>
</blockquote>
<p>Calculate the mass of the neutral precursor</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_precmass</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">),</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Fragments">Fragments<a class="anchor-link" href="#Fragments">&#182;</a></h3><p>Likewise, we can calculate the masses of the fragment ions. We employ two functions: <a href="/alphapept/fasta#get_fragmass"><code>get_fragmass</code></a> and <a href="/alphapept/fasta#get_frag_dict"><code>get_frag_dict</code></a>.</p>
<p><a href="/alphapept/fasta#get_fragmass"><code>get_fragmass</code></a> is a fast, <code>numba</code>-compatible function that calculates the fragmasses and returns an array indicating wheter the iontype was <code>b</code> or <code>y</code>.</p>
<p><a href="/alphapept/fasta#get_frag_dict"><code>get_frag_dict</code></a> instead is not <code>numba</code>-compatible and hence a bit slower. It returns a dictionary with the respective ion and can be used for plotting theoretical spectra.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_fragmass" class="doc_header"><code>get_fragmass</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L304" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_fragmass</code>(<strong><code>parsed_pep</code></strong>, <strong><code>mass_dict</code></strong>)</p>
</blockquote>
<p>Calculate the masses of the fragment ions</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_fragmass</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">),</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_frag_dict" class="doc_header"><code>get_frag_dict</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L335" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_frag_dict</code>(<strong><code>parsed_pep</code></strong>, <strong><code>mass_dict</code></strong>)</p>
</blockquote>
<p>Calculate the masses of the fragment ions</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_frag_dict</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPT&#39;</span><span class="p">),</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">peptide</span> <span class="o">=</span> <span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">)</span>

<span class="o">%</span><span class="k">timeit</span> get_frag_dict(peptide, constants.mass_dict)
<span class="o">%</span><span class="k">timeit</span> get_fragmass(peptide, constants.mass_dict)
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This allows us also to generate the theorteical isotopes for a fragment:</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">peptide</span> <span class="o">=</span> <span class="s1">&#39;PEPTIDE&#39;</span>

<span class="n">frag_dict</span> <span class="o">=</span> <span class="n">get_frag_dict</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="n">peptide</span><span class="p">),</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">)</span>

<span class="n">db_frag</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">frag_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
<span class="n">db_int</span> <span class="o">=</span> <span class="p">[</span><span class="mi">100</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">db_frag</span><span class="p">]</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">vlines</span><span class="p">(</span><span class="n">db_frag</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">db_int</span><span class="p">,</span> <span class="s2">&quot;k&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;DB&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>

<span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">frag_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">frag_dict</span><span class="p">[</span><span class="n">_</span><span class="p">],</span> <span class="mi">104</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.8</span><span class="p">)</span>
    
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Theoretical Spectrum for </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">peptide</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Mass&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Intensity&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">110</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Spectra">Spectra<a class="anchor-link" href="#Spectra">&#182;</a></h3><p>The function <code>get_spectrum</code> returns a tuple with the following content:</p>
<ul>
<li>precursor mass</li>
<li>peptide sequence</li>
<li>fragmasses</li>
<li>fragtypes</li>
</ul>
<p>Likewise, <code>get_spectra</code> returns a list of tuples. We employ a list of tuples here as this way we can sort them easily by precursor mass.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nd">@njit</span>
<span class="k">def</span> <span class="nf">get_spectrum</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="n">mass_dict</span><span class="p">):</span>
    <span class="n">parsed_peptide</span> <span class="o">=</span> <span class="n">parse</span><span class="p">(</span><span class="n">peptide</span><span class="p">)</span>

    <span class="n">fragmasses</span><span class="p">,</span> <span class="n">fragtypes</span> <span class="o">=</span> <span class="n">get_fragmass</span><span class="p">(</span><span class="n">parsed_peptide</span><span class="p">,</span> <span class="n">mass_dict</span><span class="p">)</span>
    <span class="n">sortindex</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">fragmasses</span><span class="p">)</span>
    <span class="n">fragmasses</span> <span class="o">=</span> <span class="n">fragmasses</span><span class="p">[</span><span class="n">sortindex</span><span class="p">]</span>
    <span class="n">fragtypes</span> <span class="o">=</span> <span class="n">fragtypes</span><span class="p">[</span><span class="n">sortindex</span><span class="p">]</span>

    <span class="n">precmass</span> <span class="o">=</span> <span class="n">get_precmass</span><span class="p">(</span><span class="n">parsed_peptide</span><span class="p">,</span> <span class="n">mass_dict</span><span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">precmass</span><span class="p">,</span> <span class="n">peptide</span><span class="p">,</span> <span class="n">fragmasses</span><span class="p">,</span> <span class="n">fragtypes</span><span class="p">)</span>


<span class="nd">@njit</span>
<span class="k">def</span> <span class="nf">get_spectra</span><span class="p">(</span><span class="n">peptides</span><span class="p">,</span> <span class="n">mass_dict</span><span class="p">):</span>
    <span class="c1"># Numba function for parallel calculation of spectra</span>
    <span class="c1"># prange does seem to make problems here..</span>
    <span class="n">spectra</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">peptides</span><span class="p">)):</span>
        <span class="n">spectra</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">get_spectrum</span><span class="p">(</span><span class="n">peptides</span><span class="p">[</span><span class="n">i</span><span class="p">],</span> <span class="n">mass_dict</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">spectra</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">get_spectra</span><span class="p">([</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">],</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">def</span> <span class="nf">test_get_spectra</span><span class="p">():</span>
    
    <span class="n">spectra</span> <span class="o">=</span> <span class="n">get_spectra</span><span class="p">([</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">],</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">)</span>
    
    <span class="n">precmass</span><span class="p">,</span> <span class="n">peptide</span><span class="p">,</span> <span class="n">frags</span><span class="p">,</span> <span class="n">fragtypes</span> <span class="o">=</span> <span class="n">spectra</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    
    <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">precmass</span><span class="p">,</span> <span class="mf">799.3599642034599</span><span class="p">)</span>
    
    <span class="k">assert</span> <span class="n">peptide</span> <span class="o">==</span> <span class="s1">&#39;PEPTIDE&#39;</span>
    
    <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">frags</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span> <span class="mf">98.06004033</span><span class="p">,</span> <span class="mf">148.06043425</span><span class="p">,</span> <span class="mf">227.10263343</span><span class="p">,</span> <span class="mf">263.08737735</span><span class="p">,</span>
       <span class="mf">324.15539729</span><span class="p">,</span> <span class="mf">376.17144135</span><span class="p">,</span> <span class="mf">425.20307579</span><span class="p">,</span> <span class="mf">477.21911985</span><span class="p">,</span>
       <span class="mf">538.28713979</span><span class="p">,</span> <span class="mf">574.27188371</span><span class="p">,</span> <span class="mf">653.31408289</span><span class="p">,</span> <span class="mf">703.31447681</span><span class="p">]))</span>

<span class="n">test_get_spectra</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Reading-FASTA">Reading FASTA<a class="anchor-link" href="#Reading-FASTA">&#182;</a></h2><p>To read FASTA files we use the <code>SeqIO</code> module from the <code>Biopython</code> library.
Generator expression so that we can read it one by one</p>
<p>Additionally we define the wrapper <a href="/alphapept/fasta#read_fasta"><code>read_fasta</code></a> that allows to read multiple FASTA files.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="read_fasta_file" class="doc_header"><code>read_fasta_file</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L368" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>read_fasta_file</code>(<strong><code>fasta_filename</code></strong>=<em><code>''</code></em>)</p>
</blockquote>
<p>given a fasta_file read fasta file line by line, return progress</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="read_fasta_file_entries" class="doc_header"><code>read_fasta_file_entries</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L395" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>read_fasta_file_entries</code>(<strong><code>fasta_filename</code></strong>=<em><code>''</code></em>)</p>
</blockquote>
<p>Function to count entries in fasta file</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="read_fasta" class="doc_header"><code>read_fasta</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L413" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>read_fasta</code>(<strong><code>path</code></strong>)</p>
</blockquote>
<p>Wrapper to read multiple files.</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="check_sequence" class="doc_header"><code>check_sequence</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L429" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>check_sequence</code>(<strong><code>element</code></strong>, <strong><code>AAs</code></strong>)</p>
</blockquote>
<p>Checks wheter a sequence from a FASTA entry contains valid AAs</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#load example fasta file</span>

<span class="n">fasta_path</span> <span class="o">=</span> <span class="s1">&#39;./testfiles/test.fasta&#39;</span>

<span class="nb">list</span><span class="p">(</span><span class="n">read_fasta</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Peptide-Dictionary">Peptide Dictionary<a class="anchor-link" href="#Peptide-Dictionary">&#182;</a></h2><p>In order to efficiently store peptides we rely on the python in dictionary. The idea is to have a dictionary with peptides as keys and indicies to proteins as values. This way we can quickly look up to which protein a peptide belongs to. The function <a href="/alphapept/fasta#add_to_pept_dict"><code>add_to_pept_dict</code></a> uses a regular python dictionary and allows to add peptides and stores indicies to the originating proteins as a list. If a peptide is already present in the dictionary the list is appended. The function returns a list of <code>added_peptides</code> which were not present in the dictionary yet.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_to_pept_dict" class="doc_header"><code>add_to_pept_dict</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L440" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_to_pept_dict</code>(<strong><code>pept_dict</code></strong>, <strong><code>new_peptides</code></strong>, <strong><code>i</code></strong>)</p>
</blockquote>
<p>Add peptides to the peptide dictionary</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">pept_dict</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">new_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ABC&#39;</span><span class="p">,</span><span class="s1">&#39;DEF&#39;</span><span class="p">]</span>

<span class="n">pept_dict</span><span class="p">,</span> <span class="n">added_peptides</span> <span class="o">=</span> <span class="n">add_to_pept_dict</span><span class="p">(</span><span class="n">pept_dict</span><span class="p">,</span> <span class="n">new_peptides</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

<span class="n">new_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;DEF&#39;</span><span class="p">,</span><span class="s1">&#39;GHI&#39;</span><span class="p">]</span>

<span class="n">pept_dict</span><span class="p">,</span> <span class="n">added_peptides</span> <span class="o">=</span> <span class="n">add_to_pept_dict</span><span class="p">(</span><span class="n">pept_dict</span><span class="p">,</span> <span class="n">new_peptides</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">pept_dict</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Generating-a-library">Generating a library<a class="anchor-link" href="#Generating-a-library">&#182;</a></h2><p>To wrap everything up, we employ two functions <a href="/alphapept/fasta#generate_library"><code>generate_library</code></a> and <a href="/alphapept/fasta#generate_spectra"><code>generate_spectra</code></a>. The first one reads a FASTA file and generates a list of peptides, as well as the peptide dictionary and an ordered FASTA dictionary to be able to look up the protein indices laster. For the <code>callback</code> we first read the whole FASTA file to determine the total number of entries in the FASTA file.  For a typical FASTA file of 30 Mb with 40k entries, this should take less than a second. The progress of the digestion is monitored by processing the FASTA file one by one.
The function <a href="/alphapept/fasta#generate_spectra"><code>generate_spectra</code></a> then calculates precursor masses and fragment ions. Here, we split the total_number of sequences in <code>1000</code> steps to be able to track progress with the <code>callback</code>.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="generate_library" class="doc_header"><code>generate_library</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L457" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>generate_library</code>(<strong><code>mass_dict</code></strong>, <strong><code>fasta_path</code></strong>, <strong><code>callback</code></strong>=<em><code>None</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Function to generate a library from a fasta file</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="generate_spectra" class="doc_header"><code>generate_spectra</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L487" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>generate_spectra</code>(<strong><code>to_add</code></strong>, <strong><code>mass_dict</code></strong>, <strong><code>callback</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Function to generate a library from a fasta file</p>

</div>

</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Saving">Saving<a class="anchor-link" href="#Saving">&#182;</a></h2><p>To save the generated spectra, we rely on NumPy's NPZ format. For this we create a dictionary and save all the generated elements. The container will contain the following elements:</p>
<ul>
<li><code>precursors</code>: An array containing the precursor masses</li>
<li><code>seqs</code>: An array containing the peptide sequences for the precursor masses</li>
<li><code>pept_dict</code>: A peptide dictionary to look up the peptides and return their FASTA index</li>
<li><code>fasta_dict</code>: A fasta dictionary to look up the FASTA entry based on a pept_dict index</li>
<li><code>fragmasses</code>: An array containing the fragment masses. Unoccupied cells are filled with -1</li>
<li><code>fragtypes:</code>: An array containg the fragment types. 0 equals b-ions and 1 equals y-ions. Unoccupied cells are filled with -1</li>
<li><code>bounds</code>: An integer array containing the upper bounds for the fragment masses / types array. This is needed to quickly slice the data.</li>
</ul>
<p>All arrays are sorted according to the precursor mass.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">

</div>
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="save_library" class="doc_header"><code>save_library</code><a href="__main__.py#L5" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>save_library</code>(<strong><code>spectra</code></strong>, <strong><code>pept_dict</code></strong>, <strong><code>fasta_dict</code></strong>, <strong><code>db_path</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Function to save a library to the *.npz format.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}
</div>
 

