---

title: FASTA

keywords: fastai
sidebar: home_sidebar

summary: "Functions related to generating spectra from FASTA files"
description: "Functions related to generating spectra from FASTA files"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbs\03_fasta.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This notebook contains all functions related to creating spectra from FASTA files. In brief, what we are doing is the following:</p>
<ol>
<li>Read a FASTA file and digest the sequences</li>
<li>For each peptide, calculate a synthetic spectrum and precursor mass</li>
<li>Save spectra</li>
</ol>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Cleaving">Cleaving<a class="anchor-link" href="#Cleaving"> </a></h2><p>For cleaving, we use regular expressions to find potential cleavages sites and write the wrapper <a href="/alphapept/fasta#cleave_sequence"><code>cleave_sequence</code></a> to use it.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_missed_cleavages" class="doc_header"><code>get_missed_cleavages</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L14" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_missed_cleavages</code>(<strong><code>sequences</code></strong>, <strong><code>n_missed_cleavages</code></strong>)</p>
</blockquote>
<p>Combine cleaved sequences to get sequences with missed cleavages</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="cleave_sequence" class="doc_header"><code>cleave_sequence</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L25" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>cleave_sequence</code>(<strong><code>sequence</code></strong>=<em><code>''</code></em>, <strong><code>num_missed_cleavages</code></strong>=<em><code>0</code></em>, <strong><code>protease</code></strong>=<em><code>'trypsin'</code></em>, <strong><code>min_length</code></strong>=<em><code>6</code></em>, <strong><code>max_length</code></strong>=<em><code>65</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">protease</span> <span class="o">=</span> <span class="s2">&quot;trypsin&quot;</span>
<span class="n">num_missed_cleavages</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">min_length</span><span class="p">,</span> <span class="n">max_length</span> <span class="o">=</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">65</span>

<span class="n">cleave_sequence</span><span class="p">(</span><span class="s1">&#39;ABCDEFGHIJKLMNOPQRST&#39;</span><span class="p">,</span> <span class="n">num_missed_cleavages</span><span class="p">,</span> <span class="n">protease</span><span class="p">,</span> <span class="n">min_length</span><span class="p">,</span> <span class="n">max_length</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;ABCDEFGHIJK&#39;, &#39;LMNOPQR&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Parsing">Parsing<a class="anchor-link" href="#Parsing"> </a></h2><p>Peptides are composed out of amino acids that are written in capital letters - <code>PEPTIDE</code>. to distinguish modifications, they are written in lowercase such as <code>PEPTIoxDE</code> and can be of arbitrary length. For a modified amino acid, the modification preceds the letter of the amino acid. Decoys are indicated with a underscore, hence the <a href="/alphapept/fasta#parse"><code>parse</code></a> function splits after <code>_</code>. When parsing, the peptide string is converted into a numba-compatible list so that each element can be determined with the <a href="/alphapept/constants#mass_dict"><code>mass_dict</code></a> from <code>alphapept.constants</code>.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="parse" class="doc_header"><code>parse</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L58" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>parse</code>(<strong><code>peptide</code></strong>)</p>
</blockquote>
<p>Parser to parse peptide strings</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="list_to_numba" class="doc_header"><code>list_to_numba</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L76" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>list_to_numba</code>(<strong><code>a_list</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPoxTIDE&#39;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>[P, E, P, T, I, D, E]
[P, E, P, oxT, I, D, E]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Decoy">Decoy<a class="anchor-link" href="#Decoy"> </a></h2><p>The decoy strategy that is employed is a reversal of the sequence. Additionally, we can call the functions <a href="/alphapept/fasta#swap_KR"><code>swap_KR</code></a> and and <a href="/alphapept/fasta#swap_AL"><code>swap_AL</code></a> that will swap the respective AAs. The function <a href="/alphapept/fasta#swap_KR"><code>swap_KR</code></a> will only swap terminal AAs. The swapping functions only work if the AA is not modified.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_decoy_sequence" class="doc_header"><code>get_decoy_sequence</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L85" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_decoy_sequence</code>(<strong><code>peptide</code></strong>, <strong><code>AL_swap</code></strong>=<em><code>True</code></em>, <strong><code>KR_swap</code></strong>=<em><code>True</code></em>)</p>
</blockquote>
<p>Reverses a sequence and adds the '_decoy' tag.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="swap_KR" class="doc_header"><code>swap_KR</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L105" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>swap_KR</code>(<strong><code>peptide</code></strong>)</p>
</blockquote>
<p>Swaps a terminal K or R. Note: Only if AA is not modified.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="swap_AL" class="doc_header"><code>swap_AL</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L118" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>swap_AL</code>(<strong><code>peptide</code></strong>)</p>
</blockquote>
<p>Swaps a A with L. Note: Only if AA is not modified.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_decoys" class="doc_header"><code>get_decoys</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L137" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_decoys</code>(<strong><code>peptide_list</code></strong>)</p>
</blockquote>
<p>Wrapper to get decoys for lists of peptides</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_decoy_tag" class="doc_header"><code>add_decoy_tag</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L145" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_decoy_tag</code>(<strong><code>peptides</code></strong>)</p>
</blockquote>
<p>Adds a _decoy tag to a list of peptides</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">swap_AL</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;KKKALKKK&#39;</span><span class="p">)))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">swap_KR</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;AAAKRAAA&#39;</span><span class="p">)))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>[K, K, K, L, A, K, K, K]
[A, A, A, K, R, A, A, A]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">get_decoy_sequence</span><span class="p">(</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>EDITPEP
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">get_decoys</span><span class="p">([</span><span class="s1">&#39;ABC&#39;</span><span class="p">,</span><span class="s1">&#39;DEF&#39;</span><span class="p">,</span><span class="s1">&#39;GHI&#39;</span><span class="p">]))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>[&#39;CBA&#39;, &#39;FED&#39;, &#39;IHG&#39;]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Modifications">Modifications<a class="anchor-link" href="#Modifications"> </a></h2><p>To add modifications to the peptides we distinguish fixed and variable modifications. Additionally, we make a distinciton between whether the modification is only terminal or not.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Fixed-Modifications">Fixed Modifications<a class="anchor-link" href="#Fixed-Modifications"> </a></h3><p>Fixed modifications are implemented by passing a list with modified AAs that should be replaced. As we only have one letter AAs the remainder is the modification.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_fixed_mods" class="doc_header"><code>add_fixed_mods</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L152" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_fixed_mods</code>(<strong><code>seqs</code></strong>, <strong><code>mods_fixed</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Adds fixed modifications to sequences.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">mods_fixed</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;cC&#39;</span><span class="p">,</span><span class="s1">&#39;bB&#39;</span><span class="p">]</span>
<span class="n">peptide_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ABCDEF&#39;</span><span class="p">]</span>

<span class="n">add_fixed_mods</span><span class="p">(</span><span class="n">peptide_list</span><span class="p">,</span> <span class="n">mods_fixed</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;AbBcCDEF&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Variable-Modifications">Variable Modifications<a class="anchor-link" href="#Variable-Modifications"> </a></h3><p>To employ variable modifications, we use the function <a href="/alphapept/fasta#get_mod_pos"><code>get_mod_pos</code></a> that returns a list of tuples with all possible modifications when giving a dicitionary with variable modifications.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_mod_pos" class="doc_header"><code>get_mod_pos</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L164" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_mod_pos</code>(<strong><code>variable_mods_r</code></strong>, <strong><code>sequence</code></strong>)</p>
</blockquote>
<p>Returns a list with of tuples with all possibilities for modified an unmodified AAs.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">mods_variable_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;M&#39;</span><span class="p">:</span><span class="s1">&#39;oxM&#39;</span><span class="p">}</span>
<span class="n">peptide</span> <span class="o">=</span> <span class="s1">&#39;AMAMA&#39;</span>
<span class="n">get_mod_pos</span><span class="p">(</span><span class="n">mods_variable_dict</span><span class="p">,</span> <span class="n">peptide</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[(&#39;A&#39;,), (&#39;M&#39;, &#39;oxM&#39;), (&#39;A&#39;,), (&#39;M&#39;, &#39;oxM&#39;), (&#39;A&#39;,)]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>To now generate all isoforms, we employ the function <a href="/alphapept/fasta#get_isoforms"><code>get_isoforms</code></a> that generates all isoforms for a given peptide. As the number of isoforms can become large, we restrict it with the parameter <code>max_isoforms</code>.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_isoforms" class="doc_header"><code>get_isoforms</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L180" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_isoforms</code>(<strong><code>variable_mods_r</code></strong>, <strong><code>sequence</code></strong>, <strong><code>max_isoforms</code></strong>)</p>
</blockquote>
<p>Function to generate isoforms for a given peptide - returns a list of isoforms.
The original sequence is included in the list</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">mods_variable_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;M&#39;</span><span class="p">:</span><span class="s1">&#39;oxM&#39;</span><span class="p">}</span>
<span class="n">peptide</span> <span class="o">=</span> <span class="s1">&#39;AMAMA&#39;</span>
<span class="n">max_isoforms</span> <span class="o">=</span> <span class="mi">1024</span>
<span class="n">get_isoforms</span><span class="p">(</span><span class="n">mods_variable_dict</span><span class="p">,</span> <span class="n">peptide</span><span class="p">,</span> <span class="n">max_isoforms</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;AMAMA&#39;, &#39;AMAoxMA&#39;, &#39;AoxMAMA&#39;, &#39;AoxMAoxMA&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Lastly, we define the wrapper <a href="/alphapept/fasta#add_variable_mods"><code>add_variable_mods</code></a> so that the functions can be called for lists of peptides and a list of variable modifications.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_variable_mods" class="doc_header"><code>add_variable_mods</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L201" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_variable_mods</code>(<strong><code>peptide_list</code></strong>, <strong><code>mods_variable</code></strong>, <strong><code>max_isoforms</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">peptide_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AMA&#39;</span><span class="p">,</span> <span class="s1">&#39;AAC&#39;</span><span class="p">]</span>
<span class="n">mods_variable</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;oxM&#39;</span><span class="p">,</span><span class="s1">&#39;amC&#39;</span><span class="p">]</span>
<span class="n">max_isoforms</span> <span class="o">=</span> <span class="mi">1024</span>

<span class="n">add_variable_mods</span><span class="p">(</span><span class="n">peptide_list</span><span class="p">,</span> <span class="n">mods_variable</span><span class="p">,</span> <span class="n">max_isoforms</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;AMA&#39;, &#39;AoxMA&#39;, &#39;AAC&#39;, &#39;AAamC&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Terminal-Modifications---Fixed">Terminal Modifications - Fixed<a class="anchor-link" href="#Terminal-Modifications---Fixed"> </a></h3><p>To handle terminal modifications, we use the following convention:</p>
<ul>
<li><code>&lt;</code> for the left side (N-terminal)</li>
<li><code>&gt;</code> for the right side (C-Terminal)</li>
</ul>
<p>Additionally, if we want to have a terminal modification on any AA we indicate this <code>^</code>.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_fixed_mod_terminal" class="doc_header"><code>add_fixed_mod_terminal</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L213" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_fixed_mod_terminal</code>(<strong><code>peptides</code></strong>, <strong><code>mod</code></strong>)</p>
</blockquote>
<p>Adds fixed terminal modifications</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_fixed_mods_terminal" class="doc_header"><code>add_fixed_mods_terminal</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L231" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_fixed_mods_terminal</code>(<strong><code>peptides</code></strong>, <strong><code>mods_fixed_terminal</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Wrapper to add fixed mods on sequences and lists of mods</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">peptide</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AMAMA&#39;</span><span class="p">]</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Any n-term modified with x (x&lt;^):&#39;</span><span class="p">,</span> <span class="n">add_fixed_mods_terminal</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&lt;^&#39;</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Any c-term modified with x (x&gt;^):&#39;</span><span class="p">,</span> <span class="n">add_fixed_mods_terminal</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&gt;^&#39;</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Only A on n-term modified with x (x&lt;A):&#39;</span><span class="p">,</span> <span class="n">add_fixed_mods_terminal</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&lt;A&#39;</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Only A on c-term modified with x (x&lt;A):&#39;</span><span class="p">,</span> <span class="n">add_fixed_mods_terminal</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&gt;A&#39;</span><span class="p">]))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Any n-term modified with x (x&lt;^): [&#39;xAMAMA&#39;]
Any c-term modified with x (x&gt;^): [&#39;AMAMxA&#39;]
Only A on n-term modified with x (x&lt;A): [&#39;xAMAMA&#39;]
Only A on c-term modified with x (x&lt;A): [&#39;AMAMxA&#39;]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Terminal-Modifications---Variable">Terminal Modifications - Variable<a class="anchor-link" href="#Terminal-Modifications---Variable"> </a></h3><p>Lastly, to handle terminal variable modifications we use the function <a href="/alphapept/fasta#add_variable_mods_terminal"><code>add_variable_mods_terminal</code></a>. As the modifcation can only be at the terminal end this function only adds a peptide where the terminal end is modified.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_variable_mods_terminal" class="doc_header"><code>add_variable_mods_terminal</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L244" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_variable_mods_terminal</code>(<strong><code>peptides</code></strong>, <strong><code>mods_variable_terminal</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Function to add variable terminal modifications</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_unique_peptides" class="doc_header"><code>get_unique_peptides</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L269" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_unique_peptides</code>(<strong><code>peptides</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">peptide_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AMAMA&#39;</span><span class="p">]</span>
<span class="n">add_variable_mods_terminal</span><span class="p">(</span><span class="n">peptide_list</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&lt;^&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;AMAMA&#39;, &#39;xAMAMA&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Generating-Peptides">Generating Peptides<a class="anchor-link" href="#Generating-Peptides"> </a></h3><p>Lastly we put all the functions into a wrapper <a href="/alphapept/fasta#generate_peptides"><code>generate_peptides</code></a>. It will accept a peptide and a dictionary with settings so that we can get all modified peptides</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="generate_peptides" class="doc_header"><code>generate_peptides</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L273" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>generate_peptides</code>(<strong><code>peptide</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Wrapper to get modified peptides from a peptide</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">kwargs</span> <span class="o">=</span> <span class="p">{}</span>

<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;protease&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;trypsin&quot;</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;num_missed_cleavages&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;min_length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">6</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;max_length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">27</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_variable&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;oxM&quot;</span><span class="p">]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_variable_terminal&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_fixed&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;cC&quot;</span><span class="p">]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_fixed_terminal&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_fixed_terminal_prot&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_variable_terminal_prot&quot;</span><span class="p">]</span>  <span class="o">=</span> <span class="p">[]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;max_isoforms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1024</span>

<span class="n">generate_peptides</span><span class="p">(</span><span class="s1">&#39;PEPTIDEM&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;PEPTIDEM&#39;, &#39;PEPTIDEoxM&#39;, &#39;MEDITPEP_decoy&#39;, &#39;oxMEDITPEP_decoy&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">cleave_sequence</span><span class="p">(</span><span class="s1">&#39;PEPTIDEM&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;PEPTIDEM&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Mass-Calculations">Mass Calculations<a class="anchor-link" href="#Mass-Calculations"> </a></h2><p>Using the <a href="/alphapept/constants#mass_dict"><code>mass_dict</code></a> from <a href="/alphapept/constants"><code>constants</code></a> and being able to parse sequences with <a href="/alphapept/fasta#parse"><code>parse</code></a> we can simply look up the masses for each modified or unmodified amino acid and add everything up.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Precursor">Precursor<a class="anchor-link" href="#Precursor"> </a></h3><p>To calculate the mass of the neutral precursor we start with the mass of an $H_2O$ and add the masses of all amino acids of the sequence.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_precmass" class="doc_header"><code>get_precmass</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L308" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_precmass</code>(<strong><code>parsed_pep</code></strong>, <strong><code>mass_dict</code></strong>)</p>
</blockquote>
<p>Calculate the mass of the neutral precursor</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_precmass</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">),</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>799.3599642034599</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Fragments">Fragments<a class="anchor-link" href="#Fragments"> </a></h3><p>Likewise, we can calculate the masses of the fragment ions. We employ two functions: <a href="/alphapept/fasta#get_fragmass"><code>get_fragmass</code></a> and <a href="/alphapept/fasta#get_frag_dict"><code>get_frag_dict</code></a>.</p>
<p><a href="/alphapept/fasta#get_fragmass"><code>get_fragmass</code></a> is a fast, <code>numba</code>-compatible function that calculates the fragmasses and returns an array indicating wheter the iontype was <code>b</code> or <code>y</code>.</p>
<p><a href="/alphapept/fasta#get_frag_dict"><code>get_frag_dict</code></a> instead is not <code>numba</code>-compatible and hence a bit slower. It returns a dictionary with the respective ion and can be used for plotting theoretical spectra.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_fragmass" class="doc_header"><code>get_fragmass</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L321" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_fragmass</code>(<strong><code>parsed_pep</code></strong>, <strong><code>mass_dict</code></strong>)</p>
</blockquote>
<p>Calculate the masses of the fragment ions</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_fragmass</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">),</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>(array([ 98.06004033, 227.10263343, 324.15539729, 425.20307579,
        538.28713979, 653.31408289, 148.06043425, 263.08737735,
        376.17144135, 477.21911985, 574.27188371, 703.31447681]),
 array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1], dtype=int8))</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_frag_dict" class="doc_header"><code>get_frag_dict</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L351" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_frag_dict</code>(<strong><code>parsed_pep</code></strong>, <strong><code>mass_dict</code></strong>)</p>
</blockquote>
<p>Calculate the masses of the fragment ions</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_frag_dict</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPT&#39;</span><span class="p">),</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;b1&#39;: 98.06004032687,
 &#39;b2&#39;: 227.10263342686997,
 &#39;b3&#39;: 324.15539728686997,
 &#39;y1&#39;: 120.06551965033,
 &#39;y2&#39;: 217.11828351033,
 &#39;y3&#39;: 346.16087661033}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This allows us also to generate the theorteical isotopes for a fragment:</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">peptide</span> <span class="o">=</span> <span class="s1">&#39;PEPTIDE&#39;</span>

<span class="n">frag_dict</span> <span class="o">=</span> <span class="n">get_frag_dict</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="n">peptide</span><span class="p">),</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">)</span>

<span class="n">db_frag</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">frag_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
<span class="n">db_int</span> <span class="o">=</span> <span class="p">[</span><span class="mi">100</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">db_frag</span><span class="p">]</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">vlines</span><span class="p">(</span><span class="n">db_frag</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">db_int</span><span class="p">,</span> <span class="s2">&quot;k&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;DB&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>

<span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">frag_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">frag_dict</span><span class="p">[</span><span class="n">_</span><span class="p">],</span> <span class="mi">104</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.8</span><span class="p">)</span>
    
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Theoretical Spectrum for </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">peptide</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Mass&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Intensity&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">110</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_png output_subarea ">
<img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAmQAAAFNCAYAAACuWnPfAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAgAElEQVR4nO3de7jUZbn/8fctIAcR1CDFI1rkTr3Kw0pT84iiIp5yG9phk1nm1tR+O7eamkptTbFMUyvNVCyDTaZ5Soso7eBWA/NEpngMFAEzwzMI9++P+UIjLRaLxZr1LFjv13XNNTPPd2aee+4Z4MPz/c5MZCaSJEkqZ7XSBUiSJHV1BjJJkqTCDGSSJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmdXIRcXZE/Kh0Hc2JiI0j4tWI6LaCj3NnRHy2vepaFUTN1RHx94i4r3Q9khrLQCYVVgWaRaeFEfFG3fVPlK6vXkQ8ExF7LbqemX/NzL6ZuaCBc64VEVdFxAsR8UpEPB4RpzRqvmrOzhCCPwLsDWyYmduv6INFxOCIyLr31jMRcWrd9oyI15Z4P55cbTs7IuZXYy9HxN0RsWNEnFZ32zcjYkHd9al1j/veJR7nlbrX8tKIGFRXx+7Vn4NXlzjtuKI9kDozA5lUWBVo+mZmX+CvwAF1Y9d1VB0R0b2j5lpO3wL6Au8H+gMHAk+WLKhavWr035+bAM9k5mvLe8dlvJZrVe+1I4AzI2Lfum0frH8/ZuaYum3/W91vIPB74Abg63Xv3WOA/6u775ZLmf9/M3NNYB3gEGA9YEp9KAOeX6KOvpn5f8vVBGklYyCTVg6rR8S11arC1IhoWrQhItaPiJ9GxJyIeDoiTqjb1jMiLoqI56vTRRHRs9q2e0TMiIhTIuIF4OpqfEREPFC3EvKBavyHwMbALYtWT+pWXbpXt1mn2s32fLWr7WfV+NoRcWtV49+ryxu28rl/CPhxZv49Mxdm5l8y8/q655gRcUJEPBURL0bEBfVhKSI+ExGPVvP+IiI2qdu2ZURMjIiXImJWteKzL3AaMLJ6ng9Wt70zIs6JiD8ArwObLbliWL+yVtebIyNiejX/MRHxoYh4qOrvpc094Yg4CrgS2LGqYXQ1/rmIeKKq9+aIWH+JPhwXEdOAactqahVwpgJbteZFqLvffGAstSD1ruW575KPk5lTgZHAHOBLbX0saVVgIJNWDgcC44G1gJuBSwGq4HEL8CCwATAU+GJE7FPd73Tgw8DWwAeB7YEz6h53PWorFZsAR0fEtsBVwOep/WN7OXBzRPTMzE/xzhW8+tWTRX4I9AG2BN5NbXULan/XXF3NszHwxqLn0Ar3AOdUwWbIUm5zCNAEbAscBHwGICIOphauPkptZed3wLhq25rAr4A7gPWB9wKTMvMO4FyqFaHM/GDdPJ8CjgbWBJ5tZf07AEOoBY+LqL0me1Hr0cciYrcl75CZP+CdK05nRcSewNeBjwGDqvnHL3HXg6v5tmipoGqFb+eqhj+18nksum9P4NPAjMx8cXnu25xqd/dNwC4r+ljSysxAJq0cfp+ZP6/+8fohtXAFtdWjgZn51cycl5lPAd8HDq+2fwL4ambOzsw5wGhqoWKRhcBZmflWZr4BfA64PDPvzcwFmTkWeItaqGtRtctpP+CYajVrfmbeBZCZf8vMn2bm65n5CnAO8C9BZCmOB64DvgD8uVoh2m+J25yfmS9l5l+phZ4jqvHPU9ut9mhmvk0taG1drZKNAF7IzG9m5puZ+Upm3ruMWq7JzKmZ+Xa1UtQaX6se/5fAa8C46vV4jlpA3KaVj/MJ4KrMvD8z3wK+TG0FbXDdbb5e9eGNFh7nReAlaitwp2bmpLpt91crd4tO+9Rt+1hEvAxMB7ajFv7ay/PU/mOwyPpL1PFyRKzRjvNJnU5nPWZE0ju9UHf5daBXtZtwE6p/vOq2d6P2Dz3UVn7qV3KercYWmZOZb9Zd3wQYFRHH142tvsR9lmYj4KXM/PuSGyKiD7XVsn2BtavhNSOi27I+EFCFi3OBcyOiH3Aq8JOI2DgzX6puNr3uLvXPcRPg4oj4Zn051FYTN2L5j0Wbvuyb/ItZdZffaOZ631Y+zvrA/YuuZOarEfE3as/lmeWob0AVTpuzbWY+sZRtEzLzk62sdXltQC0kLvJ8ZrZ2l7a0SnCFTFq5TQeezsy16k5rZubwavvz1ELJIhtXY4tkM493zhKP1yczxy3l9kved52IWKuZbV8CNgd2yMx+wK7VeCz7KdYVmzmXWjhbA9i0btNGdZfrn+N04PNLPJ/emXl3te09S5uqleOvUdtFu8h6rXgabfWO17JaMXoX8FzdbVp6fTqlarf7AfzzPxFSl2Qgk1Zu9wFzqwPze0dEt4jYKiI+VG0fB5wREQMjYgBwJtDS1zl8HzgmInaojjNaIyL2r463gtrqzmbN3TEzZwK3A9+pDuLvERGLgtea1FaDXo6IdYCzWvsEI+Ir1YHwq0dEL+BE4GXgsbqb/Xc150bV9v+txr8HfDkitqweq39EHFZtuxVYLyK+GLUPP6wZETvUPc/BsexPUj4AHF491ybg31v7vNrgx8CREbF1dRzXucC9mflMA+dsmKpn76f2Hl0PuLBwSVJRBjJpJVbt7juA2kH7T1M7PuhKal8PAfA/wGTgIeBharu8/qeFx5tM7TiyS4G/A09QO4B7ka9TC3gvR8RJzTzEp4D5wF+A2cAXq/GLgN5VffdQO5C+1U+T2gcCXqS2SrQ3sH9mvlp3m5uAKdQC0m3AD6rncyNwPjA+IuYCj1A7zo3qWLa9qfXvBWqfTNyjeryfVOd/i4jFuwmb8RVqq2x/p3Z83o+X43ktl+pYr68APwVmVvMe3uKdlt+D8c7v/rqonR8fqk+vUgvVNwN/A7bLzPqV2/XjX7+H7NAG1CJ1GpG50q1wS9JiEZHAkBaOfZKkTs8VMkmSpMIMZJIkSYW5y1KSJKkwV8gkSZIKM5BJkiQVtlJ/U/+AAQNy8ODBpcuQJElapilTpryYmQOb27ZSB7LBgwczefLk0mVIkiQtU0Q8u7Rt7rKUJEkqzEAmSZJUmIFMkiSpMANZKx1wwAHcd999pctQK/l6rTh72P7saTn2viz7v2wGshU0f/58Tj75ZA444ACampqYMmVK6ZKaNXHiRD7zmc+w8847c/TRR5cup5iHH36YY489lj333JO99tqLU045hRdffLHh81500UUccsgh7Lrrrhx66KHcdtttDZ+zUZ566ik+9alPsccee7DHHntw7LHH8tRTT5Uua6m+/e1vM3z4cHbddVdGjBjBVVddVbqkFl1xxRU0NTWtFP94zZ07l7322oujjjqqdClt9vzzz9PU1MQuu+yy+HTllVc2fN6zzz6bD3/4w++Yd+HChQ2ftzN68803Oe+88xg6dCi77bYbn/vc50qX1Kz77ruPT3ziE3zkIx9h+PDhTJw4sV0ff6X+lGVnsfXWW/Pxj3+cU045pXQpS9W/f3+OOOIInnnmGf74xz+WLqeYV155hY9+9KPsuOOOdOvWjTFjxjB69GguueSShs7bu3dvvvWtb7Hxxhvz5z//meOPP56NNtqID3zgAw2dtxEGDhzI+eefz6BBg8hMJkyYwGmnncb48eNLl9asgw46iM997nP07t2b2bNn84UvfIHBgwez5557li7tX8yYMYNJkyYxYMCA0qW0yre//W023XTTVSJI3HnnnXTr1q1D5/yP//gPjj322A6dszM655xzWLBgAddffz39+/fnscceK13Sv3jqqac4/fTTGT16NDvssAOvvvoqr7zySrvO4QrZcpg6dSqHHXYYe+yxB6NHj2bevHn06NGDj3/842y99dYd/od5Sddeey0nn3zyO8bGjBnDN7/5Tbbffnv23ntvBg5s9utPVknNvV477bQTe+21F2ussQa9evXiYx/7GA8++GC7zNdS/z//+c8zePBgVlttNbbaaiu22WYbHnrooXaZt5Ga6+Gaa67J+uuvT0SQmay22mpMnz69aJ0t9X6TTTahd+/ei8cjghkzZnR0iYs119NFxowZwwknnECPHj2K1bdISz0FeOihh3jyySc54IADSpTXJi31vhGW1cOuprn+P/vss9x1112cfvrprL322qy22mq8//3vL1JfS6/XD37wAz760Y+y00470a1bN/r378+GG27YrvMbyJbD7bffzqWXXspNN93Es88+2yHL2stj+PDh3H333YtT+4IFC5g4cSLDhw8vXFkZrXm97r//fjbbbLN2ma+1/X/rrbeYOnVqu83bSC31cPfdd2fHHXfkggsu4DOf+UzBKpfd+2uuuYZddtmF4cOH88Ybb7DvvvsWq3VpPf3Vr35F9+7d2XnnnYvVVq+lni5cuJDzzz+fk08+mYgoXGnrtfR+HjFiBMOHD2f06NG8/PLL7TLfst6X119/PXvuuSef/OQn+fWvf90uc3ZmzfX/kUceYdCgQVx++eUMHTqUkSNHFutFS6/Xww8/DMDIkSPZZ599+MpXvsLcuXPbdX4D2XIYOXIk6667Lv369eOoo47iF7/4RemS3mHAgAFsu+22/OpXvwLg7rvvZq211ir2v43SlvV6TZs2jSuvvJITTzyxXeZrbf/PPfdc3ve+97Hjjju2y7yN1FIP77zzTu666y5OPvlkNt9884JVLrv3n/70p/ntb3/Lddddx/7770/fvn2L1dpcT19//XUuu+wyTjrppGJ1Lamlno4fP56tttpqpfu7pbner7XWWlx77bXceuut/OhHP+K1117jjDPOaJf5Wurh4Ycfzo033sjEiRP5z//8T84+++x2W63vrJrr/6xZs3jyySfp27cvd9xxB6eccgpnnXUWTz/9dIfX19LrNXv2bH7+859zwQUXcOONN/LWW28xZsyYdp3fQLYc1l133cWXBw0axJw5cwpW07wRI0Zw++23A7X/jXTV1TFo+fWaPn06J5xwAieddBLbbLNNu825rP5ffPHFPPnkk5x33nkrxcrCst7zvXv35tBDD+XMM8/kpZde6ujy3mFZvY8INt98c3r27Mn3vve9EiUCzff08ssvZ/jw4ay//vrF6mpOcz2dM2cO48eP57jjjitc3fJrrvd9+vRhiy22oFu3bqyzzjqccsop3HPPPbz22mvtMufS3pf/9m//Rv/+/enWrRs777wz++677yq/StZc/3v16kX37t056qij6NGjB9tuuy1NTU3cc889RWpc2uvVs2dPDjjgADbeeGP69OnDkUceyR/+8Id2ndtAthxmzZq1+PILL7zQKY/H2n333Zk2bRpPPvkkv/vd79hvv/1Kl1TM0l6vmTNncuyxx/LZz3623QNrS/2//PLL+cMf/sBll13GGmus0a7zNkpr3vOZyZtvvln8Pyitfe8vWLCg6DFkzfX0j3/8I+PHj2fYsGEMGzaMWbNmceqppzJ27NhidULzPZ06dSovvvgi//7v/86wYcP4xje+wdSpUxk2bFinP7i/Ne/nRf9Rysx2mbO178tFx2Suyprr/3vf+96CFf2rpb1eQ4YMafh/og1ky2HChAnMnj2buXPnctVVVzFs2DAA5s2bt/jg0Pnz5zNv3rxif7BWX311hg4dyumnn86WW27JeuutB8DChQuZN28eCxYsIDOZN28eb7/9dpEaO0pzr9fs2bM55phjOOywwzj00EPbfc6l9f/qq6/mjjvu4Dvf+Q79+/dv93kbpbke3nvvvTz22GMsXLiQ1157jQsvvJB+/fqx6aabFq21ud4vXLiQG264gblz55KZTJ06lQkTJrD99tsXq7O5nn73u99lwoQJjBs3jnHjxjFgwABOO+00DjvssGJ1QvM93WmnnbjlllsW13rMMcew+eabM27cOFZbrXP/k9Jc7x955BGeffZZFi5cyD/+8Q8uuOACtttuu3bbrb20vxMmTZrE66+/zsKFC7nnnnu4/fbb2W233dplzs6quf5vu+22rLfeelx99dUsWLCABx98kClTphQ7pGNpr9eBBx7ILbfcwnPPPcebb77J2LFj2WWXXdp1br/2Yjnsu+++HHfcccyZM4fddttt8XfvHHroocycOROAL3zhCwDcfPPNxXY/jBgxgp/97GeceeaZi8duu+02Ro8evfj6TjvtxIgRIzj77LMLVNgxmnu9rr32Wp577jm+//3v8/3vf3/xbX/3u9+127zN9f+yyy6jR48eHHLIIYvHjjzyyOIHwy9Lcz38/e9/z5gxY5g9ezY9e/Zkiy224JJLLmH11VcvXW6zvf/Nb37DpZdeyvz58xk4cCAjR45k5MiRxWpsrqe9evV6x226detGv3796NOnT6Eq/2nJnq6++uq8613vWry9b9++dO/e/R1jnVVzvb/rrru47LLLeOmll1hjjTXYYYcdOPfcc9t13ubel+PGjeOrX/0qmckGG2zAGWecwXbbbdeu83Y2zfW/e/fuXHjhhXzta1/jmmuuYdCgQYwePZrBgwcXq7O51+vAAw9k5syZjBo1Cqj9G9rex3zGyrxE2tTUlJMnTy5dRqfzwgsvcOihh/LLX/5ypdk1tiqx/+XY+/ZnT1ecPVy5NPL1iogpmdnU3LbOvb6s5bZw4UKuu+469tlnH//gF2D/y7H37c+erjh7uHIp+Xq5y3IV8sYbbzBs2DAGDRrU8G+e17+y/+XY+/ZnT1ecPVy5lH693GUpSZLUAdxlKUmS1IkZyCRJkgprWCCLiKsiYnZEPFI3tk5ETIyIadX52nXbvhwRT0TEYxGxT6PqkiRJ6mwauUJ2DbDkL/ieCkzKzCHApOo6EbEFcDiwZXWf70REtwbWJkmS1Gk0LJBl5m+BJX/c7iBg0W+BjAUOrhsfn5lvZebTwBNAua/SliRJ6kAdfQzZupk5E6A6f3c1vgEwve52M6oxSZKkVV5nOai/uV/sbPb7OCLi6IiYHBGTO+LHjB9//HEef/zxhs+zolaWOhutVB9Wpf6vbM9lZah3ZahxkZWp1tbw74SyVpY+dIY6OzqQzYqIQQDV+exqfAawUd3tNgSeb+4BMvOKzGzKzKaBAwc2tFhJkqSO0NGB7GZgVHV5FHBT3fjhEdEzIjYFhgD3dXBtkiRJRTTsp5MiYhywOzAgImYAZwHnARMi4ijgr8BhAJk5NSImAH8G3gaOy8wFjapNkiSpM2lYIMvMI5ayaehSbn8OcE6j6pEkSeqsOstB/ZIkSV2WgUySJKkwA5kkSVJhBjJJkqTCDGSSJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmSZJUmIFMkiSpMAOZJElSYQYySZKkwgxkkiRJhRnIJEmSCjOQSZIkFWYgkyRJKsxAJkmSVJiBTJIkqTADmSRJUmEGMkmSpMIMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSrMQCZJklSYgUySJKkwA5kkSVJhBjJJkqTCDGSSJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmSZJUmIFMkiSpMAOZJElSYQYySZKkwgxkkiRJhRnIJEmSCjOQSZIkFWYgkyRJKqxIIIuI/xcRUyPikYgYFxG9ImKdiJgYEdOq87VL1CZJktTROjyQRcQGwAlAU2ZuBXQDDgdOBSZl5hBgUnVdkiRplVdql2V3oHdEdAf6AM8DBwFjq+1jgYML1SZJktShOjyQZeZzwDeAvwIzgX9k5i+BdTNzZnWbmcC7O7o2SZKkEkrsslyb2mrYpsD6wBoR8cnluP/RETE5IibPmTOnUWVKkiR1mBK7LPcCns7MOZk5H7gB2AmYFRGDAKrz2c3dOTOvyMymzGwaOHBghxUtSZLUKCUC2V+BD0dEn4gIYCjwKHAzMKq6zSjgpgK1SZIkdbjuHT1hZt4bEdcD9wNvA38CrgD6AhMi4ihqoe2wjq5NkiSphA4PZACZeRZw1hLDb1FbLZMkSepS/KZ+SZKkwgxkkiRJhRnIJEmSCjOQSZIkFWYgkyRJKsxAJkmSVJiBTJIkqTADmSRJUmEGMkmSpMIMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSrMQCZJklSYgUySJKkwA5kkSVJhBjJJkqTCDGSSJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmSZJUmIFMkiSpMAOZJElSYQYySZKkwgxkkiRJhRnIJEmSCjOQSZIkFWYgkyRJKsxAJkmSVJiBTJIkqTADmSRJUmEGMkmSpMIMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSqsVYEsItZpdCGSJEldVWtXyO6NiJ9ExPCIiIZWJEmS1MW0NpC9D7gC+BTwREScGxHva1xZkiRJXUerAlnWTMzMI4DPAqOA+yLirojYsaEVSpIkreJaewzZuyLixIiYDJwEHA8MAL4E/Hh5J42ItSLi+oj4S0Q8GhE7RsQ6ETExIqZV52sv7+NKkiStjFq7y/L/gH7AwZm5f2bekJlvZ+Zk4HttmPdi4I7M/Dfgg8CjwKnApMwcAkyqrkuSJK3yWhvIzsjMr2XmjEUDEXEYQGaevzwTRkQ/YFfgB9X952Xmy8BBwNjqZmOBg5fncSVJklZWrQ1kza1WfbmNc24GzAGujog/RcSVEbEGsG5mzgSozt/dxseXJElaqXRvaWNE7AcMBzaIiG/XbeoHvL0Cc24LHJ+Z90bExSzH7smIOBo4GmDjjTduYwmSJEmdx7JWyJ4HJgNvAlPqTjcD+7RxzhnAjMy8t7p+PbWANisiBgFU57Obu3NmXpGZTZnZNHDgwDaWIEmS1Hm0uEKWmQ8CD0bEdZnZ1hWxJR/zhYiYHhGbZ+ZjwFDgz9VpFHBedX5Te8wnSZLU2S1rl+WEzPwY8KeIyPpN1L6e7ANtnPd44LqIWB14CjiS2mrdhIg4CvgrcFgbH1uSJGml0mIgA06szke056SZ+QDQ1Mymoe05jyRJ0sqgxWPIFn3qEXgRmJ6ZzwI9qX132PMNrk2SJKlLaO3XXvwW6BURG1D70tYjgWsaVZQkSVJX0tpAFpn5OvBR4JLMPATYonFlSZIkdR2tDmTVj4h/AritGlvW8WeSJElqhdYGshOpfTP/jZk5NSI2A37TuLIkSZK6jlatcmXmb6kdR7bo+lPACY0qSpIkqStpVSCLiPcBJwGD6++TmXs2pixJkqSuo7XHgf0E+B5wJbCgceVIkiR1Pa0NZG9n5ncbWokkSVIX1dqD+m+JiGMjYlBErLPo1NDKJEmSuojWrpCNqs7/u24sgc3atxxJkqSup7Wfsty00YVIkiR1Va3aZRkRfSLijIi4oro+JCLa9QfHJUmSuqrWHkN2NTAP2Km6PgP4n4ZUJEmS1MW0NpC9JzPHAPMBMvMNIBpWlSRJUhfS2kA2LyJ6UzuQn4h4D/BWw6qSJEnqQlr7KcuzgTuAjSLiOmBn4MhGFSVJktSVtPZTlr+MiCnAh6ntqjwxM19saGWSJEldRGs/ZTkpM/+Wmbdl5q2Z+WJETGp0cZIkSV1BiytkEdEL6AMMiIi1+eeB/P2A9RtcmyRJUpewrF2Wnwe+SC18TeGfgWwucFkD65IkSeoyWgxkmXkxcHFEHJ+Zl3RQTZIkSV1Kaw/qvyQidgIG198nM69tUF2SJEldRqsCWUT8EHgP8ACwoBpOwEAmSZK0glr7PWRNwBaZmY0sRpIkqStq7Tf1PwKs18hCJEmSuqrWrpANAP4cEfdR95NJmXlgQ6qSJEnqQpbnp5MkSZLUAK39lOVdjS5EkiSpq1rWN/W/Qu3TlP+yCcjM7NeQqiRJkrqQZX0x7JodVYgkSVJX1dpPWUqSJKlBDGSSJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmSZJUmIFMkiSpMAOZJElSYQYySZKkwooFsojoFhF/iohbq+vrRMTEiJhWna9dqjZJkqSOVHKF7ETg0brrpwKTMnMIMKm6LkmStMorEsgiYkNgf+DKuuGDgLHV5bHAwR1dlyRJUgmlVsguAk4GFtaNrZuZMwGq83eXKEySJKmjdXggi4gRwOzMnNLG+x8dEZMjYvKcOXPauTpJkqSOV2KFbGfgwIh4BhgP7BkRPwJmRcQggOp8dnN3zswrMrMpM5sGDhzYUTVLkiQ1TIcHssz8cmZumJmDgcOBX2fmJ4GbgVHVzUYBN3V0bZIkSSV0pu8hOw/YOyKmAXtX1yVJklZ53UtOnpl3AndWl/8GDC1ZjyRJUgmdaYVMkiSpSzKQSZIkFWYgkyRJKsxAJkmSVJiBTJIkqTADmSRJUmEGMkmSpMIMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSrMQCZJklSYgUySJKkwA5kkSVJhBjJJkqTCDGSSJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmSZJUmIFMkiSpMAOZJElSYQYySZKkwgxkkiRJhRnIJEmSCjOQSZIkFWYgkyRJKsxAJkmSVJiBTJIkqTADmSRJUmEGMkmSpMIMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSrMQCZJklSYgUySJKkwA5kkSVJhBjJJkqTCDGSSJEmFGcgkSZIK6/BAFhEbRcRvIuLRiJgaESdW4+tExMSImFadr93RtUmSJJVQYoXsbeBLmfl+4MPAcRGxBXAqMCkzhwCTquuSJEmrvA4PZJk5MzPvry6/AjwKbAAcBIytbjYWOLija5MkSSqh6DFkETEY2Aa4F1g3M2dCLbQB717KfY6OiMkRMXnOnDkdVaokSVLDFAtkEdEX+Cnwxcyc29r7ZeYVmdmUmU0DBw5sXIGSJEkdpEggi4ge1MLYdZl5QzU8KyIGVdsHAbNL1CZJktTRSnzKMoAfAI9m5oV1m24GRlWXRwE3dXRtkiRJJXQvMOfOwKeAhyPigWrsNOA8YEJEHAX8FTisQG2SJEkdrsMDWWb+HoilbB7akbVIkiR1Bn5TvyRJUmEGMkmSpMIMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSrMQCZJklSYgUySJKkwA5kkSVJhBjJJkqTCDGSSJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmSZJUmIFMkiSpMAOZJElSYQYySZKkwgxkkiRJhRnIJEmSCjOQSZIkFWYgkyRJKsxAJkmSVJiBTE66BlYAAAbXSURBVJIkqTADmSRJUmEGMkmSpMIMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSrMQCZJklSYgUySJKkwA5kkSVJhBjJJkqTCDGSSJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmSZJUmIFMkiSpsE4XyCJi34h4LCKeiIhTS9cjSZLUaJ0qkEVEN+AyYD9gC+CIiNiibFWSJEmN1akCGbA98ERmPpWZ84DxwEGFa5IkSWqozhbINgCm112fUY1JkiStsrqXLmAJ0cxYvuMGEUcDR1dXX42IxxpeVc0A4MUOmqsrsJ/ty362L/vZvuxn+7GX7auj+7nJ0jZ0tkA2A9io7vqGwPP1N8jMK4ArOrIogIiYnJlNHT3vqsp+ti/72b7sZ/uyn+3HXravztTPzrbL8o/AkIjYNCJWBw4Hbi5ckyRJUkN1qhWyzHw7Ir4A/ALoBlyVmVMLlyVJktRQnSqQAWTmz4Gfl66jGR2+m3QVZz/bl/1sX/azfdnP9mMv21en6Wdk5rJvJUmSpIbpbMeQSZIkdTkGskpEXBURsyPikbqxdSJiYkRMq87Xrtv25ernnR6LiH3KVN05RcRGEfGbiHg0IqZGxInVuP1sg4joFRH3RcSDVT9HV+P2cwVERLeI+FNE3Fpdt59tFBHPRMTDEfFAREyuxuxnG0XEWhFxfUT8pfp7dEf7ufwiYvPqPbnoNDcivthpe5mZnmq7bXcFtgUeqRsbA5xaXT4VOL+6vAXwINAT2BR4EuhW+jl0lhMwCNi2urwm8HjVM/vZtn4G0Le63AO4F/iw/Vzhvv4X8GPg1uq6/Wx7L58BBiwxZj/b3s+xwGery6sDa9nPFe5pN+AFat8D1il76QpZJTN/C7y0xPBB1P5gUJ0fXDc+PjPfysyngSeo/eyTgMycmZn3V5dfAR6l9osL9rMNsubV6mqP6pTYzzaLiA2B/YEr64btZ/uyn20QEf2oLRD8ACAz52Xmy9jPFTUUeDIzn6WT9tJA1rJ1M3Mm1EIG8O5q3J94aqWIGAxsQ21Vx362UbV77QFgNjAxM+3nirkIOBlYWDdmP9sugV9GxJTq11TAfrbVZsAc4Opql/qVEbEG9nNFHQ6Mqy53yl4ayNpmmT/xJIiIvsBPgS9m5tyWbtrMmP2sk5kLMnNrar9esX1EbNXCze1nCyJiBDA7M6e09i7NjNnPd9o5M7cF9gOOi4hdW7it/WxZd2qHz3w3M7cBXqO2W21p7OcyVF80fyDwk2XdtJmxDuulgaxlsyJiEEB1PrsaX+ZPPHV1EdGDWhi7LjNvqIbt5wqqdl3cCeyL/WyrnYEDI+IZYDywZ0T8CPvZZpn5fHU+G7iR2m4e+9k2M4AZ1So4wPXUApr9bLv9gPszc1Z1vVP20kDWspuBUdXlUcBNdeOHR0TPiNgUGALcV6C+TikigtrxD49m5oV1m+xnG0TEwIhYq7rcG9gL+Av2s00y88uZuWFmDqa2G+PXmflJ7GebRMQaEbHmosvAMOAR7GebZOYLwPSI2LwaGgr8Gfu5Io7gn7srobP2svQnHzrLidqLNROYTy0lHwW8C5gETKvO16m7/enUPoHxGLBf6fo70wn4CLVl3oeAB6rTcPvZ5n5+APhT1c9HgDOrcfu54r3dnX9+ytJ+tq2Hm1H7ZNqDwFTgdPu5wj3dGphc/Zn/GbC2/WxzL/sAfwP61411yl76Tf2SJEmFuctSkiSpMAOZJElSYQYySZKkwgxkkiRJhRnIJEmSCjOQSeoSIiIj4od117tHxJyIuLVkXZIEBjJJXcdrwFbVl+sC7A08V7AeSVrMQCapK7kd2L+6/I5v746I7SPi7uoHne9e9E3pEbFlRNwXEQ9ExEMRMaT6dvrbIuLBiHgkIkYWeC6SViEGMkldyXhqP43Si9ovINxbt+0vwK5Z+0HnM4Fzq/FjgIuz9uPuTdR+yWNf4PnM/GBmbgXc0VFPQNKqqXvpAiSpo2TmQxExmNrq2M+X2NwfGBsRQ6j99FePavz/gNMjYkPghsycFhEPA9+IiPOp/fTS7zrkCUhaZblCJqmruRn4Bu/8sWGArwG/qVa8DgB6AWTmj4EDgTeAX0TEnpn5OLAd8DDw9Yg4s6OKl7RqcoVMUldzFfCPzHw4InavG+/PPw/y//SiwYjYDHgqM79dXf5ARPwFeCkzfxQRr9bfXpLawkAmqUvJzBnAxc1sGkNtl+V/Ab+uGx8JfDIi5gMvAF8FPgRcEBELgfnAfza2akmrusjM0jVIkiR1aR5DJkmSVJiBTJIkqTADmSRJUmEGMkmSpMIMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSrs/wPekDfCKLnORAAAAABJRU5ErkJggg==
"
>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Spectra">Spectra<a class="anchor-link" href="#Spectra"> </a></h3><p>The function <a href="/alphapept/fasta#get_spectrum"><code>get_spectrum</code></a> returns a tuple with the following content:</p>
<ul>
<li>precursor mass</li>
<li>peptide sequence</li>
<li>fragmasses</li>
<li>fragtypes</li>
</ul>
<p>Likewise, <a href="/alphapept/fasta#get_spectra"><code>get_spectra</code></a> returns a list of tuples. We employ a list of tuples here as this way we can sort them easily by precursor mass.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_spectrum" class="doc_header"><code>get_spectrum</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L380" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_spectrum</code>(<strong><code>peptide</code></strong>, <strong><code>mass_dict</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_spectra" class="doc_header"><code>get_spectra</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L394" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_spectra</code>(<strong><code>peptides</code></strong>, <strong><code>mass_dict</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">get_spectra</span><span class="p">([</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">],</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>[(799.3599642034599, &#39;PEPTIDE&#39;, array([ 98.06004033, 148.06043425, 227.10263343, 263.08737735,
       324.15539729, 376.17144135, 425.20307579, 477.21911985,
       538.28713979, 574.27188371, 653.31408289, 703.31447681]), array([0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1], dtype=int8))]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Reading-FASTA">Reading FASTA<a class="anchor-link" href="#Reading-FASTA"> </a></h2><p>To read FASTA files we use the <code>SeqIO</code> module from the <code>Biopython</code> library.
Generator expression so that we can read it one by one</p>
<p>Additionally we define the wrapper <a href="/alphapept/fasta#read_fasta"><code>read_fasta</code></a> that allows to read multiple FASTA files.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="read_fasta_file" class="doc_header"><code>read_fasta_file</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L410" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>read_fasta_file</code>(<strong><code>fasta_filename</code></strong>=<em><code>''</code></em>)</p>
</blockquote>
<p>given a fasta_file read fasta file line by line, return progress</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="read_fasta_file_entries" class="doc_header"><code>read_fasta_file_entries</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L437" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>read_fasta_file_entries</code>(<strong><code>fasta_filename</code></strong>=<em><code>''</code></em>)</p>
</blockquote>
<p>Function to count entries in fasta file</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="read_fasta" class="doc_header"><code>read_fasta</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L455" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>read_fasta</code>(<strong><code>path</code></strong>)</p>
</blockquote>
<p>Wrapper to read multiple files.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="check_sequence" class="doc_header"><code>check_sequence</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L471" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>check_sequence</code>(<strong><code>element</code></strong>, <strong><code>AAs</code></strong>)</p>
</blockquote>
<p>Checks wheter a sequence from a FASTA entry contains valid AAs</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="c1">#load example fasta file</span>

<span class="n">fasta_path</span> <span class="o">=</span> <span class="s1">&#39;../testfiles/test.fasta&#39;</span>

<span class="nb">list</span><span class="p">(</span><span class="n">read_fasta</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;id&#39;: &#39;A0PJZ0&#39;,
 &#39;name&#39;: &#39;sp|A0PJZ0|A20A5_HUMAN&#39;,
 &#39;description&#39;: &#39;sp|A0PJZ0|A20A5_HUMAN Putative ankyrin repeat domain-containing protein 20A5 OS=Homo sapiens OX=9606 GN=ANKRD20A5P PE=5 SV=1&#39;,
 &#39;sequence&#39;: &#39;MKLFGFRSRRGQTVLGSIDHLYTGSGYRIRYSELQKIHKAAVKGDAAEMERCLARRSGDLDALDKQHRTALHLACASGHVKVVTLLVNRKCQIDIYDKENRTPLIQAVHCQEEACAVILLEHGANPNLKDIYGNTALHYAVYSESTSLAEKLLFHGENIEALDKV&#39;}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Peptide-Dictionary">Peptide Dictionary<a class="anchor-link" href="#Peptide-Dictionary"> </a></h2><p>In order to efficiently store peptides we rely on the python in dictionary. The idea is to have a dictionary with peptides as keys and indicies to proteins as values. This way we can quickly look up to which protein a peptide belongs to. The function <a href="/alphapept/fasta#add_to_pept_dict"><code>add_to_pept_dict</code></a> uses a regular python dictionary and allows to add peptides and stores indicies to the originating proteins as a list. If a peptide is already present in the dictionary the list is appended. The function returns a list of <code>added_peptides</code> which were not present in the dictionary yet.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_to_pept_dict" class="doc_header"><code>add_to_pept_dict</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L482" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_to_pept_dict</code>(<strong><code>pept_dict</code></strong>, <strong><code>new_peptides</code></strong>, <strong><code>i</code></strong>)</p>
</blockquote>
<p>Add peptides to the peptide dictionary</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">pept_dict</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">new_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ABC&#39;</span><span class="p">,</span><span class="s1">&#39;DEF&#39;</span><span class="p">]</span>

<span class="n">pept_dict</span><span class="p">,</span> <span class="n">added_peptides</span> <span class="o">=</span> <span class="n">add_to_pept_dict</span><span class="p">(</span><span class="n">pept_dict</span><span class="p">,</span> <span class="n">new_peptides</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

<span class="n">new_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;DEF&#39;</span><span class="p">,</span><span class="s1">&#39;GHI&#39;</span><span class="p">]</span>

<span class="n">pept_dict</span><span class="p">,</span> <span class="n">added_peptides</span> <span class="o">=</span> <span class="n">add_to_pept_dict</span><span class="p">(</span><span class="n">pept_dict</span><span class="p">,</span> <span class="n">new_peptides</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">pept_dict</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>{&#39;ABC&#39;: [0], &#39;DEF&#39;: [0, 1], &#39;GHI&#39;: [1]}
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Generating-a-library">Generating a library<a class="anchor-link" href="#Generating-a-library"> </a></h2><p>To wrap everything up, we employ two functions <a href="/alphapept/fasta#generate_library"><code>generate_library</code></a> and <a href="/alphapept/fasta#generate_spectra"><code>generate_spectra</code></a>. The first one reads a FASTA file and generates a list of peptides, as well as the peptide dictionary and an ordered FASTA dictionary to be able to look up the protein indices laster. For the <code>callback</code> we first read the whole FASTA file to determine the total number of entries in the FASTA file.  For a typical FASTA file of 30 Mb with 40k entries, this should take less than a second. The progress of the digestion is monitored by processing the FASTA file one by one.
The function <a href="/alphapept/fasta#generate_spectra"><code>generate_spectra</code></a> then calculates precursor masses and fragment ions. Here, we split the total_number of sequences in <code>1000</code> steps to be able to track progress with the <code>callback</code>.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="generate_library" class="doc_header"><code>generate_library</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L499" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>generate_library</code>(<strong><code>mass_dict</code></strong>, <strong><code>fasta_path</code></strong>, <strong><code>callback</code></strong>=<em><code>None</code></em>, <strong><code>contaminants_path</code></strong>=<em><code>None</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Function to generate a library from a fasta file</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="generate_spectra" class="doc_header"><code>generate_spectra</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L542" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>generate_spectra</code>(<strong><code>to_add</code></strong>, <strong><code>mass_dict</code></strong>, <strong><code>callback</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Function to generate a library from a fasta file</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Saving">Saving<a class="anchor-link" href="#Saving"> </a></h2><p>To save the generated spectra, we rely on NumPy's NPZ format. For this we create a dictionary and save all the generated elements. The container will contain the following elements:</p>
<ul>
<li><code>precursors</code>: An array containing the precursor masses</li>
<li><code>seqs</code>: An array containing the peptide sequences for the precursor masses</li>
<li><code>pept_dict</code>: A peptide dictionary to look up the peptides and return their FASTA index</li>
<li><code>fasta_dict</code>: A fasta dictionary to look up the FASTA entry based on a pept_dict index</li>
<li><code>fragmasses</code>: An array containing the fragment masses. Unoccupied cells are filled with -1</li>
<li><code>fragtypes:</code>: An array containg the fragment types. 0 equals b-ions and 1 equals y-ions. Unoccupied cells are filled with -1</li>
<li><code>bounds</code>: An integer array containing the upper bounds for the fragment masses / types array. This is needed to quickly slice the data.</li>
</ul>
<p>All arrays are sorted according to the precursor mass.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="save_library" class="doc_header"><code>save_library</code><a href="__main__.py#L5" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>save_library</code>(<strong><code>spectra</code></strong>, <strong><code>pept_dict</code></strong>, <strong><code>fasta_dict</code></strong>, <strong><code>library_path</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Function to save a library to the *.npz format.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

