---

title: FASTA


keywords: fastai
sidebar: home_sidebar

summary: "Functions related to generating spectra from FASTA files"
description: "Functions related to generating spectra from FASTA files"
nb_path: "nbs\03_fasta.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbs\03_fasta.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This notebook contains all functions related to creating spectra from FASTA files. In brief, what we are doing is the following:</p>
<ol>
<li>Read a FASTA file and digest the proteins to generate peptides</li>
<li>For each peptide, calculate a theoretical spectrum and precursor mass</li>
<li>Save spectra</li>
</ol>
<p>Currently, <code>numba</code> has only limited string support. A lot of the functions are therefore Python-native.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Cleaving">Cleaving<a class="anchor-link" href="#Cleaving"> </a></h2><p>For cleaving, we use regular expressions to find potential cleavages sites and write the wrapper <a href="/alphapept/fasta.html#cleave_sequence"><code>cleave_sequence</code></a> to use it.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_missed_cleavages" class="doc_header"><code>get_missed_cleavages</code><a href="__main__.py#L5" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_missed_cleavages</code>(<strong><code>sequences</code></strong>, <strong><code>n_missed_cleavages</code></strong>)</p>
</blockquote>
<p>Combine cleaved sequences to get sequences with missed cleavages</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="cleave_sequence" class="doc_header"><code>cleave_sequence</code><a href="__main__.py#L16" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>cleave_sequence</code>(<strong><code>sequence</code></strong>=<em><code>''</code></em>, <strong><code>num_missed_cleavages</code></strong>=<em><code>0</code></em>, <strong><code>protease</code></strong>=<em><code>'trypsin'</code></em>, <strong><code>min_length</code></strong>=<em><code>6</code></em>, <strong><code>max_length</code></strong>=<em><code>65</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Cleave a sequence with a given protease. Filters to have a minimum and maximum length.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">protease</span> <span class="o">=</span> <span class="s2">&quot;trypsin&quot;</span>
<span class="n">num_missed_cleavages</span> <span class="o">=</span> <span class="mi">0</span>
<span class="n">min_length</span><span class="p">,</span> <span class="n">max_length</span> <span class="o">=</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">65</span>

<span class="n">cleave_sequence</span><span class="p">(</span><span class="s1">&#39;ABCDEFGHIJKLMNOPQRST&#39;</span><span class="p">,</span> <span class="n">num_missed_cleavages</span><span class="p">,</span> <span class="n">protease</span><span class="p">,</span> <span class="n">min_length</span><span class="p">,</span> <span class="n">max_length</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;ABCDEFGHIJK&#39;, &#39;LMNOPQR&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Counting-missed-and-internal-cleavages">Counting missed and internal cleavages<a class="anchor-link" href="#Counting-missed-and-internal-cleavages"> </a></h2><p>The following are helper functions to retrieve the number of missed cleavages and internal cleavage sites for each sequence.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="count_missed_cleavages" class="doc_header"><code>count_missed_cleavages</code><a href="__main__.py#L5" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>count_missed_cleavages</code>(<strong><code>sequence</code></strong>=<em><code>''</code></em>, <strong><code>protease</code></strong>=<em><code>'trypsin'</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Counts the number of missed cleavages for a given sequence and protease</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="count_internal_cleavages" class="doc_header"><code>count_internal_cleavages</code><a href="__main__.py#L15" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>count_internal_cleavages</code>(<strong><code>sequence</code></strong>=<em><code>''</code></em>, <strong><code>protease</code></strong>=<em><code>'trypsin'</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Counts the number of internal cleavage sites for a given sequence and protease</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">protease</span> <span class="o">=</span> <span class="s2">&quot;trypsin&quot;</span>
<span class="nb">print</span><span class="p">(</span><span class="n">count_missed_cleavages</span><span class="p">(</span><span class="s1">&#39;ABCDEFGHIJKLMNOPQRST&#39;</span><span class="p">,</span> <span class="n">protease</span><span class="p">))</span>

<span class="n">protease</span> <span class="o">=</span> <span class="s2">&quot;trypsin&quot;</span>
<span class="nb">print</span><span class="p">(</span><span class="n">count_internal_cleavages</span><span class="p">(</span><span class="s1">&#39;ABCDEFGHIJKLMNOPQRST&#39;</span><span class="p">,</span> <span class="n">protease</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>2
1
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Parsing">Parsing<a class="anchor-link" href="#Parsing"> </a></h2><p>Peptides are composed out of amino acids that are written in capital letters - <code>PEPTIDE</code>. to distinguish modifications, they are written in lowercase such as <code>PEPTIoxDE</code> and can be of arbitrary length. For a modified amino acid, the modification precedes the letter of the amino acid. Decoys are indicated with an underscore. Therfore, the <a href="/alphapept/fasta.html#parse"><code>parse</code></a> function splits after <code>_</code>. When parsing, the peptide string is converted into a numba-compatible list, like so: <code>PEPoxTIDE</code> -&gt; <code>[P, E, P, oxT, I, D, E]</code>. This allows that we can use the <a href="/alphapept/constants.html#mass_dict"><code>mass_dict</code></a> from <code>alphapept.constants</code> to directly determine the masses for the corresponding amino acids.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="parse" class="doc_header"><code>parse</code><a href="__main__.py#L5" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>parse</code>(<strong><code>peptide</code></strong>)</p>
</blockquote>
<p>Parser to parse peptide strings</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="list_to_numba" class="doc_header"><code>list_to_numba</code><a href="__main__.py#L23" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>list_to_numba</code>(<strong><code>a_list</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPoxTIDE&#39;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>[P, E, P, T, I, D, E]
[P, E, P, oxT, I, D, E]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Decoy">Decoy<a class="anchor-link" href="#Decoy"> </a></h2><p>The decoy strategy employed is a pseudo-reversal of the peptide sequence, keeping only the terminal amino-acid and reversing the rest. Additionally, we can call the functions <a href="/alphapept/fasta.html#swap_KR"><code>swap_KR</code></a> and and <a href="/alphapept/fasta.html#swap_AL"><code>swap_AL</code></a> that will swap the respective AAs. The function <a href="/alphapept/fasta.html#swap_KR"><code>swap_KR</code></a> will only swap terminal AAs. The swapping functions only work if the AA is not modified.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_decoy_sequence" class="doc_header"><code>get_decoy_sequence</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L117" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_decoy_sequence</code>(<strong><code>peptide</code></strong>, <strong><code>pseudo_reverse</code></strong>=<em><code>True</code></em>, <strong><code>AL_swap</code></strong>=<em><code>False</code></em>, <strong><code>KR_swap</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>Reverses a sequence and adds the '_decoy' tag.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="swap_KR" class="doc_header"><code>swap_KR</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L141" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>swap_KR</code>(<strong><code>peptide</code></strong>)</p>
</blockquote>
<p>Swaps a terminal K or R. Note: Only if AA is not modified.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="swap_AL" class="doc_header"><code>swap_AL</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L154" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>swap_AL</code>(<strong><code>peptide</code></strong>)</p>
</blockquote>
<p>Swaps a A with L. Note: Only if AA is not modified.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_decoys" class="doc_header"><code>get_decoys</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L173" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_decoys</code>(<strong><code>peptide_list</code></strong>)</p>
</blockquote>
<p>Wrapper to get decoys for lists of peptides</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_decoy_tag" class="doc_header"><code>add_decoy_tag</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L181" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_decoy_tag</code>(<strong><code>peptides</code></strong>)</p>
</blockquote>
<p>Adds a _decoy tag to a list of peptides</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">swap_AL</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;KKKALKKK&#39;</span><span class="p">)))</span>
<span class="nb">print</span><span class="p">(</span><span class="n">swap_KR</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;AAAKRAAA&#39;</span><span class="p">)))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>[K, K, K, L, A, K, K, K]
[A, A, A, K, R, A, A, A]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">get_decoy_sequence</span><span class="p">(</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>DITPEPE
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">get_decoys</span><span class="p">([</span><span class="s1">&#39;ABC&#39;</span><span class="p">,</span><span class="s1">&#39;DEF&#39;</span><span class="p">,</span><span class="s1">&#39;GHI&#39;</span><span class="p">]))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>[&#39;BAC&#39;, &#39;EDF&#39;, &#39;HGI&#39;]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Modifications">Modifications<a class="anchor-link" href="#Modifications"> </a></h2><p>To add modifications to the peptides, we distinguish fixed and variable modifications. Additionally, we make a distinction between whether the modification is only terminal or not.</p>
<h3 id="Fixed-Modifications">Fixed Modifications<a class="anchor-link" href="#Fixed-Modifications"> </a></h3><p>Fixed modifications are implemented by passing a list with modified AAs that should be replaced. As a AA is only one letter, the remainder is the modification.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_fixed_mods" class="doc_header"><code>add_fixed_mods</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L188" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_fixed_mods</code>(<strong><code>seqs</code></strong>, <strong><code>mods_fixed</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Adds fixed modifications to sequences.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">mods_fixed</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;cC&#39;</span><span class="p">,</span><span class="s1">&#39;bB&#39;</span><span class="p">]</span>
<span class="n">peptide_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ABCDEF&#39;</span><span class="p">]</span>

<span class="n">add_fixed_mods</span><span class="p">(</span><span class="n">peptide_list</span><span class="p">,</span> <span class="n">mods_fixed</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;AbBcCDEF&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Variable-Modifications">Variable Modifications<a class="anchor-link" href="#Variable-Modifications"> </a></h3><p>To employ variable modifications, we use the function <a href="/alphapept/fasta.html#get_mod_pos"><code>get_mod_pos</code></a> that returns a list of tuples with all possible modifications when giving a dicitionary with variable modifications.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_mod_pos" class="doc_header"><code>get_mod_pos</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L200" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_mod_pos</code>(<strong><code>variable_mods_r</code></strong>, <strong><code>sequence</code></strong>)</p>
</blockquote>
<p>Returns a list with of tuples with all possibilities for modified an unmodified AAs.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">mods_variable_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;M&#39;</span><span class="p">:</span><span class="s1">&#39;oxM&#39;</span><span class="p">}</span>
<span class="n">peptide</span> <span class="o">=</span> <span class="s1">&#39;AMAMA&#39;</span>
<span class="n">get_mod_pos</span><span class="p">(</span><span class="n">mods_variable_dict</span><span class="p">,</span> <span class="n">peptide</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[(&#39;A&#39;,), (&#39;M&#39;, &#39;oxM&#39;), (&#39;A&#39;,), (&#39;M&#39;, &#39;oxM&#39;), (&#39;A&#39;,)]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>To now generate all isoforms, we employ the function <a href="/alphapept/fasta.html#get_isoforms"><code>get_isoforms</code></a> that generates all isoforms for a given peptide. As the number of isoforms can become large, we restrict it with the parameter <code>max_isoforms</code>.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_isoforms" class="doc_header"><code>get_isoforms</code><a href="__main__.py#L4" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_isoforms</code>(<strong><code>variable_mods_r</code></strong>, <strong><code>sequence</code></strong>, <strong><code>max_isoforms</code></strong>)</p>
</blockquote>
<p>Function to generate isoforms for a given peptide - returns a list of isoforms.
The original sequence is included in the list</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">mods_variable_dict</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;M&#39;</span><span class="p">:</span><span class="s1">&#39;oxM&#39;</span><span class="p">}</span>
<span class="n">peptide</span> <span class="o">=</span> <span class="s1">&#39;AMAMA&#39;</span>
<span class="n">max_isoforms</span> <span class="o">=</span> <span class="mi">1024</span>
<span class="n">get_isoforms</span><span class="p">(</span><span class="n">mods_variable_dict</span><span class="p">,</span> <span class="n">peptide</span><span class="p">,</span> <span class="n">max_isoforms</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;AMAMA&#39;, &#39;AMAoxMA&#39;, &#39;AoxMAMA&#39;, &#39;AoxMAoxMA&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Lastly, we define the wrapper <a href="/alphapept/fasta.html#add_variable_mods"><code>add_variable_mods</code></a> so that the functions can be called for lists of peptides and a list of variable modifications.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_variable_mods" class="doc_header"><code>add_variable_mods</code><a href="__main__.py#L4" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_variable_mods</code>(<strong><code>peptide_list</code></strong>, <strong><code>mods_variable</code></strong>, <strong><code>max_isoforms</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">peptide_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AMA&#39;</span><span class="p">,</span> <span class="s1">&#39;AAC&#39;</span><span class="p">]</span>
<span class="n">mods_variable</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;oxM&#39;</span><span class="p">,</span><span class="s1">&#39;amC&#39;</span><span class="p">]</span>
<span class="n">max_isoforms</span> <span class="o">=</span> <span class="mi">1024</span>

<span class="n">add_variable_mods</span><span class="p">(</span><span class="n">peptide_list</span><span class="p">,</span> <span class="n">mods_variable</span><span class="p">,</span> <span class="n">max_isoforms</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;AMA&#39;, &#39;AoxMA&#39;, &#39;AAC&#39;, &#39;AAamC&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Terminal-Modifications---Fixed">Terminal Modifications - Fixed<a class="anchor-link" href="#Terminal-Modifications---Fixed"> </a></h3><p>To handle terminal modifications, we use the following convention:</p>
<ul>
<li><code>&lt;</code> for the left side (N-terminal)</li>
<li><code>&gt;</code> for the right side (C-Terminal)</li>
</ul>
<p>Additionally, if we want to have a terminal modification on any AA we indicate this <code>^</code>.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_fixed_mod_terminal" class="doc_header"><code>add_fixed_mod_terminal</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L249" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_fixed_mod_terminal</code>(<strong><code>peptides</code></strong>, <strong><code>mod</code></strong>)</p>
</blockquote>
<p>Adds fixed terminal modifications</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_fixed_mods_terminal" class="doc_header"><code>add_fixed_mods_terminal</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L267" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_fixed_mods_terminal</code>(<strong><code>peptides</code></strong>, <strong><code>mods_fixed_terminal</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Wrapper to add fixed mods on sequences and lists of mods</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">peptide</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AMAMA&#39;</span><span class="p">]</span>

<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Any n-term modified with x (x&lt;^):&#39;</span><span class="p">,</span> <span class="n">add_fixed_mods_terminal</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&lt;^&#39;</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Any c-term modified with x (x&gt;^):&#39;</span><span class="p">,</span> <span class="n">add_fixed_mods_terminal</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&gt;^&#39;</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Only A on n-term modified with x (x&lt;A):&#39;</span><span class="p">,</span> <span class="n">add_fixed_mods_terminal</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&lt;A&#39;</span><span class="p">]))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Only A on c-term modified with x (x&lt;A):&#39;</span><span class="p">,</span> <span class="n">add_fixed_mods_terminal</span><span class="p">(</span><span class="n">peptide</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&gt;A&#39;</span><span class="p">]))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Any n-term modified with x (x&lt;^): [&#39;xAMAMA&#39;]
Any c-term modified with x (x&gt;^): [&#39;AMAMxA&#39;]
Only A on n-term modified with x (x&lt;A): [&#39;xAMAMA&#39;]
Only A on c-term modified with x (x&lt;A): [&#39;AMAMxA&#39;]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Terminal-Modifications---Variable">Terminal Modifications - Variable<a class="anchor-link" href="#Terminal-Modifications---Variable"> </a></h3><p>Lastly, to handle terminal variable modifications we use the function <a href="/alphapept/fasta.html#add_variable_mods_terminal"><code>add_variable_mods_terminal</code></a>. As the modifcation can only be at the terminal end this function only adds a peptide where the terminal end is modified.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_variable_mods_terminal" class="doc_header"><code>add_variable_mods_terminal</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L280" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_variable_mods_terminal</code>(<strong><code>peptides</code></strong>, <strong><code>mods_variable_terminal</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Function to add variable terminal modifications</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_unique_peptides" class="doc_header"><code>get_unique_peptides</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L305" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_unique_peptides</code>(<strong><code>peptides</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">peptide_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;AMAMA&#39;</span><span class="p">]</span>
<span class="n">add_variable_mods_terminal</span><span class="p">(</span><span class="n">peptide_list</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;x&lt;^&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;AMAMA&#39;, &#39;xAMAMA&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Generating-Peptides">Generating Peptides<a class="anchor-link" href="#Generating-Peptides"> </a></h3><p>Lastly we put all the functions into a wrapper <a href="/alphapept/fasta.html#generate_peptides"><code>generate_peptides</code></a>. It will accept a peptide and a dictionary with settings so that we can get all modified peptides</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="generate_peptides" class="doc_header"><code>generate_peptides</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L309" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>generate_peptides</code>(<strong><code>peptide</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Wrapper to get modified peptides from a peptide</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">kwargs</span> <span class="o">=</span> <span class="p">{}</span>

<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;protease&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;trypsin&quot;</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;num_missed_cleavages&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;min_length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">6</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;max_length&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">27</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_variable&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;oxM&quot;</span><span class="p">]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_variable_terminal&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_fixed&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;cC&quot;</span><span class="p">]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_fixed_terminal&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_fixed_terminal_prot&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;mods_variable_terminal_prot&quot;</span><span class="p">]</span>  <span class="o">=</span> <span class="p">[]</span>
<span class="n">kwargs</span><span class="p">[</span><span class="s2">&quot;max_isoforms&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1024</span>

<span class="n">generate_peptides</span><span class="p">(</span><span class="s1">&#39;PEPTIDEM&#39;</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;PEPTIDEM&#39;, &#39;PEPTIDEoxM&#39;, &#39;EDITPEPM_decoy&#39;, &#39;EDITPEPoxM_decoy&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">cleave_sequence</span><span class="p">(</span><span class="s1">&#39;PEPTIDEM&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>[&#39;PEPTIDEM&#39;]</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Mass-Calculations">Mass Calculations<a class="anchor-link" href="#Mass-Calculations"> </a></h2><p>Using the <a href="/alphapept/constants.html#mass_dict"><code>mass_dict</code></a> from <a href="/alphapept/constants.html"><code>constants</code></a> and being able to parse sequences with <a href="/alphapept/fasta.html#parse"><code>parse</code></a> one can simply look up the masses for each modified or unmodified amino acid and add everything up.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Precursor">Precursor<a class="anchor-link" href="#Precursor"> </a></h3><p>To calculate the mass of the neutral precursor we start with the mass of an $H_2O$ and add the masses of all amino acids of the sequence.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_precmass" class="doc_header"><code>get_precmass</code><a href="__main__.py#L6" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_precmass</code>(<strong><code>parsed_pep</code></strong>, <strong><code>mass_dict</code></strong>)</p>
</blockquote>
<p>Calculate the mass of the neutral precursor</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_precmass</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">),</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>799.3599642034599</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Fragments">Fragments<a class="anchor-link" href="#Fragments"> </a></h3><p>Likewise, we can calculate the masses of the fragment ions. We employ two functions: <a href="/alphapept/fasta.html#get_fragmass"><code>get_fragmass</code></a> and <a href="/alphapept/fasta.html#get_frag_dict"><code>get_frag_dict</code></a>.</p>
<p><a href="/alphapept/fasta.html#get_fragmass"><code>get_fragmass</code></a> is a fast, <code>numba</code>-compatible function that calculates the fragmasses and returns an array indicating wheter the iontype was <code>b</code> or <code>y</code>.</p>
<p><a href="/alphapept/fasta.html#get_frag_dict"><code>get_frag_dict</code></a> instead is not <code>numba</code>-compatible and hence a bit slower. It returns a dictionary with the respective ion and can be used for plotting theoretical spectra.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_fragmass" class="doc_header"><code>get_fragmass</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L357" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_fragmass</code>(<strong><code>parsed_pep</code></strong>, <strong><code>mass_dict</code></strong>)</p>
</blockquote>
<p>Calculate the masses of the fragment ions</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_fragmass</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">),</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>(array([ 98.06004033, 227.10263343, 324.15539729, 425.20307579,
        538.28713979, 653.31408289, 148.06043425, 263.08737735,
        376.17144135, 477.21911985, 574.27188371, 703.31447681]),
 array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1], dtype=int8))</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_frag_dict" class="doc_header"><code>get_frag_dict</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L387" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_frag_dict</code>(<strong><code>parsed_pep</code></strong>, <strong><code>mass_dict</code></strong>)</p>
</blockquote>
<p>Calculate the masses of the fragment ions</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">get_frag_dict</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="s1">&#39;PEPT&#39;</span><span class="p">),</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;b1&#39;: 98.06004032687,
 &#39;b2&#39;: 227.10263342686997,
 &#39;b3&#39;: 324.15539728686997,
 &#39;y1&#39;: 120.06551965033,
 &#39;y2&#39;: 217.11828351033,
 &#39;y3&#39;: 346.16087661033}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This allows us also to generate the theorteical isotopes for a fragment:</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="o">%</span><span class="k">matplotlib</span> inline
<span class="n">peptide</span> <span class="o">=</span> <span class="s1">&#39;PEPTIDE&#39;</span>

<span class="n">frag_dict</span> <span class="o">=</span> <span class="n">get_frag_dict</span><span class="p">(</span><span class="n">parse</span><span class="p">(</span><span class="n">peptide</span><span class="p">),</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">)</span>

<span class="n">db_frag</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">frag_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
<span class="n">db_int</span> <span class="o">=</span> <span class="p">[</span><span class="mi">100</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">db_frag</span><span class="p">]</span>

<span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">vlines</span><span class="p">(</span><span class="n">db_frag</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="n">db_int</span><span class="p">,</span> <span class="s2">&quot;k&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s2">&quot;DB&quot;</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.2</span><span class="p">)</span>

<span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">frag_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">text</span><span class="p">(</span><span class="n">frag_dict</span><span class="p">[</span><span class="n">_</span><span class="p">],</span> <span class="mi">104</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.8</span><span class="p">)</span>
    
<span class="n">plt</span><span class="o">.</span><span class="n">title</span><span class="p">(</span><span class="s1">&#39;Theoretical Spectrum for </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">peptide</span><span class="p">))</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s1">&#39;Mass&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s1">&#39;Intensity&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">110</span><span class="p">])</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_png output_subarea ">
<img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAmQAAAFNCAYAAACuWnPfAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO3de7xUdb3/8ddHQAEV1EDFK1pmqY+O2j5e84qhEl6KDCs7ZJaZmnZO/tTUTOroUSzLW+UlFY8Gh0zTvCWSl9LUoPKCpuQtEATMCu8ofH5/zNo60maz2ezZ3w379Xw85jEz37Vm1mc+M8Cb71qzJjITSZIklbNS6QIkSZK6OwOZJElSYQYySZKkwgxkkiRJhRnIJEmSCjOQSZIkFWYgk7q4iDgtIq4qXUdLIuKzEXFbBzxPRsT7OqKmFUVErBMRd0fESxHxvdL1SGosA5lUWES8XHdZGBGv1d3/bOn6mkXE4Co49Wwey8yrM3Nog7e7ZUTcFhF/j4h/RMSUiBjW4G3eGRFfbOQ22uBw4AWgX2Z+fVmfLCI+HxELqs/VvIj4U0QMr5btXn32Xl7ksmO1/M6IeL0aeyEiro2IQRFxS926b0bE/Lr7P66ed0ZdDc3P81JVw5SIODEiVqlb57Tquerr+Meyvn6pqzOQSYVl5mrNF+CvwH51Y1d3Vh31QauL+SUwEVgHWBs4BphXsqBO6tXGwKPZjrN3t1Lf76rP2RrAT4AJEbFWtWxm/Wexuvyu7rFHV499f/X472fmvnWf3auBMXWPPWIxNRydmasDg4CvAwcDN0dE1K3zf4vUscbS9kBa3hjIpOXDyhFxZTWzMDUimpoXRMR6EfHziJgbEU9HxDF1y1aJiB9ExMzq8oPm2Yjm2YuIOCEingcuj4iVqhmLJyPibxFR/w/23dX1P5pnT6pZl9/WbW/LiJgYES9GxOyIOKka3y4iflfNcM2KiAsiYuUlveiIGABsAlySmfOryz2Z+dtFXsNJ1czNM/WzitXr/25E/LWq58cR0adu+QHVTNG86jXvExGnA7sAF1Sv84Jq3YyIoyJiGjCtpRnD+pm1qjf3RMT3q9f9VETsVI1Pj4g5ETFqMa/7CmAUcHxVw15L+1621tfMXAhcBvQBNl3S+7DIY18Efg5stTSPa+F5XsnMO4H9gR2Bjy3L80nLOwOZtHzYHxhPbWbiBqA5JKxEbQbpQWB9YAjwtYjYu3rcycAOwNbAvwHbAafUPe+6wFrUZmMOpzb7dCCwG7Ae8HfgwmrdXavrNVqYPSEiVgduB26tHvs+YFK1eAHwn8AAav/4DgGObMPr/hvwF+CqiDgwItZpYZ11q+ddn1qIuTgiNq+WnUVtRmfrqp71gVOrercDrgT+H7W+7go8k5knA7+hmhHKzKPrtnUgsD2wRRtqp1r3IeA9wE+pvYf/XtVyCLXQt9qiD8rMz/PuGafbWfr3crGqEPlF4GVgWhtfS/NjBwAjgD8uzeMWJzP/CkymFoKlbstAJi0ffpuZN2fmAuB/qf2DDLV/3Adm5rer2aOngEuo7QYC+Czw7cyck5lzgdHA5+qedyHwrcx8IzNfA74MnJyZMzLzDeA04JNt3EU3HHg+M7+Xma9n5kuZeT9AZk7JzPsy863MfAa4iFroa1W1u24P4Bnge8CsqB3ovtkiq36zeg13ATcBn6p2gX0J+M/MfDEzXwLOqOvNYcBlmTkxMxdm5nOZ+ecllPQ/1XO91oZ+ADydmZdX79v/ARtSez/eyMzbgPnUwllbLO172ZIdquOxngc+DXw8M/9ZLVuvmsmrv6xa99jzqsc+CMwC/quNdbfFTGphstmnFqnjjg7cltQlddVjRiS92/N1t18FelchaWOqf0jrlvegNsMDtZmqZ+uWPVuNNZubma/X3d8YuC4iFtaNLaB2/NaSbAg82dKCiHg/cA7QBPSl9nfPlDY8J5k5Azi6ep4NgYupzWztWK3y98x8pe4hza9xYLWtKXWHJwW1/jTXe3NbaqgzfSnXn113+zWAzFx07F9myBZjad/LltyXmR9ZzLKZmblBK489JjMvbUOd7bE+cG/d/QmZeUiDtiV1Sc6QScu36dRmYdaou6yemc3fQpxJLWQ126gaa7boAePTgX0Xeb7emflcC+u2VMt7F7PsR8Cfgc0ysx9wErVwtFQyczq1Xaj1xy+tuchMTvNrfIFa4Nmy7rX0rw5AX1K9i3ut9ePNIbBv3di6bXgZ7bW07+VyoQrZH+ad/0RI3ZKBTFq+PQDMqw7m7hMRPSJiq4j492r5OOCUiBhYHftzKtDaOc1+DJweERsDVI87oFo2l9puscUdBH4jsG5EfK06AH31iNi+WrY6tW9GvhwRHwC+0pYXFxFrRsToiHhf9YWDAcAXgPsWWXV0RKwcEbtQ23X6s+rA9UuA70fE2tXzrV93fN1PgEMjYkj13OtXtUFtZqvVg92r3YbPAYdUff8Ciw94HWFp38suLSL6RsRuwPXUPsdLO1sprVAMZNJyrDo2aT9qB3o/TW1W6FKgf7XKf1M7YPoh4GHgD9XY4pxL7UsDt0XES9SCz/bVtl4FTgfuqY7r2WGRWl4CPlrV8zy1g8X3qBYfB3wGeIlaSPq/Nr7E+cBgal8WmAc8ArwBfL5uneepfflgJrUD4Y+oOxbsBGpfCrgvIuZVz7N5Ve8DwKHA94F/AnfxzgzUudSOnft7RJzXSn1fovalgL8BW/Lu3W4dbWnfy6W1XvzrechGdODzN7ug+mzNBn5A7Rub+1QButnIFmpZuwG1SF1GtOMUN5LUJUTE7sBVSzj2SZK6PGfIJEmSCjOQSZIkFeYuS0mSpMKcIZMkSSrMQCZJklTYcn2m/gEDBuTgwYNLlyFJkrREU6ZMeSEzB7a0bLkOZIMHD2by5Mmly5AkSVqiiHh2ccvcZSlJklSYgUySJKkwA5kkSVJhBrI22m+//XjggQdKl6E28v1advaw49nTcux9WfZ/yQxky+jNN9/k+OOPZ7/99qOpqYkpU6aULqlFEydO5Atf+AI777wzhx9+eOlyinn44Yc58sgj2XPPPdlrr7044YQTeOGFFxq+3R/84Ad8/OMfZ9ddd2XEiBHcdNNNDd9mozz11FN87nOfY4899mCPPfbgyCOP5Kmnnipd1mKdd955DBs2jF133ZXhw4dz2WWXlS6pVRdffDFNTU3LxT9e8+bNY6+99uKwww4rXUq7zZw5k6amJnbZZZe3L5deemnDt3vaaaexww47vGu7CxcuXPIDV0Cvv/46Z555JkOGDGG33XbjS1/6UumSWvTAAw/w2c9+lo985CMMGzaMiRMndujzL9ffsuwqtt56az7zmc9wwgknlC5lsfr378+nP/1pnnnmGX7/+9+XLqeYl156iU984hPsuOOO9OjRgzFjxjB69GjOP//8hm63T58+fP/732ejjTbi0Ucf5atf/SobbrghH/rQhxq63UYYOHAgZ511FoMGDSIzmTBhAieddBLjx48vXVqLDjjgAL70pS/Rp08f5syZw9FHH83gwYPZc889S5f2L2bMmMGkSZMYMGBA6VLa5LzzzmOTTTZZIYLEnXfeSY8ePTp1m//xH//BkUce2anb7IpOP/10FixYwDXXXEP//v15/PHHS5f0L5566ilOPvlkRo8ezfbbb8/LL7/MSy+91KHbcIZsKUydOpWDDjqIPfbYg9GjRzN//nx69erFZz7zGbbeeutO/8O8qCuvvJLjjz/+XWNjxozhe9/7Httttx0f/ehHGTiwxdOfrJBaer922mkn9tprL1ZddVV69+7Npz71KR588MEO2V5r/f/yl7/M4MGDWWmlldhqq63YZptteOihhzpku43UUg9XX3111ltvPSKCzGSllVZi+vTpRetsrfcbb7wxffr0eXs8IpgxY0Znl/i2lnrabMyYMRxzzDH06tWrWH3NWuspwEMPPcSTTz7JfvvtV6K8dmmt942wpB52Ny31/9lnn+Wuu+7i5JNPZs0112SllVbigx/8YJH6Wnu/fvKTn/CJT3yCnXbaiR49etC/f3822GCDDt2+gWwp3HLLLVxwwQVcf/31PPvss50yrb00hg0bxr333vt2al+wYAETJ05k2LBhhSsroy3v1x/+8Ac23XTTDtleW/v/xhtvMHXq1A7bbiO11sPdd9+dHXfckbPPPpsvfOELBatccu+vuOIKdtllF4YNG8Zrr73GPvvsU6zWxfX09ttvp2fPnuy8887FaqvXWk8XLlzIWWedxfHHH09EFK607Vr7PA8fPpxhw4YxevRo/vGPf3TI9pb0ubzmmmvYc889OeSQQ/j1r3/dIdvsylrq/yOPPMKgQYO46KKLGDJkCCNHjizWi9ber4cffhiAkSNHsvfee/PNb36TefPmdej2DWRLYeTIkayzzjr069ePww47jF/96lelS3qXAQMGsO2223L77bcDcO+997LGGmsU+99GaUt6v6ZNm8all17Kscce2yHba2v/zzjjDN7//vez4447dsh2G6m1Ht55553cddddHH/88Wy++eYFq1xy7z//+c9z9913c/XVV/Oxj32M1VZbrVitLfX01Vdf5cILL+S4444rVteiWuvp+PHj2WqrrZa7v1ta6v0aa6zBlVdeyY033shVV13FK6+8wimnnNIh22uthwcffDDXXXcdEydO5Ctf+QqnnXZah83Wd1Ut9X/27Nk8+eSTrLbaatx6662ccMIJfOtb3+Lpp5/u9Ppae7/mzJnDzTffzNlnn811113HG2+8wZgxYzp0+waypbDOOuu8fXvQoEHMnTu3YDUtGz58OLfccgtQ+99Id50dg9bfr+nTp3PMMcdw3HHHsc0223TYNpfU/3PPPZcnn3ySM888c7mYWVjSZ75Pnz6MGDGCU089lRdffLGzy3uXJfU+Ith8881ZZZVV+PGPf1yiRKDlnl500UUMGzaM9dZbr1hdLWmpp3PnzmX8+PEcddRRhatbei31vm/fvmyxxRb06NGDtdZaixNOOIH77ruPV155pUO2ubjP5Qc+8AH69+9Pjx492Hnnndlnn31W+Fmylvrfu3dvevbsyWGHHUavXr3YdtttaWpq4r777itS4+Ler1VWWYX99tuPjTbaiL59+3LooYdyzz33dOi2DWRLYfbs2W/ffv7557vk8Vi7774706ZN48knn+Q3v/kN++67b+mSilnc+zVr1iyOPPJIvvjFL3Z4YG2t/xdddBH33HMPF154IauuumqHbrdR2vKZz0xef/314v9Baetnf8GCBUWPIWupp7///e8ZP348Q4cOZejQocyePZsTTzyRsWPHFqsTWu7p1KlTeeGFF/jkJz/J0KFD+e53v8vUqVMZOnRolz+4vy2f5+b/KGVmh2yzrZ/L5mMyV2Qt9f9973tfwYr+1eLer80226zh/4k2kC2FCRMmMGfOHObNm8dll13G0KFDAZg/f/7bB4e++eabzJ8/v9gfrJVXXpkhQ4Zw8skns+WWW7LuuusCsHDhQubPn8+CBQvITObPn89bb71VpMbO0tL7NWfOHI444ggOOuggRowY0eHbXFz/L7/8cm699VZ++MMf0r9//w7fbqO01MP777+fxx9/nIULF/LKK69wzjnn0K9fPzbZZJOitbbU+4ULF3Lttdcyb948MpOpU6cyYcIEtttuu2J1ttTTH/3oR0yYMIFx48Yxbtw4BgwYwEknncRBBx1UrE5ouac77bQTv/zlL9+u9YgjjmDzzTdn3LhxrLRS1/4npaXeP/LIIzz77LMsXLiQf/7zn5x99tl8+MMf7rDd2ov7O2HSpEm8+uqrLFy4kPvuu49bbrmF3XbbrUO22VW11P9tt92Wddddl8svv5wFCxbw4IMPMmXKlGKHdCzu/dp///355S9/yXPPPcfrr7/O2LFj2WWXXTp02572Yinss88+HHXUUcydO5fddtvt7XPvjBgxglmzZgFw9NFHA3DDDTcU2/0wfPhwfvGLX3Dqqae+PXbTTTcxevTot+/vtNNODB8+nNNOO61AhZ2jpffryiuv5LnnnuOSSy7hkksueXvd3/zmNx223Zb6f+GFF9KrVy8+/vGPvz126KGHFj8Yfkla6uFvf/tbxowZw5w5c1hllVXYYostOP/881l55ZVLl9ti7++44w4uuOAC3nzzTQYOHMjIkSMZOXJksRpb6mnv3r3ftU6PHj3o168fffv2LVTlOxbt6corr8x73vOet5evttpq9OzZ811jXVVLvb/rrru48MILefHFF1l11VXZfvvtOeOMMzp0uy19LseNG8e3v/1tMpP111+fU045hQ9/+MMdut2upqX+9+zZk3POOYfvfOc7XHHFFQwaNIjRo0czePDgYnW29H7tv//+zJo1i1GjRgG1f0M7+pjPWJ6nSJuamnLy5Mmly+hynn/+eUaMGMFtt9223OwaW5HY/3Lsfcezp8vOHi5fGvl+RcSUzGxqaVnXnl/WUlu4cCFXX301e++9t3/wC7D/5dj7jmdPl509XL6UfL/cZbkCee211xg6dCiDBg1q+Jnn9a/sfzn2vuPZ02VnD5cvpd8vd1lKkiR1AndZSpIkdWEGMkmSpMIaFsgi4rKImBMRj9SNrRUREyNiWnW9Zt2yb0TEXyLi8YjYu1F1SZIkdTWNnCG7Alj0F3xPBCZl5mbApOo+EbEFcDCwZfWYH0ZEjwbWJkmS1GU0LJBl5t3Aoj9udwDQ/FsgY4ED68bHZ+Ybmfk08Beg3Km0JUmSOlFnH0O2TmbOAqiu167G1wem1603oxqTJEla4XWVg/pb+sXOFs/HERGHR8TkiJjcGT9m/MQTT/DEE080fDvLanmps9FK9WFF6v/y9lqWh3qXhxqbLU+1toV/J5S1vPShK9TZ2YFsdkQMAqiu51TjM4AN69bbAJjZ0hNk5sWZ2ZSZTQMHDmxosZIkSZ2hswPZDcCo6vYo4Pq68YMjYpWI2ATYDHigk2uTJEkqomE/nRQR44DdgQERMQP4FnAmMCEiDgP+ChwEkJlTI2IC8CjwFnBUZi5oVG2SJEldScMCWWZ+ejGLhixm/dOB0xtVjyRJUlfVVQ7qlyRJ6rYMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSrMQCZJklSYgUySJKkwA5kkSVJhBjJJkqTCDGSSJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmSZJUmIFMkiSpMAOZJElSYQYySZKkwgxkkiRJhRnIJEmSCjOQSZIkFWYgkyRJKsxAJkmSVJiBTJIkqTADmSRJUmEGMkmSpMIMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSrMQCZJklSYgUySJKkwA5kkSVJhBjJJkqTCDGSSJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmSZJUmIFMkiSpMAOZJElSYUUCWUT8Z0RMjYhHImJcRPSOiLUiYmJETKuu1yxRmyRJUmfr9EAWEesDxwBNmbkV0AM4GDgRmJSZmwGTqvuSJEkrvFK7LHsCfSKiJ9AXmAkcAIytlo8FDixUmyRJUqfq9ECWmc8B3wX+CswC/pmZtwHrZOasap1ZwNqdXZskSVIJJXZZrkltNmwTYD1g1Yg4ZCkef3hETI6IyXPnzm1UmZIkSZ2mxC7LvYCnM3NuZr4JXAvsBMyOiEEA1fWclh6cmRdnZlNmNg0cOLDTipYkSWqUEoHsr8AOEdE3IgIYAjwG3ACMqtYZBVxfoDZJkqRO17OzN5iZ90fENcAfgLeAPwIXA6sBEyLiMGqh7aDOrk2SJKmETg9kAJn5LeBbiwy/QW22TJIkqVvxTP2SJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmSZJUmIFMkiSpMAOZJElSYQYySZKkwgxkkiRJhRnIJEmSCjOQSZIkFWYgkyRJKsxAJkmSVJiBTJIkqTADmSRJUmEGMkmSpMIMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSrMQCZJklSYgUySJKkwA5kkSVJhBjJJkqTCDGSSJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmSZJUmIFMkiSpMAOZJElSYQYySZKkwgxkkiRJhRnIJEmSCjOQSZIkFWYgkyRJKsxAJkmSVFibAllErNXoQiRJkrqrts6Q3R8RP4uIYRERDa1IkiSpm2lrIHs/cDHwOeAvEXFGRLy/cWVJkiR1H20KZFkzMTM/DXwRGAU8EBF3RcSODa1QkiRpBdfWY8jeExHHRsRk4Djgq8AA4OvAT5d2oxGxRkRcExF/jojHImLHiFgrIiZGxLTqes2lfV5JkqTlUVt3Wf4O6AccmJkfy8xrM/OtzJwM/Lgd2z0XuDUzPwD8G/AYcCIwKTM3AyZV9yVJklZ4bQ1kp2TmdzJzRvNARBwEkJlnLc0GI6IfsCvwk+rx8zPzH8ABwNhqtbHAgUvzvJIkScurtgaylmarvtHObW4KzAUuj4g/RsSlEbEqsE5mzgKortdu5/NLkiQtV3q2tjAi9gWGAetHxHl1i/oBby3DNrcFvpqZ90fEuSzF7smIOBw4HGCjjTZqZwmSJEldx5JmyGYCk4HXgSl1lxuAvdu5zRnAjMy8v7p/DbWANjsiBgFU13NaenBmXpyZTZnZNHDgwHaWIEmS1HW0OkOWmQ8CD0bE1ZnZ3hmxRZ/z+YiYHhGbZ+bjwBDg0eoyCjizur6+I7YnSZLU1S1pl+WEzPwU8MeIyPpF1E5P9qF2bverwNURsTLwFHAotdm6CRFxGPBX4KB2PrckSdJypdVABhxbXQ/vyI1m5p+AphYWDenI7UiSJC0PWj2GrPlbj8ALwPTMfBZYhdq5w2Y2uDZJkqRuoa2nvbgb6B0R61M7aeuhwBWNKkqSJKk7aWsgi8x8FfgEcH5mfhzYonFlSZIkdR9tDmTVj4h/FripGlvS8WeSJElqg7YGsmOpnZn/usycGhGbAnc0rixJkqTuo02zXJl5N7XjyJrvPwUc06iiJEmSupM2BbKIeD9wHDC4/jGZuWdjypIkSeo+2noc2M+AHwOXAgsaV44kSVL309ZA9lZm/qihlUiSJHVTbT2o/5cRcWREDIqItZovDa1MkiSpm2jrDNmo6vr/1Y0lsGnHliNJktT9tPVblps0uhBJkqTuqk27LCOib0ScEhEXV/c3i4gO/cFxSZKk7qqtx5BdDswHdqruzwD+uyEVSZIkdTNtDWTvzcwxwJsAmfkaEA2rSpIkqRtpayCbHxF9qB3IT0S8F3ijYVVJkiR1I239luVpwK3AhhFxNbAzcGijipIkSepO2voty9siYgqwA7Vdlcdm5gsNrUySJKmbaOu3LCdl5t8y86bMvDEzX4iISY0uTpIkqTtodYYsInoDfYEBEbEm7xzI3w9Yr8G1SZIkdQtL2mX5ZeBr1MLXFN4JZPOACxtYlyRJUrfRaiDLzHOBcyPiq5l5fifVJEmS1K209aD+8yNiJ2Bw/WMy88oG1SVJktRttCmQRcT/Au8F/gQsqIYTMJBJkiQto7aeh6wJ2CIzs5HFSJIkdUdtPVP/I8C6jSxEkiSpu2rrDNkA4NGIeIC6n0zKzP0bUpUkSVI3sjQ/nSRJkqQGaOu3LO9qdCGSJEnd1ZLO1P8StW9T/ssiIDOzX0OqkiRJ6kaWdGLY1TurEEmSpO6qrd+ylCRJUoMYyCRJkgozkEmSJBVmIJMkSSrMQCZJklSYgUySJKkwA5kkSVJhBjJJkqTCDGSSJEmFFQtkEdEjIv4YETdW99eKiIkRMa26XrNUbZIkSZ2p5AzZscBjdfdPBCZl5mbApOq+JEnSCq9IIIuIDYCPAZfWDR8AjK1ujwUO7Oy6JEmSSig1Q/YD4HhgYd3YOpk5C6C6XrtEYZIkSZ2t0wNZRAwH5mTmlHY+/vCImBwRk+fOndvB1UmSJHW+EjNkOwP7R8QzwHhgz4i4CpgdEYMAqus5LT04My/OzKbMbBo4cGBn1SxJktQwnR7IMvMbmblBZg4GDgZ+nZmHADcAo6rVRgHXd3ZtkiRJJXSl85CdCXw0IqYBH63uS5IkrfB6ltx4Zt4J3Fnd/hswpGQ9kiRJJXSlGTJJkqRuyUAmSZJUmIFMkiSpMAOZJElSYQYySZKkwgxkkiRJhRnIJEmSCjOQSZIkFWYgkyRJKsxAJkmSVJiBTJIkqTADmSRJUmEGMkmSpMIMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSrMQCZJklSYgUySJKkwA5kkSVJhBjJJkqTCDGSSJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmSZJUmIFMkiSpMAOZJElSYQYySZKkwgxkkiRJhRnIJEmSCjOQSZIkFWYgkyRJKsxAJkmSVJiBTJIkqTADmSRJUmEGMkmSpMIMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSqs0wNZRGwYEXdExGMRMTUijq3G14qIiRExrbpes7NrkyRJKqHEDNlbwNcz84PADsBREbEFcCIwKTM3AyZV9yVJklZ4nR7IMnNWZv6huv0S8BiwPnAAMLZabSxwYGfXJkmSVELRY8giYjCwDXA/sE5mzoJaaAPWXsxjDo+IyRExee7cuZ1VqiRJUsMUC2QRsRrwc+BrmTmvrY/LzIszsykzmwYOHNi4AiVJkjpJkUAWEb2ohbGrM/Paanh2RAyqlg8C5pSoTZIkqbOV+JZlAD8BHsvMc+oW3QCMqm6PAq7v7NokSZJK6FlgmzsDnwMejog/VWMnAWcCEyLiMOCvwEEFapMkSep0nR7IMvO3QCxm8ZDOrEWSJKkr8Ez9kiRJhRnIJEmSCjOQSZIkFWYgkyRJKsxAJkmSVJiBTJIkqTADmSRJUmEGMkmSpMIMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSrMQCZJklSYgUySJKkwA5kkSVJhBjJJkqTCDGSSJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmSZJUmIFMkiSpMAOZJElSYQYySZKkwgxkkiRJhRnIJEmSCjOQSZIkFWYgkyRJKsxAJkmSVJiBTJIkqTADmSRJUmEGMkmSpMIMZJIkSYUZyCRJkgozkEmSJBVmIJMkSSrMQCZJklSYgUySJKkwA5kkSVJhBjJJkqTCulwgiyA305AAAAZsSURBVIh9IuLxiPhLRJxYuh5JkqRG61KBLCJ6ABcC+wJbAJ+OiC3KViVJktRYXSqQAdsBf8nMpzJzPjAeOKBwTZIkSQ3V1QLZ+sD0uvszqjFJkqQVVs/SBSwiWhjLd60QcThweHX35Yh4vOFV1QwAXuikbXUH9rNj2c+OZT87lv3sOPayY3V2Pzde3IKuFshmABvW3d8AmFm/QmZeDFzcmUUBRMTkzGzq7O2uqOxnx7KfHct+diz72XHsZcfqSv3sarssfw9sFhGbRMTKwMHADYVrkiRJaqguNUOWmW9FxNHAr4AewGWZObVwWZIkSQ3VpQIZQGbeDNxcuo4WdPpu0hWc/exY9rNj2c+OZT87jr3sWF2mn5GZS15LkiRJDdPVjiGTJEnqdgxklYi4LCLmRMQjdWNrRcTEiJhWXa9Zt+wb1c87PR4Re5epumuKiA0j4o6IeCwipkbEsdW4/WyHiOgdEQ9ExINVP0dX4/ZzGUREj4j4Y0TcWN23n+0UEc9ExMMR8aeImFyN2c92iog1IuKaiPhz9ffojvZz6UXE5tVnsvkyLyK+1mV7mZlearttdwW2BR6pGxsDnFjdPhE4q7q9BfAgsAqwCfAk0KP0a+gqF2AQsG11e3Xgiapn9rN9/Qxgtep2L+B+YAf7ucx9/S/gp8CN1X372f5ePgMMWGTMfra/n2OBL1a3VwbWsJ/L3NMewPPUzgPWJXvpDFklM+8GXlxk+ABqfzCorg+sGx+fmW9k5tPAX6j97JOAzJyVmX+obr8EPEbtFxfsZztkzcvV3V7VJbGf7RYRGwAfAy6tG7afHct+tkNE9KM2QfATgMycn5n/wH4uqyHAk5n5LF20lway1q2TmbOgFjKAtatxf+KpjSJiMLANtVkd+9lO1e61PwFzgImZaT+XzQ+A44GFdWP2s/0SuC0iplS/pgL2s702BeYCl1e71C+NiFWxn8vqYGBcdbtL9tJA1j5L/IknQUSsBvwc+Fpmzmtt1RbG7GedzFyQmVtT+/WK7SJiq1ZWt5+tiIjhwJzMnNLWh7QwZj/fbefM3BbYFzgqInZtZV372bqe1A6f+VFmbgO8Qm232uLYzyWoTjS/P/CzJa3awlin9dJA1rrZETEIoLqeU40v8SeeuruI6EUtjF2dmddWw/ZzGVW7Lu4E9sF+ttfOwP4R8QwwHtgzIq7CfrZbZs6srucA11HbzWM/22cGMKOaBQe4hlpAs5/tty/wh8ycXd3vkr00kLXuBmBUdXsUcH3d+MERsUpEbAJsBjxQoL4uKSKC2vEPj2XmOXWL7Gc7RMTAiFijut0H2Av4M/azXTLzG5m5QWYOprYb49eZeQj2s10iYtWIWL35NjAUeAT72S6Z+TwwPSI2r4aGAI9iP5fFp3lndyV01V6W/uZDV7lQe7NmAW9SS8mHAe8BJgHTquu16tY/mdo3MB4H9i1df1e6AB+hNs37EPCn6jLMfra7nx8C/lj18xHg1Grcfi57b3fnnW9Z2s/29XBTat9MexCYCpxsP5e5p1sDk6s/878A1rSf7e5lX+BvQP+6sS7ZS8/UL0mSVJi7LCVJkgozkEmSJBVmIJMkSSrMQCZJklSYgUySJKkwA5mkbiEiMiL+t+5+z4iYGxE3lqxLksBAJqn7eAXYqjq5LsBHgecK1iNJbzOQSepObgE+Vt1+19m7I2K7iLi3+kHne5vPlB4RW0bEAxHxp4h4KCI2q85Of1NEPBgRj0TEyAKvRdIKxEAmqTsZT+2nUXpT+wWE++uW/RnYNWs/6HwqcEY1fgRwbtZ+3L2J2i957APMzMx/y8ytgFs76wVIWjH1LF2AJHWWzHwoIgZTmx27eZHF/YGxEbEZtZ/+6lWN/w44OSI2AK7NzGkR8TDw3Yg4i9pPL/2mU16ApBWWM2SSupsbgO/y7h8bBvgOcEc147Uf0BsgM38K7A+8BvwqIvbMzCeADwMPA/8TEad2VvGSVkzOkEnqbi4D/pmZD0fE7nXj/XnnIP/PNw9GxKbAU5l5XnX7QxHxZ+DFzLwqIl6uX1+S2sNAJqlbycwZwLktLBpDbZflfwG/rhsfCRwSEW8CzwPfBv4dODsiFgJvAl9pbNWSVnSRmaVrkCRJ6tY8hkySJKkwA5kkSVJhBjJJkqTCDGSSJEmFGcgkSZIKM5BJkiQVZiCTJEkqzEAmSZJU2P8HEHgsdkJfrskAAAAASUVORK5CYII=
"
>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Spectra">Spectra<a class="anchor-link" href="#Spectra"> </a></h3><p>The function <a href="/alphapept/fasta.html#get_spectrum"><code>get_spectrum</code></a> returns a tuple with the following content:</p>
<ul>
<li>precursor mass</li>
<li>peptide sequence</li>
<li>fragmasses</li>
<li>fragtypes</li>
</ul>
<p>Likewise, <a href="/alphapept/fasta.html#get_spectra"><code>get_spectra</code></a> returns a list of tuples. We employ a list of tuples here as this way we can sort them easily by precursor mass.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_spectrum" class="doc_header"><code>get_spectrum</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L416" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_spectrum</code>(<strong><code>peptide</code></strong>, <strong><code>mass_dict</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_spectra" class="doc_header"><code>get_spectra</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L429" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_spectra</code>(<strong><code>peptides</code></strong>, <strong><code>mass_dict</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">get_spectra</span><span class="p">(</span><span class="n">List</span><span class="p">([</span><span class="s1">&#39;PEPTIDE&#39;</span><span class="p">]),</span> <span class="n">constants</span><span class="o">.</span><span class="n">mass_dict</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>[(799.3599642034599, &#39;PEPTIDE&#39;, array([ 98.06004033, 148.06043425, 227.10263343, 263.08737735,
       324.15539729, 376.17144135, 425.20307579, 477.21911985,
       538.28713979, 574.27188371, 653.31408289, 703.31447681]), array([0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1], dtype=int8))]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Reading-FASTA">Reading FASTA<a class="anchor-link" href="#Reading-FASTA"> </a></h2><p>To read FASTA files we use the <code>SeqIO</code> module from the <code>Biopython</code> library. This is a generator expression so that we read one FASTA entry after another until the <code>StopIteration</code> is reached, which is implemented in <a href="/alphapept/fasta.html#read_fasta_file"><code>read_fasta_file</code></a>. Additionally, we define the funciton <a href="/alphapept/fasta.html#read_fasta_file_entries"><code>read_fasta_file_entries</code></a> that simply counts the number of FASTA entries.</p>
<p>All FASTA entries that contain AAs which are not in the mass_dict can be checked with <a href="/alphapept/fasta.html#check_sequence"><code>check_sequence</code></a> and will be ignored.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="read_fasta_file" class="doc_header"><code>read_fasta_file</code><a href="__main__.py#L7" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>read_fasta_file</code>(<strong><code>fasta_filename</code></strong>=<em><code>''</code></em>)</p>
</blockquote>
<p>Read a FASTA file line by line</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="read_fasta_file_entries" class="doc_header"><code>read_fasta_file_entries</code><a href="__main__.py#L34" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>read_fasta_file_entries</code>(<strong><code>fasta_filename</code></strong>=<em><code>''</code></em>)</p>
</blockquote>
<p>Function to count entries in fasta file</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="check_sequence" class="doc_header"><code>check_sequence</code><a href="__main__.py#L51" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>check_sequence</code>(<strong><code>element</code></strong>, <strong><code>AAs</code></strong>)</p>
</blockquote>
<p>Checks wheter a sequence from a FASTA entry contains valid AAs</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">fasta_path</span> <span class="o">=</span> <span class="s1">&#39;../testfiles/test.fasta&#39;</span>

<span class="nb">list</span><span class="p">(</span><span class="n">read_fasta_file</span><span class="p">(</span><span class="n">fasta_path</span><span class="p">))[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;id&#39;: &#39;A0PJZ0&#39;,
 &#39;name&#39;: &#39;sp|A0PJZ0|A20A5_HUMAN&#39;,
 &#39;description&#39;: &#39;sp|A0PJZ0|A20A5_HUMAN Putative ankyrin repeat domain-containing protein 20A5 OS=Homo sapiens OX=9606 GN=ANKRD20A5P PE=5 SV=1&#39;,
 &#39;sequence&#39;: &#39;MKLFGFRSRRGQTVLGSIDHLYTGSGYRIRYSELQKIHKAAVKGDAAEMERCLARRSGDLDALDKQHRTALHLACASGHVKVVTLLVNRKCQIDIYDKENRTPLIQAVHCQEEACAVILLEHGANPNLKDIYGNTALHYAVYSESTSLAEKLLFHGENIEALDKV&#39;}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Peptide-Dictionary">Peptide Dictionary<a class="anchor-link" href="#Peptide-Dictionary"> </a></h2><p>In order to efficiently store peptides, we rely on the Python dictionary. The idea is to have a dictionary with peptides as keys and indices to proteins as values. This way, one can quickly look up to which protein a peptide belongs to. The function <a href="/alphapept/fasta.html#add_to_pept_dict"><code>add_to_pept_dict</code></a> uses a regular python dictionary and allows to add peptides and stores indices to the originating proteins as a list. If a peptide is already present in the dictionary, the list is appended. The function returns a list of <code>added_peptides</code>, which were not present in the dictionary yet. One can use the function <a href="/alphapept/fasta.html#merge_pept_dicts"><code>merge_pept_dicts</code></a> to merge multiple peptide dicts.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_to_pept_dict" class="doc_header"><code>add_to_pept_dict</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L499" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_to_pept_dict</code>(<strong><code>pept_dict</code></strong>, <strong><code>new_peptides</code></strong>, <strong><code>i</code></strong>)</p>
</blockquote>
<p>Add peptides to the peptide dictionary</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">pept_dict</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">new_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ABC&#39;</span><span class="p">,</span><span class="s1">&#39;DEF&#39;</span><span class="p">]</span>

<span class="n">pept_dict</span><span class="p">,</span> <span class="n">added_peptides</span> <span class="o">=</span> <span class="n">add_to_pept_dict</span><span class="p">(</span><span class="n">pept_dict</span><span class="p">,</span> <span class="n">new_peptides</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>

<span class="n">new_peptides</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;DEF&#39;</span><span class="p">,</span><span class="s1">&#39;GHI&#39;</span><span class="p">]</span>

<span class="n">pept_dict</span><span class="p">,</span> <span class="n">added_peptides</span> <span class="o">=</span> <span class="n">add_to_pept_dict</span><span class="p">(</span><span class="n">pept_dict</span><span class="p">,</span> <span class="n">new_peptides</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">pept_dict</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>{&#39;ABC&#39;: [0], &#39;DEF&#39;: [0, 1], &#39;GHI&#39;: [1]}
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="merge_pept_dicts" class="doc_header"><code>merge_pept_dicts</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L515" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>merge_pept_dicts</code>(<strong><code>list_of_pept_dicts</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">pept_dict_1</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ABC&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;DEF&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;GHI&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">]}</span>
<span class="n">pept_dict_2</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ABC&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="s1">&#39;JKL&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span> <span class="s1">&#39;MNO&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">7</span><span class="p">]}</span>

<span class="n">merge_pept_dicts</span><span class="p">([</span><span class="n">pept_dict_1</span><span class="p">,</span> <span class="n">pept_dict_2</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;ABC&#39;: [0, 3, 4], &#39;DEF&#39;: [0, 1], &#39;GHI&#39;: [1], &#39;JKL&#39;: [5, 6], &#39;MNO&#39;: [7]}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">def</span> <span class="nf">test_merge_pept_dicts</span><span class="p">():</span>
    <span class="n">pept_dict_1</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ABC&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s1">&#39;DEF&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;GHI&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">]}</span>
    <span class="n">pept_dict_2</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;ABC&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">],</span> <span class="s1">&#39;JKL&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span> <span class="s1">&#39;MNO&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">7</span><span class="p">]}</span>

    <span class="k">assert</span> <span class="n">merge_pept_dicts</span><span class="p">([</span><span class="n">pept_dict_1</span><span class="p">,</span> <span class="n">pept_dict_2</span><span class="p">])</span> <span class="o">==</span> <span class="p">{</span><span class="s1">&#39;ABC&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span> <span class="s1">&#39;DEF&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;GHI&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s1">&#39;JKL&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">],</span> <span class="s1">&#39;MNO&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">7</span><span class="p">]}</span>
    
<span class="n">test_merge_pept_dicts</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Generating-a-database">Generating a database<a class="anchor-link" href="#Generating-a-database"> </a></h2><p>To wrap everything up, we employ two functions <a href="/alphapept/fasta.html#generate_database"><code>generate_database</code></a> and <a href="/alphapept/fasta.html#generate_spectra"><code>generate_spectra</code></a>. The first one reads a FASTA file and generates a list of peptides, as well as the peptide dictionary and an ordered FASTA dictionary to be able to look up the protein indices laster. For the <code>callback</code> we first read the whole FASTA file to determine the total number of entries in the FASTA file.  For a typical FASTA file of 30 Mb with 40k entries, this should take less than a second. The progress of the digestion is monitored by processing the FASTA file one by one.
The function <a href="/alphapept/fasta.html#generate_spectra"><code>generate_spectra</code></a> then calculates precursor masses and fragment ions. Here, we split the total_number of sequences in <code>1000</code> steps to be able to track progress with the <code>callback</code>.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="generate_fasta_list" class="doc_header"><code>generate_fasta_list</code><a href="__main__.py#L4" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>generate_fasta_list</code>(<strong><code>fasta_paths</code></strong>, <strong><code>callback</code></strong>=<em><code>None</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Function to generate a database from a fasta file</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="generate_database" class="doc_header"><code>generate_database</code><a href="__main__.py#L35" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>generate_database</code>(<strong><code>mass_dict</code></strong>, <strong><code>fasta_paths</code></strong>, <strong><code>callback</code></strong>=<em><code>None</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Function to generate a database from a fasta file</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="generate_spectra" class="doc_header"><code>generate_spectra</code><a href="__main__.py#L73" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>generate_spectra</code>(<strong><code>to_add</code></strong>, <strong><code>mass_dict</code></strong>, <strong><code>callback</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Function to generate a database from a fasta file</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Parallelized-version">Parallelized version<a class="anchor-link" href="#Parallelized-version"> </a></h2><p>To speed up spectra generated, one can use the parallelized version. The function <a href="/alphapept/fasta.html#generate_database_parallel"><code>generate_database_parallel</code></a> reads an entire FASTA file and splits it into multiple blocks. Each block will be processed, and the generated pept_dicts will be merged.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="block_idx" class="doc_header"><code>block_idx</code><a href="__main__.py#L6" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>block_idx</code>(<strong><code>len_list</code></strong>, <strong><code>block_size</code></strong>=<em><code>1000</code></em>)</p>
</blockquote>
<p>Create indices for a list of length len_list</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="blocks" class="doc_header"><code>blocks</code><a href="__main__.py#L21" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>blocks</code>(<strong><code>l</code></strong>, <strong><code>n</code></strong>)</p>
</blockquote>
<p>Create blocks from a given list</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="digest_fasta_block" class="doc_header"><code>digest_fasta_block</code><a href="__main__.py#L28" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>digest_fasta_block</code>(<strong><code>to_process</code></strong>)</p>
</blockquote>
<p>Digest and create spectra for a whole fasta_block</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="generate_database_parallel" class="doc_header"><code>generate_database_parallel</code><a href="__main__.py#L56" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>generate_database_parallel</code>(<strong><code>settings</code></strong>, <strong><code>callback</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Function to generate a database from a fasta file</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Parallel-search-on-large-files">Parallel search on large files<a class="anchor-link" href="#Parallel-search-on-large-files"> </a></h3><p>In some cases (i.e., a lot of modifications), it will not be useful to save the database as it will consume too much memory. Here, we use the function <a href="/alphapept/search.html#search_parallel"><code>search_parallel</code></a> from search. It creates theoretical spectra on the fly and directly searches against them. As we cannot create a pept_dict here, we need to create one from the search results. For this, we group peptides by their FASTA index and generate a lookup dictionary that can be used as a pept_dict.</p>
<blockquote><p>Note that we are passing the settings argument here. Search results should be stored in the corresponding path in the hdf file.</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="pept_dict_from_search" class="doc_header"><code>pept_dict_from_search</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/fasta.py#L715" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>pept_dict_from_search</code>(<strong><code>settings</code></strong>)</p>
</blockquote>
<p>Generates a peptide dict from a large search</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Saving">Saving<a class="anchor-link" href="#Saving"> </a></h2><p>To save the generated spectra, we rely on NumPy's NPZ format. For this, we create a dictionary and save all the generated elements. The container will contain the following elements:</p>
<ul>
<li><code>precursors</code>: An array containing the precursor masses</li>
<li><code>seqs</code>: An array containing the peptide sequences for the precursor masses</li>
<li><code>pept_dict</code>: A peptide dictionary to look up the peptides and return their FASTA index</li>
<li><code>fasta_dict</code>: A fasta dictionary to look up the FASTA entry based on a pept_dict index</li>
<li><code>fragmasses</code>: An array containing the fragment masses. Unoccupied cells are filled with -1</li>
<li><code>fragtypes:</code>: An array containg the fragment types. 0 equals b-ions and 1 equals y-ions. Unoccupied cells are filled with -1</li>
<li><code>bounds</code>: An integer array containing the upper bounds for the fragment masses / types array. This is needed to quickly slice the data.</li>
</ul>
<p>All arrays are sorted according to the precursor mass.</p>
<blockquote><p>To access the dictionaries such as <code>pept_dict</code> or <code>fasta_dict</code>, one needs to extract them using the <code>.item()</code> method like so:<code>container['pept_dict'].item()</code>.</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="save_database" class="doc_header"><code>save_database</code><a href="__main__.py#L5" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>save_database</code>(<strong><code>spectra</code></strong>, <strong><code>pept_dict</code></strong>, <strong><code>fasta_dict</code></strong>, <strong><code>database_path</code></strong>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Function to save a database to the *.npz format.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

