---

title: Feature Finding


keywords: fastai
sidebar: home_sidebar

summary: "Functions related to feature finding"
description: "Functions related to feature finding"
nb_path: "nbs\04_feature_finding.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbs\04_feature_finding.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This part describes the implementation of the feature finding algorithm. The core of the algorithm is described in the <a href="https://www.nature.com/articles/nbt.1511">MaxQuant-Paper</a>.
The supplementary material explains the underlying methodology in great detail and is the foundation of the theoretical background that is described here.
A refined version of the algorithm was presented with <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4933939/">Dinosaur</a>, which was also used as a reference for the Python implementation.</p>
<p>For the algorithm, we need serval modules:</p>
<ol>
<li><p>Connecting Centroids to Hills</p>
</li>
<li><p>Refinement of Hills</p>
</li>
<li><p>Calculating Hill Statistics</p>
</li>
<li><p>Combining Hills to Isotope Patterns</p>
</li>
<li><p>Deconvolution of Isotope Patterns</p>
</li>
</ol>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Loading-Data">Loading Data<a class="anchor-link" href="#Loading-Data"> </a></h2><p>From the <code>IO</code> library, we already have <code>*.ms_data.hdf</code> container that contains centroided data. To use it in feature finding, we directly load the data.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Connecting-Centroids-to-Hills">Connecting Centroids to Hills<a class="anchor-link" href="#Connecting-Centroids-to-Hills"> </a></h2><h3 id="Connecting-Centroids">Connecting Centroids<a class="anchor-link" href="#Connecting-Centroids"> </a></h3><p>In this step, we search for centroids that are connected (i.e., within a specific mass difference) over time. The connected centroids are termed hills. As we have a list of centroid lists, we ideally would have a list of tuples with indices to the centroids that describe the connected hills.</p>
<h4 id="Pairs-of-Centroids">Pairs of Centroids<a class="anchor-link" href="#Pairs-of-Centroids"> </a></h4><p>First, we define the function <code>get_pairs_centroids</code> that checks two lists of centroids and checks for pairs of centroids. As the masses in the centroid lists are sorted, we can loop through the centroids with two indices and compare the mass difference. If the mass difference is within the defined threshold, <code>ppm_tol</code> (default is 7ppm), we have found ourselves a pair. If no match is found, we can advance either index, depending on whether the mass difference is positive or negative. A pair is a <code>tuple</code> containing the index to the first centroid list <code>i</code> and the second centroid list <code>j</code> as well as the difference in ppm <code>delta</code>. We add all found pairs to a list termed <code>pairs</code>.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="connect_centroids_unidirection" class="doc_header"><code>connect_centroids_unidirection</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L20" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>connect_centroids_unidirection</code>(<strong><code>idx</code></strong>, <strong><code>row_borders</code></strong>, <strong><code>connections</code></strong>, <strong><code>scores</code></strong>, <strong><code>centroids</code></strong>, <strong><code>max_gap</code></strong>, <strong><code>ppm_tol</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="convert_connections_to_array" class="doc_header"><code>convert_connections_to_array</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L61" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>convert_connections_to_array</code>(<strong><code>idx</code></strong>, <strong><code>from_r</code></strong>, <strong><code>from_c</code></strong>, <strong><code>to_r</code></strong>, <strong><code>to_c</code></strong>, <strong><code>row_borders</code></strong>, <strong><code>out_from_idx</code></strong>, <strong><code>out_to_idx</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="eliminate_overarching_vertex" class="doc_header"><code>eliminate_overarching_vertex</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L82" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>eliminate_overarching_vertex</code>(<strong><code>idx</code></strong>, <strong><code>from_idx</code></strong>, <strong><code>to_idx</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="path_finder" class="doc_header"><code>path_finder</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L94" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>path_finder</code>(<strong><code>idx</code></strong>, <strong><code>from_idx</code></strong>, <strong><code>to_idx</code></strong>, <strong><code>forward</code></strong>, <strong><code>backward</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="find_path_start" class="doc_header"><code>find_path_start</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L106" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>find_path_start</code>(<strong><code>idx</code></strong>, <strong><code>forward</code></strong>, <strong><code>backward</code></strong>, <strong><code>path_starts</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="find_path_length" class="doc_header"><code>find_path_length</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L116" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>find_path_length</code>(<strong><code>idx</code></strong>, <strong><code>path_starts</code></strong>, <strong><code>forward</code></strong>, <strong><code>path_cnt</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="fill_path_matrix" class="doc_header"><code>fill_path_matrix</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L130" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>fill_path_matrix</code>(<strong><code>idx</code></strong>, <strong><code>path_start</code></strong>, <strong><code>forwards</code></strong>, <strong><code>out_hill_data</code></strong>, <strong><code>out_hill_ptr</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="convert_to_coord_path" class="doc_header"><code>convert_to_coord_path</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L144" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>convert_to_coord_path</code>(<strong><code>idx</code></strong>, <strong><code>out_rows</code></strong>, <strong><code>out_cols</code></strong>, <strong><code>connected_comps</code></strong>, <strong><code>d_width</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h4 id="Connecting-all-Centroids">Connecting all Centroids<a class="anchor-link" href="#Connecting-all-Centroids"> </a></h4><p>As we now have a function that can find pairs of centroids when given two lists of centroids, the next step is to write a function that connects all the centroids within a dataset. For this, we define the function <a href="/alphapept/feature_finding.html#connect_centroids"><code>connect_centroids</code></a>. Here, we use the <code>get_pairs_centroids</code> function to connect hills in consecutive scans <code>(n-&gt;n+1)</code> whenever their centroid m/z positions differ less than the defined threshold. If no matching peak is found in the next scan <code>(n+1)</code>, additional scans can be considered when looking in the scan after next <code>(n+2)</code>. The maximum distance between two hills can be set via the <code>gap</code> parameter.</p>
<p>For an efficient implementation to find all connected hills over all scans, we create a sparse matrix, named <code>results</code> that represent indices to all possible centroids. As the number of centroids varies for each scan, not all positions within the matrix represent a valid centroid. To distinguish these values, we fill the matrix with <code>-1</code> upon initialization. We can use the matrix to store the matching centroid as an integer that encodes its position. The x,y index for each cell represents the indices to the centroid list. The idea here is to have an efficient data structure to represent connections between centroids.</p>
<p>To calculate indexes back and forth we define the functions <code>tup_to_ind</code> and <code>ind_to_tup</code>. This could in principle be also done with <code>numpy.unravel_index</code>, however, the numba implementation seems to be much faster.</p>
<p>Additionally, we define a sparse matrix named <code>scores</code> that is intended to store the mass difference. If there are multiple possible connections between two centroids, only the best (i.e., shortest distance) will be stored.</p>
<p>The idea behind the matrix-indexing is explained with the following example:</p>
<h5 id="Example:">Example:<a class="anchor-link" href="#Example:"> </a></h5><p>Imagine you have three scans with the following centroids:</p>
<ul>
<li>Scan 0: 10, 20, 30</li>
<li>Scan 1: 10.2, 40.1</li>
<li>Scan 2: 40, 50, 60</li>
</ul>
<p>When comparing consecutive scans and defining the maximum delta mass to be 0.5 find the following connections: (Scan No, Centroid No) -&gt; (Scan No, Centroid No). As we cannot easily store tuples in the matrix, we convert tuple containing the position of the connected centroid to an integer.</p>
<ul>
<li>(0,0) -&gt; (1,0) -&gt; (3): 10 &amp; 10.2 -&gt; delta = 0.2</li>
<li>(1,1) -&gt; (2,0) -&gt; (6): 40.1 &amp; 40 -&gt; delta = 0.1</li>
</ul>
<p>Finally, we store this in the <code>results</code> matrix:</p>
<p>$\begin{bmatrix}
3 &amp; -1  &amp; -1 \\ 
-1 &amp; 6 &amp; -1\\ 
-1 &amp; -1 &amp; -1 
\end{bmatrix}$</p>
<p>The coressponding <code>scores</code> matrix will look as follows:</p>
<p>$\begin{bmatrix}
0.2 &amp; -1  &amp; -1 \\ 
-1 &amp; 0.1 &amp; -1\\ 
-1 &amp; -1 &amp; -1 
\end{bmatrix}$</p>
<p>This allows us to not only easily store connections between centroids but also perform a quick lookup for the delta of an existing connection. Note that it also only stores the best connection for each centroid. To extract the connected centroids, we can use <code>np.where(results &gt;= 0)</code>. This implementation allows getting millions of connections within seconds.</p>
<h5 id="Foward-and-Backward-Search">Foward and Backward Search<a class="anchor-link" href="#Foward-and-Backward-Search"> </a></h5><p>One thing that still needs to be considered is that when searching with a large mass tolerance, we could get multiple matches that we do not want. Consider the following example:</p>
<ul>
<li>Scan 0: 10 (a), </li>
<li>Scan 1: 10 (b) , 10.1 (c)</li>
<li>Scan 2: 10 (d)</li>
</ul>
<p>As we only allow one connection per mass, we would have the following connections:</p>
<ul>
<li>a -&gt; b</li>
<li>b -&gt; d</li>
</ul>
<p>Now if we have a large delta mass, e.g. of <code>0.1</code>, there would be an additional connection:</p>
<ul>
<li>c-&gt; d</li>
</ul>
<p><img src="/alphapept/images/undirected.png" alt="undirected"></p>
<p>The resulting hill would have a branch to the c centroid, which is obviously intended. To circumvent this, we additionally perform the search backwards. (Note the functions <code>connect_centroids_forward</code> and <code>connect_centroids_backward</code>).  Now, as we only store, the best connection between two centroids  <code>d</code> will only be connected to <code>b</code>. To, therefore, remove branches, we remove all connections that are not double.</p>
<p><img src="/alphapept/images/directed.png" alt="directed"></p>
<p>We define a wrapper function named <a href="/alphapept/feature_finding.html#connect_centroids"><code>connect_centroids</code></a> that allows to search with a gaps ize and returns lists of indices of the connected centroids.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Connecting-hills">Connecting hills<a class="anchor-link" href="#Connecting-hills"> </a></h3><p>Having all the connected centroids, we would like to find <code>hills</code>, referring to centroids that are connected. We can achieve this by creating a graph <code>G</code> with the <code>networkx</code>-package. Here, each connection is added as an edge using the <code>add_edge</code> method to the graph. We can then use the <code>connected_components</code> function to extract all the connected components.</p>
<p><code>NetworkX</code> is known to be memory <a href="https://groups.google.com/forum/#!topic/networkx-discuss/Etd4GpkjPdA">consuming</a> and can take hundreds of bytes per edge. For millions of connections, we will, therefore, consume Gigabytes of memory. To keep the graph small, we continuously feed the graph until we reach a certain scan number <code>buffer size</code>. Once the limit is reached, we extract all completed hills and remove all nodes that are part of a completed hill or cannot be connected anymore. Once all edges are added, we perform a final extraction step. Finally, everything is put into the <a href="/alphapept/feature_finding.html#get_hills"><code>get_hills</code></a> function.</p>
<h4 id="Parallelization-Strategies">Parallelization Strategies<a class="anchor-link" href="#Parallelization-Strategies"> </a></h4><p>Currently, the parallel connection of hills is not implemented yet. A possible way could be to process individual chunks of edges in parallel and then combine them together, as it is done in <code>Dinosaur</code>.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="find_centroid_connections" class="doc_header"><code>find_centroid_connections</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L160" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>find_centroid_connections</code>(<strong><code>rowwise_peaks</code></strong>, <strong><code>row_borders</code></strong>, <strong><code>centroids</code></strong>, <strong><code>max_gap</code></strong>=<em><code>2</code></em>, <strong><code>ppm_tol</code></strong>=<em><code>8</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_hills" class="doc_header"><code>get_hills</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L194" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_hills</code>(<strong><code>centroids</code></strong>, <strong><code>from_idx</code></strong>, <strong><code>to_idx</code></strong>, <strong><code>min_hill_length</code></strong>=<em><code>3</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="connect_centroids" class="doc_header"><code>connect_centroids</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L226" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>connect_centroids</code>(<strong><code>rowwise_peaks</code></strong>, <strong><code>row_borders</code></strong>, <strong><code>centroids</code></strong>, <strong><code>max_gap</code></strong>=<em><code>2</code></em>, <strong><code>ppm_tol</code></strong>=<em><code>8</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="extract_hills" class="doc_header"><code>extract_hills</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L256" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>extract_hills</code>(<strong><code>query_data</code></strong>, <strong><code>max_gap</code></strong>=<em><code>2</code></em>, <strong><code>ppm_tol</code></strong>=<em><code>8</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Hill-Splitting">Hill Splitting<a class="anchor-link" href="#Hill-Splitting"> </a></h3><p>When having a hill with two or more maxima, we would like to split it at the minimum position. For this, we use a recursive approach. First, the minimum of a hill is detected. A hill is split at this minimum if the smaller of the surrounding maxima is at least the factor <code>split_level</code> larger than the minimum. For each split, the process is repeated.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="fast_minima" class="doc_header"><code>fast_minima</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L284" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>fast_minima</code>(<strong><code>y</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="split" class="doc_header"><code>split</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L303" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>split</code>(<strong><code>idx</code></strong>, <strong><code>hill_ptrs_range</code></strong>, <strong><code>hill_ptrs</code></strong>, <strong><code>int_data</code></strong>, <strong><code>hill_data</code></strong>, <strong><code>splits</code></strong>, <strong><code>split_level</code></strong>, <strong><code>window</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="split_hills" class="doc_header"><code>split_hills</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L350" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>split_hills</code>(<strong><code>hill_ptrs</code></strong>, <strong><code>hill_data</code></strong>, <strong><code>int_data</code></strong>, <strong><code>split_level</code></strong>, <strong><code>window</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Filter-Hills">Filter Hills<a class="anchor-link" href="#Filter-Hills"> </a></h3><p>To filter hills, we define a minimum length <code>hill_min_length</code>. All peaks below the threshold <code>hill_peak_min_length</code> are accepted as is. For longer hills, the intensity at the start and the end are compared to the maximum intensity. If the ratio of the maximum raw intensity to the smoothed intensity and the beginning and end are larger than <code>hill_peak_factor</code> the hills are accepted.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="check_large_hills" class="doc_header"><code>check_large_hills</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L369" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>check_large_hills</code>(<strong><code>idx</code></strong>, <strong><code>large_peaks</code></strong>, <strong><code>hill_ptrs</code></strong>, <strong><code>hill_data</code></strong>, <strong><code>int_data</code></strong>, <strong><code>to_remove</code></strong>, <strong><code>large_peak</code></strong>=<em><code>40</code></em>, <strong><code>hill_peak_factor</code></strong>=<em><code>2</code></em>, <strong><code>window</code></strong>=<em><code>1</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="filter_hills" class="doc_header"><code>filter_hills</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L403" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>filter_hills</code>(<strong><code>hill_data</code></strong>, <strong><code>hill_ptrs</code></strong>, <strong><code>int_data</code></strong>, <strong><code>large_peak</code></strong>=<em><code>40</code></em>, <strong><code>window</code></strong>=<em><code>1</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Since the mass estimate min the equation above is more complicated than just an average of the mj, a standard deviation based estimate of the error would not be appropriate. Therefore we calculate the error as a bootstrap2 estimate over B=150 bootstrap replications</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Calculating-Hill-Statistics">Calculating Hill Statistics<a class="anchor-link" href="#Calculating-Hill-Statistics"> </a></h2><p>Next, we calculate summary statistics for the connected centroids. We can obtain a high precision mass estimate for each hill by taking the average of the the masses and weighting this by their intensiteis:</p>
$$
\overline{m} = \frac{\sum_{j=1}^nm_jI_j}{\sum_{j=1}^nI_j}
$$<p>To estimate the mass error, we calculate the error as a boostrap estimate:</p>
<p>{% raw %}
$$\Delta \overline{m} = \sqrt{\frac{\sum_{b=1}^{B}(\overline{m}_b - \overline{m} )}{(B-1)}}$$
{% endraw %}</p>
<p>The calculation of hill statistics for a single hill is implemented in <code>get_hill_stats</code>. To calculate the hill stats for a list of hills, we can call the wrapper <a href="/alphapept/feature_finding.html#get_hill_data"><code>get_hill_data</code></a>.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="hill_stats" class="doc_header"><code>hill_stats</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L432" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>hill_stats</code>(<strong><code>idx</code></strong>, <strong><code>hill_range</code></strong>, <strong><code>hill_ptrs</code></strong>, <strong><code>hill_data</code></strong>, <strong><code>int_data</code></strong>, <strong><code>mass_data</code></strong>, <strong><code>rt_</code></strong>, <strong><code>rt_idx</code></strong>, <strong><code>stats</code></strong>, <strong><code>hill_nboot_max</code></strong>, <strong><code>hill_nboot</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_hill_data" class="doc_header"><code>get_hill_data</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L483" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_hill_data</code>(<strong><code>query_data</code></strong>, <strong><code>hill_ptrs</code></strong>, <strong><code>hill_data</code></strong>, <strong><code>hill_nboot_max</code></strong>=<em><code>300</code></em>, <strong><code>hill_nboot</code></strong>=<em><code>150</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Combining-Hills-to-Isotope-Patterns">Combining Hills to Isotope Patterns<a class="anchor-link" href="#Combining-Hills-to-Isotope-Patterns"> </a></h2><p>After obtaining summary statistics of hills, the next step is to check whether they belong together to form an isotope pattern. For this, we check wheter it is possible that they are neighbors in an isotope pattern, e.g. one having a 12C atom that has been replaced by a 13C version. The detailed criterion for the check is implemented in <a href="/alphapept/feature_finding.html#check_isotope_pattern"><code>check_isotope_pattern</code></a> and is as follows:</p>
<p>{% raw %}
$$\left | \Delta m-\frac{\Delta M}{z} \right |\leq \sqrt{\left ( \frac{\Delta S}{z}  \right )^{2}+\Delta {m_{1}}^{2} +\Delta {m_{2}}^{2}}$$
{% endraw %}</p>
<p>The left side contains $\Delta m$, being the delta of the precise mass estimates from the summary statistics and $\Delta M = 1.00286864$, which is the mass difference ebtween the 13C peak and the monoisotopic peak in an averagine molecule of 1500 Da mass divided by the charge $z$.</p>
<p>The right side contains $\Delta S = 0.0109135$, which is the maximum shift that a sulphur atom can cause ($\Delta S = 2m(^{13}C) - 2m(^{12}C) - m(^{34}S) + m(^{32}S)$) and $\Delta {m_{1}}$ and $\Delta {m_{2}}$, which are the bootstrapped mass standard deviations.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="check_isotope_pattern" class="doc_header"><code>check_isotope_pattern</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L510" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>check_isotope_pattern</code>(<strong><code>mass1</code></strong>, <strong><code>mass2</code></strong>, <strong><code>delta_mass1</code></strong>, <strong><code>delta_mass2</code></strong>, <strong><code>charge</code></strong>, <strong><code>mass_range</code></strong>=<em><code>5</code></em>)</p>
</blockquote>
<p>Check if two masses could belong to the same isotope pattern</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">charge</span> <span class="o">=</span> <span class="mi">1</span>

<span class="n">mass1</span><span class="p">,</span> <span class="n">delta_mass1</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span> <span class="mf">0.1</span>
<span class="n">mass2</span><span class="p">,</span> <span class="n">delta_mass2</span> <span class="o">=</span> <span class="mf">101.1</span><span class="p">,</span> <span class="mf">0.05</span>

<span class="nb">print</span><span class="p">(</span><span class="n">check_isotope_pattern</span><span class="p">(</span><span class="n">mass1</span><span class="p">,</span> <span class="n">mass2</span><span class="p">,</span> <span class="n">delta_mass1</span><span class="p">,</span> <span class="n">delta_mass2</span><span class="p">,</span> <span class="n">charge</span><span class="p">))</span>

<span class="n">mass2</span><span class="p">,</span> <span class="n">delta_mass2</span> <span class="o">=</span> <span class="mf">102.1</span><span class="p">,</span> <span class="mf">0.05</span>

<span class="nb">print</span><span class="p">(</span><span class="n">check_isotope_pattern</span><span class="p">(</span><span class="n">mass1</span><span class="p">,</span> <span class="n">mass2</span><span class="p">,</span> <span class="n">delta_mass1</span><span class="p">,</span> <span class="n">delta_mass2</span><span class="p">,</span> <span class="n">charge</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>True
False
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Cosine-Correlation-of-two-hills">Cosine Correlation of two hills<a class="anchor-link" href="#Cosine-Correlation-of-two-hills"> </a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>An additional criterion that is being checked is that the intensity profiles have sufficient overalp in retention time. This is validated by ensuring that two hills have a cosine correlation of at least 0.6.</p>
<p>{% raw %}
$$\frac{\sum_{s=s_{min}}^{s_{max}}I_sJ_s}{\sum_{s=s_{min}}^{s_{max}}I_s^{2} \sum_{s=s_{min}}^{s_{max}}J_s^{2}} \geq 0.6$$
{% endraw %}</p>
<p>The intensities of two hills are only compared if both have an intensity value in a particular scan. Otherwise, the intensity is set to zero. Additionally, an overlap of at least three elements is required.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="correlate" class="doc_header"><code>correlate</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L527" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>correlate</code>(<strong><code>scans_</code></strong>, <strong><code>scans_2</code></strong>, <strong><code>int_</code></strong>, <strong><code>int_2</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Extracting-pre-Isotope-Patterns">Extracting pre-Isotope Patterns<a class="anchor-link" href="#Extracting-pre-Isotope-Patterns"> </a></h3><p>Now having two criteria to check whether hills could, in principle, belong together, we define the wrapper functions <a href="/alphapept/feature_finding.html#extract_edge"><code>extract_edge</code></a> and <code>get_edges</code> to extract the connected hills. To minimize the number of comparisons we need to perform, we only compare the hills that overlap in time (i.e., the start of one hill <code>rt_min</code> needs to be before the end of the other hill <code>rt_max</code>) and are less than the sum of $\Delta M$ and $\Delta S$ apart.</p>
<p>To extract all hills that belong together, we again rely on the <code>NetworkX</code>-package to extract the connected components.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="extract_edge" class="doc_header"><code>extract_edge</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L554" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>extract_edge</code>(<strong><code>stats</code></strong>, <strong><code>idxs_upper</code></strong>, <strong><code>runner</code></strong>, <strong><code>max_index</code></strong>, <strong><code>maximum_offset</code></strong>, <strong><code>min_charge</code></strong>=<em><code>1</code></em>, <strong><code>max_charge</code></strong>=<em><code>6</code></em>, <strong><code>mass_range</code></strong>=<em><code>5</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="edge_correlation" class="doc_header"><code>edge_correlation</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L572" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>edge_correlation</code>(<strong><code>idx</code></strong>, <strong><code>to_keep</code></strong>, <strong><code>sortindex_</code></strong>, <strong><code>pre_edges</code></strong>, <strong><code>hill_ptrs</code></strong>, <strong><code>hill_data</code></strong>, <strong><code>int_data</code></strong>, <strong><code>scan_idx</code></strong>, <strong><code>cc_cutoff</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_pre_isotope_patterns" class="doc_header"><code>get_pre_isotope_patterns</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L601" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_pre_isotope_patterns</code>(<strong><code>stats</code></strong>, <strong><code>idxs_upper</code></strong>, <strong><code>sortindex_</code></strong>, <strong><code>hill_ptrs</code></strong>, <strong><code>hill_data</code></strong>, <strong><code>int_data</code></strong>, <strong><code>scan_idx</code></strong>, <strong><code>maximum_offset</code></strong>, <strong><code>min_charge</code></strong>=<em><code>1</code></em>, <strong><code>max_charge</code></strong>=<em><code>6</code></em>, <strong><code>mass_range</code></strong>=<em><code>5</code></em>, <strong><code>cc_cutoff</code></strong>=<em><code>0.6</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Extracting-Isotope-Patterns">Extracting Isotope Patterns<a class="anchor-link" href="#Extracting-Isotope-Patterns"> </a></h3><p>The extracted pre-isotope patterns may not be consistent because their pair-wise mass differences may not correspond to the same charge. To extract isotope patterns from pre-isotope patterns, we need to ensure that they are consistent for a single charge.</p>
<p>To do this, we start with the 100 most intense peaks from a pre-isotope pattern to be used as a seed. For each seed and charge we then try to extract the longest consistent isotope pattern. To check wheter a hill is consistent with the seed we employ a modified checking criterion (<a href="/alphapept/feature_finding.html#check_isotope_pattern_directed"><code>check_isotope_pattern_directed</code></a>) to be as follows:</p>
<p>{% raw %}
$$\left | m-m_j-\frac{j\Delta M}{z} \right |\leq \sqrt{\left ( \frac{\Delta S}{z}  \right )^{2}+\Delta {m}^{2} +\Delta {m_{j}}^{2}}$$
{% endraw %}</p>
<p>Here $m$ is the mass of a seed peak, and $m_{j}$ refers to a peak relative to the seed. $j$ refers to the peaks to the left or right (negative or positive index) within the pattern. $j$ needs to run over consecutive values so that gaps are not allowed. Besides this consistency check, two hills are also checked to have a cosine correlation of at least 0.6.</p>
<p>Programmatically, this is implemented in <a href="/alphapept/feature_finding.html#grow_trail"><code>grow_trail</code></a> and <a href="/alphapept/feature_finding.html#grow"><code>grow</code></a>. These function uses a recursive approach that adds matching hills to the seed on the left and right side until no more hills can be added.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="check_isotope_pattern_directed" class="doc_header"><code>check_isotope_pattern_directed</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L627" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>check_isotope_pattern_directed</code>(<strong><code>mass1</code></strong>, <strong><code>mass2</code></strong>, <strong><code>delta_mass1</code></strong>, <strong><code>delta_mass2</code></strong>, <strong><code>charge</code></strong>, <strong><code>index</code></strong>, <strong><code>mass_range</code></strong>)</p>
</blockquote>
<p>Check if two masses could belong to the same isotope pattern</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="grow" class="doc_header"><code>grow</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L642" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>grow</code>(<strong><code>trail</code></strong>, <strong><code>seed</code></strong>, <strong><code>direction</code></strong>, <strong><code>relative_pos</code></strong>, <strong><code>index</code></strong>, <strong><code>stats</code></strong>, <strong><code>pattern</code></strong>, <strong><code>charge</code></strong>, <strong><code>mass_range</code></strong>, <strong><code>sortindex_</code></strong>, <strong><code>hill_ptrs</code></strong>, <strong><code>hill_data</code></strong>, <strong><code>int_data</code></strong>, <strong><code>scan_idx</code></strong>, <strong><code>cc_cutoff</code></strong>)</p>
</blockquote>
<p>Grows isotope pattern based on a seed and direction</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="grow_trail" class="doc_header"><code>grow_trail</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L703" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>grow_trail</code>(<strong><code>seed</code></strong>, <strong><code>pattern</code></strong>, <strong><code>stats</code></strong>, <strong><code>charge</code></strong>, <strong><code>mass_range</code></strong>, <strong><code>sortindex_</code></strong>, <strong><code>hill_ptrs</code></strong>, <strong><code>hill_data</code></strong>, <strong><code>int_data</code></strong>, <strong><code>scan_idx</code></strong>, <strong><code>cc_cutoff</code></strong>)</p>
</blockquote>
<p>Wrapper to grow an isotope pattern to the left and right side</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_trails" class="doc_header"><code>get_trails</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L717" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_trails</code>(<strong><code>seed</code></strong>, <strong><code>pattern</code></strong>, <strong><code>stats</code></strong>, <strong><code>charge_range</code></strong>, <strong><code>mass_range</code></strong>, <strong><code>sortindex_</code></strong>, <strong><code>hill_ptrs</code></strong>, <strong><code>hill_data</code></strong>, <strong><code>int_data</code></strong>, <strong><code>scan_idx</code></strong>, <strong><code>cc_cutoff</code></strong>)</p>
</blockquote>
<p>Wrapper to extract trails for a given charge range</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="plot_pattern" class="doc_header"><code>plot_pattern</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L732" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>plot_pattern</code>(<strong><code>pattern</code></strong>, <strong><code>sorted_hills</code></strong>, <strong><code>centroids</code></strong>, <strong><code>hill_data</code></strong>)</p>
</blockquote>
<p>Helper function to plot a pattern</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_minpos" class="doc_header"><code>get_minpos</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L921" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_minpos</code>(<strong><code>y</code></strong>, <strong><code>split</code></strong>=<em><code>5</code></em>)</p>
</blockquote>
<p>Function to get a list of minima in a trace.
A minimum is returned if the ratio of lower of the surrounding maxima to the minimum is larger than the splitting factor.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_local_minima" class="doc_header"><code>get_local_minima</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L946" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_local_minima</code>(<strong><code>y</code></strong>)</p>
</blockquote>
<p>Function to return all local minima of a array</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="is_local_minima" class="doc_header"><code>is_local_minima</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L958" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>is_local_minima</code>(<strong><code>y</code></strong>, <strong><code>i</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="truncate" class="doc_header"><code>truncate</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L964" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>truncate</code>(<strong><code>array</code></strong>, <strong><code>intensity_profile</code></strong>, <strong><code>seedpos</code></strong>)</p>
</blockquote>
<p>Function to truncate an intensity profile around its seedposition</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Isolating-Isotope_patterns">Isolating Isotope_patterns<a class="anchor-link" href="#Isolating-Isotope_patterns"> </a></h2><p>The extraction of the longest consistent isotope pattern is implemented in <a href="/alphapept/feature_finding.html#isolate_isotope_pattern"><code>isolate_isotope_pattern</code></a>. Here, three additional checks for an isotope pattern are implemented.</p>
<p>The first one is <a href="/alphapept/feature_finding.html#truncate"><code>truncate</code></a>. Here, one checks the seed position, whether it has a minimum to its left or right side. If a minimum is found, the isotope pattern is cut off at this position.</p>
<p>The second one is a mass filter. If the seed has a mass of smaller than 1000, the intensity maximum is detected, and all smaller masses are discarded. This reflects the averagine distribution for small masses where no minimum on the left side can be found.</p>
<p>The third one is <a href="/alphapept/feature_finding.html#check_averagine"><code>check_averagine</code></a> that relies on <a href="/alphapept/feature_finding.html#pattern_to_mz"><code>pattern_to_mz</code></a> and <a href="/alphapept/feature_finding.html#cosine_averagine"><code>cosine_averagine</code></a>. It is used to ensure that the extracted isotope pattern has a cosine correlation of the averagine isotope pattern of the same mass of at least 0.6.</p>
<p>After the longest consistent isotope pattern is found, the hills are removed from the pre-isotope pattern, and the process is repeated until no more isotope patterns can be extracted from the pre-isotope patterns.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="check_averagine" class="doc_header"><code>check_averagine</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L836" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>check_averagine</code>(<strong><code>stats</code></strong>, <strong><code>pattern</code></strong>, <strong><code>charge</code></strong>, <strong><code>averagine_aa</code></strong>, <strong><code>isotopes</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="pattern_to_mz" class="doc_header"><code>pattern_to_mz</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L850" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>pattern_to_mz</code>(<strong><code>stats</code></strong>, <strong><code>pattern</code></strong>, <strong><code>charge</code></strong>)</p>
</blockquote>
<p>Function to calculate masses and intensities from pattern for a given charge</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="cosine_averagine" class="doc_header"><code>cosine_averagine</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L870" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>cosine_averagine</code>(<strong><code>int_one</code></strong>, <strong><code>int_two</code></strong>, <strong><code>spec_one</code></strong>, <strong><code>spec_two</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="int_list_to_array" class="doc_header"><code>int_list_to_array</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L893" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>int_list_to_array</code>(<strong><code>numba_list</code></strong>)</p>
</blockquote>
<p>Numba compatbilte function to convert a numba list with integers to a numpy array</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="mz_to_mass" class="doc_header"><code>mz_to_mass</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L908" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>mz_to_mass</code>(<strong><code>mz</code></strong>, <strong><code>charge</code></strong>)</p>
</blockquote>
<p>Function to calculate the mass from a mz value.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_minpos" class="doc_header"><code>get_minpos</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L921" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_minpos</code>(<strong><code>y</code></strong>, <strong><code>split</code></strong>=<em><code>5</code></em>)</p>
</blockquote>
<p>Function to get a list of minima in a trace.
A minimum is returned if the ratio of lower of the surrounding maxima to the minimum is larger than the splitting factor.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_local_minima" class="doc_header"><code>get_local_minima</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L946" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_local_minima</code>(<strong><code>y</code></strong>)</p>
</blockquote>
<p>Function to return all local minima of a array</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="is_local_minima" class="doc_header"><code>is_local_minima</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L958" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>is_local_minima</code>(<strong><code>y</code></strong>, <strong><code>i</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="truncate" class="doc_header"><code>truncate</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L964" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>truncate</code>(<strong><code>array</code></strong>, <strong><code>intensity_profile</code></strong>, <strong><code>seedpos</code></strong>)</p>
</blockquote>
<p>Function to truncate an intensity profile around its seedposition</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Isotope-Patterns">Isotope Patterns<a class="anchor-link" href="#Isotope-Patterns"> </a></h2><p>The wrapper function <a href="/alphapept/feature_finding.html#get_isotope_patterns"><code>get_isotope_patterns</code></a> iterates over all pre_isotope_patterns.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="isolate_isotope_pattern" class="doc_header"><code>isolate_isotope_pattern</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L992" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>isolate_isotope_pattern</code>(<strong><code>pre_pattern</code></strong>, <strong><code>hill_ptrs</code></strong>, <strong><code>hill_data</code></strong>, <strong><code>int_data</code></strong>, <strong><code>scan_idx</code></strong>, <strong><code>stats</code></strong>, <strong><code>sortindex_</code></strong>, <strong><code>mass_range</code></strong>, <strong><code>charge_range</code></strong>, <strong><code>averagine_aa</code></strong>, <strong><code>isotopes</code></strong>, <strong><code>seed_masses</code></strong>, <strong><code>cc_cutoff</code></strong>)</p>
</blockquote>
<p>Isolate isotope patterns</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">if</span> <span class="kc">False</span><span class="p">:</span>
    <span class="k">def</span> <span class="nf">test_get_isotope_patterns</span><span class="p">():</span>

        <span class="n">test_centroids</span> <span class="o">=</span> <span class="p">[</span>
            <span class="p">[</span>
                <span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">300.501</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">301.003</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">301.504</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">302.006</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span>
            <span class="p">],</span>
            <span class="p">[</span>
                <span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">300.501</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">301.003</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">301.504</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">302.006</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span>
            <span class="p">],</span>
            <span class="p">[</span>
                <span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">300.501</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">301.003</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">301.504</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">302.006</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span>
            <span class="p">],</span>
            <span class="p">[</span>
                <span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">300.501</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">301.003</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">301.504</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">302.006</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span>
            <span class="p">],</span>
            <span class="p">[</span>
                <span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="mi">50</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">300.501</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">301.003</span><span class="p">,</span> <span class="mi">30</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">301.504</span><span class="p">,</span> <span class="mi">20</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
                <span class="p">(</span><span class="mf">302.006</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">),</span>
            <span class="p">],</span>
            <span class="p">[(</span><span class="mi">400</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span> <span class="p">(</span><span class="mi">401</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">),</span> <span class="p">(</span><span class="mi">402</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">6</span><span class="p">)],</span>
            <span class="p">[(</span><span class="mi">400</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">7</span><span class="p">),</span> <span class="p">(</span><span class="mi">401</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">7</span><span class="p">),</span> <span class="p">(</span><span class="mi">402</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">7</span><span class="p">)],</span>
            <span class="p">[(</span><span class="mi">400</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span> <span class="p">(</span><span class="mi">401</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">),</span> <span class="p">(</span><span class="mi">402</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">8</span><span class="p">)],</span>
            <span class="p">[(</span><span class="mi">400</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">9</span><span class="p">),</span> <span class="p">(</span><span class="mi">401</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">9</span><span class="p">),</span> <span class="p">(</span><span class="mi">402</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">9</span><span class="p">)],</span>
        <span class="p">]</span>

        <span class="n">centroid_dtype</span> <span class="o">=</span> <span class="p">[(</span><span class="s2">&quot;mz&quot;</span><span class="p">,</span> <span class="nb">float</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;int&quot;</span><span class="p">,</span> <span class="nb">float</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;scan_no&quot;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;rt&quot;</span><span class="p">,</span> <span class="nb">float</span><span class="p">)]</span>
        <span class="n">test_centroids_tmp</span> <span class="o">=</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">centroid_dtype</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">test_centroids</span><span class="p">]</span>

        <span class="n">test_centroids</span> <span class="o">=</span> <span class="n">List</span><span class="p">([</span><span class="n">_</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">test_centroids_tmp</span><span class="p">])</span>

        <span class="n">test_hills</span> <span class="o">=</span> <span class="n">get_hills</span><span class="p">(</span><span class="n">test_centroids</span><span class="p">)</span>
        <span class="n">sorted_hills</span><span class="p">,</span> <span class="n">stats</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="n">get_hill_data</span><span class="p">(</span><span class="n">test_hills</span><span class="p">,</span> <span class="n">test_centroids</span><span class="p">)</span>
        <span class="n">pre_patterns</span> <span class="o">=</span> <span class="n">get_edges</span><span class="p">(</span><span class="n">stats</span><span class="p">,</span> <span class="n">data</span><span class="p">)</span>

        <span class="n">isotope_patterns</span><span class="p">,</span> <span class="n">isotope_charges</span> <span class="o">=</span> <span class="n">get_isotope_patterns</span><span class="p">(</span><span class="n">pre_patterns</span><span class="p">,</span> <span class="n">stats</span><span class="p">,</span> <span class="n">data</span><span class="p">,</span> <span class="n">averagine_aa</span><span class="p">,</span> <span class="n">isotopes</span><span class="p">)</span>
        <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">isotope_patterns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]))</span>
        <span class="k">assert</span> <span class="n">isotope_charges</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">2</span>
        <span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">isotope_patterns</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">]))</span>
        <span class="k">assert</span> <span class="n">isotope_charges</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span>

    <span class="n">test_get_isotope_patterns</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_isotope_patterns" class="doc_header"><code>get_isotope_patterns</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L1046" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_isotope_patterns</code>(<strong><code>pre_isotope_patterns</code></strong>, <strong><code>hill_ptrs</code></strong>, <strong><code>hill_data</code></strong>, <strong><code>int_data</code></strong>, <strong><code>scan_idx</code></strong>, <strong><code>stats</code></strong>, <strong><code>sortindex_</code></strong>, <strong><code>averagine_aa</code></strong>, <strong><code>isotopes</code></strong>, <strong><code>min_charge</code></strong>=<em><code>1</code></em>, <strong><code>max_charge</code></strong>=<em><code>6</code></em>, <strong><code>mass_range</code></strong>=<em><code>5</code></em>, <strong><code>seed_masses</code></strong>=<em><code>100</code></em>, <strong><code>cc_cutoff</code></strong>=<em><code>0.6</code></em>, <strong><code>callback</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Wrapper function to iterate over pre_isotope_patterns</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="report_" class="doc_header"><code>report_</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L1103" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>report_</code>(<strong><code>idx</code></strong>, <strong><code>isotope_charges</code></strong>, <strong><code>isotope_patterns</code></strong>, <strong><code>iso_idx</code></strong>, <strong><code>stats</code></strong>, <strong><code>sortindex_</code></strong>, <strong><code>hill_ptrs</code></strong>, <strong><code>hill_data</code></strong>, <strong><code>int_data</code></strong>, <strong><code>rt_</code></strong>, <strong><code>rt_idx</code></strong>, <strong><code>results</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="feature_finder_report" class="doc_header"><code>feature_finder_report</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L1216" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>feature_finder_report</code>(<strong><code>query_data</code></strong>, <strong><code>isotope_patterns</code></strong>, <strong><code>isotope_charges</code></strong>, <strong><code>iso_idx</code></strong>, <strong><code>stats</code></strong>, <strong><code>sortindex_</code></strong>, <strong><code>hill_ptrs</code></strong>, <strong><code>hill_data</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Data-Output">Data Output<a class="anchor-link" href="#Data-Output"> </a></h2><p>For each feature that is found we extract summary statistics and put it in tabular form to be used as as pandas dataframe.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Plotting">Plotting<a class="anchor-link" href="#Plotting"> </a></h2><p>For quality control reasons we also employ a function to plot a feature in its local environment.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="plot_isotope_pattern" class="doc_header"><code>plot_isotope_pattern</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L1236" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>plot_isotope_pattern</code>(<strong><code>index</code></strong>, <strong><code>df</code></strong>, <strong><code>sorted_stats</code></strong>, <strong><code>centroids</code></strong>, <strong><code>scan_range</code></strong>=<em><code>100</code></em>, <strong><code>mz_range</code></strong>=<em><code>2</code></em>, <strong><code>plot_hills</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>Plot an isotope pattern in its local environment</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="External-Feature-Finder">External Feature Finder<a class="anchor-link" href="#External-Feature-Finder"> </a></h2><p>To utilize the command-line Feature Finder from Bruker <code>4DFF-3.13</code> - <code>uff-cmdline2.exe</code>, we call it via a subprocess and wait until completion.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="extract_bruker" class="doc_header"><code>extract_bruker</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L1329" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>extract_bruker</code>(<strong><code>file</code></strong>, <strong><code>base_dir</code></strong>=<em><code>'ext/bruker/FF'</code></em>, <strong><code>config</code></strong>=<em><code>'proteomics_4d.config'</code></em>)</p>
</blockquote>
<p>Call Bruker Feautre Finder via subprocess</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="convert_bruker" class="doc_header"><code>convert_bruker</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L1387" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>convert_bruker</code>(<strong><code>feature_path</code></strong>)</p>
</blockquote>
<p>Reads feature table and converts to feature table to be used with AlphaPept</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="map_bruker" class="doc_header"><code>map_bruker</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L1407" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>map_bruker</code>(<strong><code>feature_path</code></strong>, <strong><code>feature_table</code></strong>, <strong><code>query_data</code></strong>)</p>
</blockquote>
<p>Map Ms1 to Ms2 via Table FeaturePrecursorMapping from Bruker FF</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Wrapper">Wrapper<a class="anchor-link" href="#Wrapper"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="find_features" class="doc_header"><code>find_features</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L1468" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>find_features</code>(<strong><code>to_process</code></strong>, <strong><code>callback</code></strong>=<em><code>None</code></em>, <strong><code>parallel</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>Wrapper for feature finding</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Mapping">Mapping<a class="anchor-link" href="#Mapping"> </a></h2><p>Mapping MS1 to MS2</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="map_ms2" class="doc_header"><code>map_ms2</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/feature_finding.py#L1591" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>map_ms2</code>(<strong><code>feature_table</code></strong>, <strong><code>query_data</code></strong>, <strong><code>ppm_range</code></strong>=<em><code>20</code></em>, <strong><code>rt_range</code></strong>=<em><code>0.5</code></em>, <strong><code>mob_range</code></strong>=<em><code>0.3</code></em>, <strong><code>n_neighbors</code></strong>=<em><code>5</code></em>, <strong><code>search_unidentified</code></strong>=<em><code>False</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Map MS1 features to MS2 based on rt and mz
if ccs is included also add</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

</div>
 

