---

title: Input / Output

keywords: fastai
sidebar: home_sidebar

summary: "Functions related to input and output"
description: "Functions related to input and output"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: nbs\02_io.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>This notebook contains all functions related to importing and exporting files. The current way to store raw data is to use the numpy native *.npz container. It allows dictionary-type access and provides reasonable access speeds.</p>
<p>To access proprietary data formats, we have import functions to access <code>Bruker</code> and <code>Thermo</code> data.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Conversion-functions">Conversion functions<a class="anchor-link" href="#Conversion-functions"> </a></h2><p><a href="/alphapept/io#get_most_abundant"><code>get_most_abundant</code></a>: In order to save spectra in a more memory efficient form, we only keep the n most abundant peaks. This allows us to save data in a fast accessible matrix format.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_most_abundant" class="doc_header"><code>get_most_abundant</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L21" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_most_abundant</code>(<strong><code>mass</code></strong>, <strong><code>intensity</code></strong>, <strong><code>n_max</code></strong>)</p>
</blockquote>
<p>Returns the n_max most abundant peaks of a spectrum</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Reading-Thermo-Files">Reading Thermo Files<a class="anchor-link" href="#Reading-Thermo-Files"> </a></h2><blockquote><p>This implementation is based on <code>pymsfilereader</code>. It requires that MSFileReader from Thermo is installed.</p>
<p>The current implementation uses a lot of lists and fills them with list comprehensions. This creates a lot of variables but seems to work reasonably fast. This code could be refactored as all variables end up in a dictionary-type container anyhow.</p>
</blockquote>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="load_thermo_raw" class="doc_header"><code>load_thermo_raw</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L34" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>load_thermo_raw</code>(<strong><code>raw_file</code></strong>, <strong><code>most_abundant</code></strong>, <strong><code>callback</code></strong>=<em><code>None</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Load thermo raw file and extract spectra</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Wrapper">Wrapper<a class="anchor-link" href="#Wrapper"> </a></h2><p>We use <code>multiprocessing - pool</code> to be able to convert multiple files to raw in parallel.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="raw_to_npz" class="doc_header"><code>raw_to_npz</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L137" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>raw_to_npz</code>(<strong><code>to_process</code></strong>, <strong><code>callback</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Wrapper function to convert raw to npz</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="raw_to_npz_parallel" class="doc_header"><code>raw_to_npz_parallel</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L164" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>raw_to_npz_parallel</code>(<strong><code>path_list</code></strong>, <strong><code>settings</code></strong>, <strong><code>callback</code></strong>=<em><code>None</code></em>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Bruker">Bruker<a class="anchor-link" href="#Bruker"> </a></h2><p>For accessing Bruker files, we rely on the external <code>timsdata</code> library. 
For <code>ccs</code> values, we need some functions from this library. As the live feature-finder might not be able to determine some charge values, it is intended to perform this calculation at a later stage once we have charge values from the post-processing feature finder.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="load_bruker_raw" class="doc_header"><code>load_bruker_raw</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L181" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>load_bruker_raw</code>(<strong><code>raw_file</code></strong>, <strong><code>most_abundant</code></strong>, <strong><code>callback</code></strong>=<em><code>None</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Load bruker raw file and extract spectra</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="one_over_k0_to_CCS" class="doc_header"><code>one_over_k0_to_CCS</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L248" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>one_over_k0_to_CCS</code>(<strong><code>one_over_k0s</code></strong>, <strong><code>charges</code></strong>, <strong><code>mzs</code></strong>)</p>
</blockquote>
<p>convert one_over_k0 to CCS</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="MZML">MZML<a class="anchor-link" href="#MZML"> </a></h2><p>To access mzml files, we rely on the pyteomics package.</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="check_sanity" class="doc_header"><code>check_sanity</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L269" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>check_sanity</code>(<strong><code>mass_list</code></strong>)</p>
</blockquote>
<p>Sanity check for mass list to make sure the masses are sorted</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="extract_mzml_info" class="doc_header"><code>extract_mzml_info</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L280" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>extract_mzml_info</code>(<strong><code>input_dict</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="extract_mzxml_info" class="doc_header"><code>extract_mzxml_info</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L297" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>extract_mzxml_info</code>(<strong><code>input_dict</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="read_mzML" class="doc_header"><code>read_mzML</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L310" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>read_mzML</code>(<strong><code>filename</code></strong>, <strong><code>most_abundant</code></strong>)</p>
</blockquote>
<p>Read spectral data from an mzML file and return various lists separately for ms1 and ms2 data.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="read_mzXML" class="doc_header"><code>read_mzXML</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L383" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>read_mzXML</code>(<strong><code>filename</code></strong>, <strong><code>most_abundant</code></strong>)</p>
</blockquote>
<p>Read spectral data from an mzXML file and return various lists separately for ms1 and ms2 data.</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Saving">Saving<a class="anchor-link" href="#Saving"> </a></h2><p>For saving, we are currently relying on the NumPy-native npz-container. It offers reasonable speed, dictionary-type access, and does not need individual type definitions.</p>
<p>While we could, in principle, store the mz and int arrays as a list of variable length, this will come at a performance decrease. We, therefore, create an array of the dimensions of the n most abundant peaks and the number of spectra with the function <a href="/alphapept/io#list_to_numpy_f32"><code>list_to_numpy_f32</code></a> and fill the unoccupied cells with <code>-1</code>. This allows an increase in accessing times at the cost of additional disk space.</p>
<p>Implementation Note: For large files (e.g., choosing a large number of peaks that should be kept, the npz array can fail and trigger an ZIP64 error. This is supposed to be fixed in a later NumPy version.)</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="list_to_numpy_f32" class="doc_header"><code>list_to_numpy_f32</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L457" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>list_to_numpy_f32</code>(<strong><code>long_list</code></strong>)</p>
</blockquote>
<p>Function to convert a list to float32 array</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="save_query_as_npz" class="doc_header"><code>save_query_as_npz</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L474" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>save_query_as_npz</code>(<strong><code>raw_file_npz</code></strong>, <strong><code>query_data</code></strong>)</p>
</blockquote>
<p>Saves query_data as npz</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Parsing-other-Files">Parsing other Files<a class="anchor-link" href="#Parsing-other-Files"> </a></h2><p>Benchmarking proteomics software against each other is not straightforward as various naming conventions exist, and different algorithms are implemented. In this section, we define some helper functions that allow us to facilitate the comparison of different tools.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Reading-MaxQuant-xml-settings-file">Reading MaxQuant xml settings file<a class="anchor-link" href="#Reading-MaxQuant-xml-settings-file"> </a></h3>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="extract_nested" class="doc_header"><code>extract_nested</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L496" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>extract_nested</code>(<strong><code>child</code></strong>)</p>
</blockquote>
<p>Helper function to extract nested entries</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="extract_mq_settings" class="doc_header"><code>extract_mq_settings</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L514" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>extract_mq_settings</code>(<strong><code>path</code></strong>)</p>
</blockquote>
<p>Function to return MaxQuant values as a dictionary for a given xml file</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">mq_dict</span> <span class="o">=</span> <span class="n">extract_mq_settings</span><span class="p">(</span><span class="s1">&#39;../testfiles/test_mqpar.xml&#39;</span><span class="p">)</span>
<span class="n">mq_dict</span><span class="p">[</span><span class="s1">&#39;fastaFiles&#39;</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>{&#39;FastaFileInfo&#39;: {&#39;fastaFilePath&#39;: &#39;testfile.fasta&#39;,
  &#39;identifierParseRule&#39;: &#39;&gt;([^\\s]*)&#39;,
  &#39;descriptionParseRule&#39;: &#39;&gt;(.*)&#39;,
  &#39;taxonomyParseRule&#39;: None,
  &#39;variationParseRule&#39;: None,
  &#39;modificationParseRule&#39;: None,
  &#39;taxonomyId&#39;: None}}</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="parse_mq_seq" class="doc_header"><code>parse_mq_seq</code><a href="https://github.com/mannlabs/alphapept/tree/master/alphapept/io.py#L533" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>parse_mq_seq</code>(<strong><code>peptide</code></strong>)</p>
</blockquote>
<p>Replaces maxquant convention to alphapept convention
ToDo: include more sequences</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">parse_mq_seq</span><span class="p">(</span><span class="s1">&#39;_AFQPFFVELTM(Oxidation (M))PYSVIR_&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">



<div class="output_text output_subarea output_execute_result">
<pre>&#39;AFQPFFVELToxMPYSVIR&#39;</pre>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

